{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8dbd8ad3-9f87-41c3-9881-85e1a05b43d9",
   "metadata": {},
   "source": [
    "# BioRAG-A-Retrieval-Augmented-Agent-for-Scientific-Literature-Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "733d7bc3",
   "metadata": {},
   "source": [
    "## Extraction and Initial Document Creation - Indexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6fb2025c-831a-4b1c-8d74-25efeb21d86d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Jan 17 19:13:39 2026       \n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 570.124.06             Driver Version: 570.124.06     CUDA Version: 12.8     |\n",
      "|-----------------------------------------+------------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                        |               MIG M. |\n",
      "|=========================================+========================+======================|\n",
      "|   0  NVIDIA A100 80GB PCIe          On  |   00000000:CA:00.0 Off |                    0 |\n",
      "| N/A   30C    P0             45W /  300W |       4MiB /  81920MiB |      0%      Default |\n",
      "|                                         |                        |             Disabled |\n",
      "+-----------------------------------------+------------------------+----------------------+\n",
      "                                                                                         \n",
      "+-----------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                              |\n",
      "|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |\n",
      "|        ID   ID                                                               Usage      |\n",
      "|=========================================================================================|\n",
      "|  No running processes found                                                             |\n",
      "+-----------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4de3b380-e5d1-466c-b2fb-5770585dc57f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found existing installation: numpy 1.26.3\n",
      "Uninstalling numpy-1.26.3:\n",
      "  Successfully uninstalled numpy-1.26.3\n",
      "\u001b[33mWARNING: Skipping ml_dtypes as it is not installed.\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Skipping onnx as it is not installed.\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Skipping onnxruntime as it is not installed.\u001b[0m\u001b[33m\n",
      "\u001b[0mCollecting numpy==1.26.4\n",
      "  Downloading numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m61.0/61.0 kB\u001b[0m \u001b[31m2.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.3 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.3/18.3 MB\u001b[0m \u001b[31m110.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: numpy\n",
      "Successfully installed numpy-1.26.4\n",
      "Collecting ml_dtypes==0.2.0\n",
      "  Downloading ml_dtypes-0.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (20 kB)\n",
      "Requirement already satisfied: numpy>1.20 in /opt/conda/lib/python3.11/site-packages (from ml_dtypes==0.2.0) (1.26.4)\n",
      "Downloading ml_dtypes-0.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.0/1.0 MB\u001b[0m \u001b[31m11.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: ml_dtypes\n",
      "Successfully installed ml_dtypes-0.2.0\n",
      "Collecting onnx==1.14.1\n",
      "  Downloading onnx-1.14.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (15 kB)\n",
      "Requirement already satisfied: numpy in /opt/conda/lib/python3.11/site-packages (from onnx==1.14.1) (1.26.4)\n",
      "Requirement already satisfied: protobuf>=3.20.2 in /opt/conda/lib/python3.11/site-packages (from onnx==1.14.1) (4.24.4)\n",
      "Requirement already satisfied: typing-extensions>=3.6.2.1 in /opt/conda/lib/python3.11/site-packages (from onnx==1.14.1) (4.9.0)\n",
      "Downloading onnx-1.14.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (14.6 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.6/14.6 MB\u001b[0m \u001b[31m105.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m0:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: onnx\n",
      "Successfully installed onnx-1.14.1\n",
      "Collecting onnxruntime==1.16.3\n",
      "  Downloading onnxruntime-1.16.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.3 kB)\n",
      "Collecting coloredlogs (from onnxruntime==1.16.3)\n",
      "  Downloading coloredlogs-15.0.1-py2.py3-none-any.whl.metadata (12 kB)\n",
      "Collecting flatbuffers (from onnxruntime==1.16.3)\n",
      "  Downloading flatbuffers-25.12.19-py2.py3-none-any.whl.metadata (1.0 kB)\n",
      "Requirement already satisfied: numpy>=1.24.2 in /opt/conda/lib/python3.11/site-packages (from onnxruntime==1.16.3) (1.26.4)\n",
      "Requirement already satisfied: packaging in /opt/conda/lib/python3.11/site-packages (from onnxruntime==1.16.3) (23.2)\n",
      "Requirement already satisfied: protobuf in /opt/conda/lib/python3.11/site-packages (from onnxruntime==1.16.3) (4.24.4)\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.11/site-packages (from onnxruntime==1.16.3) (1.12)\n",
      "Collecting humanfriendly>=9.1 (from coloredlogs->onnxruntime==1.16.3)\n",
      "  Downloading humanfriendly-10.0-py2.py3-none-any.whl.metadata (9.2 kB)\n",
      "Requirement already satisfied: mpmath>=0.19 in /opt/conda/lib/python3.11/site-packages (from sympy->onnxruntime==1.16.3) (1.3.0)\n",
      "Downloading onnxruntime-1.16.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.4 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.4/6.4 MB\u001b[0m \u001b[31m38.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.0/46.0 kB\u001b[0m \u001b[31m12.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading flatbuffers-25.12.19-py2.py3-none-any.whl (26 kB)\n",
      "Downloading humanfriendly-10.0-py2.py3-none-any.whl (86 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.8/86.8 kB\u001b[0m \u001b[31m17.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: flatbuffers, humanfriendly, coloredlogs, onnxruntime\n",
      "Successfully installed coloredlogs-15.0.1 flatbuffers-25.12.19 humanfriendly-10.0 onnxruntime-1.16.3\n",
      "Found existing installation: scipy 1.12.0\n",
      "Uninstalling scipy-1.12.0:\n",
      "  Successfully uninstalled scipy-1.12.0\n",
      "Collecting scipy==1.10.1\n",
      "  Downloading scipy-1.10.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (58 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.9/58.9 kB\u001b[0m \u001b[31m2.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: numpy<1.27.0,>=1.19.5 in /opt/conda/lib/python3.11/site-packages (from scipy==1.10.1) (1.26.4)\n",
      "Downloading scipy-1.10.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m34.1/34.1 MB\u001b[0m \u001b[31m43.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: scipy\n",
      "Successfully installed scipy-1.10.1\n",
      "Collecting pdf2image\n",
      "  Downloading pdf2image-1.17.0-py3-none-any.whl.metadata (6.2 kB)\n",
      "Requirement already satisfied: pillow in /opt/conda/lib/python3.11/site-packages (from pdf2image) (10.2.0)\n",
      "Downloading pdf2image-1.17.0-py3-none-any.whl (11 kB)\n",
      "Installing collected packages: pdf2image\n",
      "Successfully installed pdf2image-1.17.0\n",
      "Collecting unstructured_inference\n",
      "  Downloading unstructured_inference-1.1.4-py3-none-any.whl.metadata (5.5 kB)\n",
      "Collecting python-multipart (from unstructured_inference)\n",
      "  Downloading python_multipart-0.0.21-py3-none-any.whl.metadata (1.8 kB)\n",
      "Requirement already satisfied: huggingface-hub in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (0.20.3)\n",
      "Requirement already satisfied: numpy in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (1.26.4)\n",
      "Collecting opencv-python>=4.12 (from unstructured_inference)\n",
      "  Downloading opencv_python-4.12.0.88-cp37-abi3-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (19 kB)\n",
      "Requirement already satisfied: onnx in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (1.14.1)\n",
      "Collecting onnxruntime>=1.18.0 (from unstructured_inference)\n",
      "  Downloading onnxruntime-1.23.2-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (5.1 kB)\n",
      "Requirement already satisfied: matplotlib in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (3.8.2)\n",
      "Requirement already satisfied: torch in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (2.1.2)\n",
      "Collecting timm (from unstructured_inference)\n",
      "  Downloading timm-1.0.24-py3-none-any.whl.metadata (38 kB)\n",
      "Requirement already satisfied: transformers>=4.25.1 in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (4.37.1)\n",
      "Collecting accelerate (from unstructured_inference)\n",
      "  Downloading accelerate-1.12.0-py3-none-any.whl.metadata (19 kB)\n",
      "Collecting rapidfuzz (from unstructured_inference)\n",
      "  Downloading rapidfuzz-3.14.3-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (12 kB)\n",
      "Requirement already satisfied: pandas in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (2.2.0)\n",
      "Requirement already satisfied: scipy in /opt/conda/lib/python3.11/site-packages (from unstructured_inference) (1.10.1)\n",
      "Collecting pypdfium2 (from unstructured_inference)\n",
      "  Downloading pypdfium2-5.3.0-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (67 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m67.8/67.8 kB\u001b[0m \u001b[31m16.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting pdfminer-six (from unstructured_inference)\n",
      "  Downloading pdfminer_six-20260107-py3-none-any.whl.metadata (4.3 kB)\n",
      "Requirement already satisfied: coloredlogs in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.18.0->unstructured_inference) (15.0.1)\n",
      "Requirement already satisfied: flatbuffers in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.18.0->unstructured_inference) (25.12.19)\n",
      "Requirement already satisfied: packaging in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.18.0->unstructured_inference) (23.2)\n",
      "Requirement already satisfied: protobuf in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.18.0->unstructured_inference) (4.24.4)\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.18.0->unstructured_inference) (1.12)\n",
      "Collecting numpy (from unstructured_inference)\n",
      "  Downloading numpy-2.2.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m62.0/62.0 kB\u001b[0m \u001b[31m16.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: filelock in /opt/conda/lib/python3.11/site-packages (from transformers>=4.25.1->unstructured_inference) (3.13.1)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /opt/conda/lib/python3.11/site-packages (from transformers>=4.25.1->unstructured_inference) (6.0.1)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /opt/conda/lib/python3.11/site-packages (from transformers>=4.25.1->unstructured_inference) (2023.12.25)\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.11/site-packages (from transformers>=4.25.1->unstructured_inference) (2.31.0)\n",
      "Requirement already satisfied: tokenizers<0.19,>=0.14 in /opt/conda/lib/python3.11/site-packages (from transformers>=4.25.1->unstructured_inference) (0.15.1)\n",
      "Requirement already satisfied: safetensors>=0.3.1 in /opt/conda/lib/python3.11/site-packages (from transformers>=4.25.1->unstructured_inference) (0.4.2)\n",
      "Requirement already satisfied: tqdm>=4.27 in /opt/conda/lib/python3.11/site-packages (from transformers>=4.25.1->unstructured_inference) (4.66.1)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in /opt/conda/lib/python3.11/site-packages (from huggingface-hub->unstructured_inference) (2023.12.2)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /opt/conda/lib/python3.11/site-packages (from huggingface-hub->unstructured_inference) (4.9.0)\n",
      "Requirement already satisfied: psutil in /opt/conda/lib/python3.11/site-packages (from accelerate->unstructured_inference) (5.9.8)\n",
      "Collecting huggingface-hub (from unstructured_inference)\n",
      "  Downloading huggingface_hub-0.36.0-py3-none-any.whl.metadata (14 kB)\n",
      "Collecting safetensors>=0.3.1 (from transformers>=4.25.1->unstructured_inference)\n",
      "  Downloading safetensors-0.7.0-cp38-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.1 kB)\n",
      "Collecting hf-xet<2.0.0,>=1.1.3 (from huggingface-hub->unstructured_inference)\n",
      "  Downloading hf_xet-1.2.0-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.9 kB)\n",
      "Requirement already satisfied: networkx in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (3.2.1)\n",
      "Requirement already satisfied: jinja2 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (3.1.3)\n",
      "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (8.9.2.26)\n",
      "Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (12.1.3.1)\n",
      "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (11.0.2.54)\n",
      "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (10.3.2.106)\n",
      "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (11.4.5.107)\n",
      "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (12.1.0.106)\n",
      "Requirement already satisfied: nvidia-nccl-cu12==2.18.1 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (2.18.1)\n",
      "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (12.1.105)\n",
      "Requirement already satisfied: triton==2.1.0 in /opt/conda/lib/python3.11/site-packages (from torch->unstructured_inference) (2.1.0)\n",
      "Requirement already satisfied: nvidia-nvjitlink-cu12 in /opt/conda/lib/python3.11/site-packages (from nvidia-cusolver-cu12==11.4.5.107->torch->unstructured_inference) (12.3.101)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /opt/conda/lib/python3.11/site-packages (from matplotlib->unstructured_inference) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /opt/conda/lib/python3.11/site-packages (from matplotlib->unstructured_inference) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /opt/conda/lib/python3.11/site-packages (from matplotlib->unstructured_inference) (4.47.2)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /opt/conda/lib/python3.11/site-packages (from matplotlib->unstructured_inference) (1.4.5)\n",
      "INFO: pip is looking at multiple versions of matplotlib to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting matplotlib (from unstructured_inference)\n",
      "  Downloading matplotlib-3.10.8-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (52 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m52.8/52.8 kB\u001b[0m \u001b[31m13.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: pillow>=8 in /opt/conda/lib/python3.11/site-packages (from matplotlib->unstructured_inference) (10.2.0)\n",
      "Requirement already satisfied: pyparsing>=3 in /opt/conda/lib/python3.11/site-packages (from matplotlib->unstructured_inference) (3.1.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /opt/conda/lib/python3.11/site-packages (from matplotlib->unstructured_inference) (2.8.2)\n",
      "INFO: pip is looking at multiple versions of pandas to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting pandas (from unstructured_inference)\n",
      "  Downloading pandas-2.3.3-cp311-cp311-manylinux_2_24_x86_64.manylinux_2_28_x86_64.whl.metadata (91 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m91.2/91.2 kB\u001b[0m \u001b[31m14.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.11/site-packages (from pandas->unstructured_inference) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/conda/lib/python3.11/site-packages (from pandas->unstructured_inference) (2023.4)\n",
      "Requirement already satisfied: charset-normalizer>=2.0.0 in /opt/conda/lib/python3.11/site-packages (from pdfminer-six->unstructured_inference) (3.3.2)\n",
      "Requirement already satisfied: cryptography>=36.0.0 in /opt/conda/lib/python3.11/site-packages (from pdfminer-six->unstructured_inference) (41.0.7)\n",
      "INFO: pip is looking at multiple versions of scipy to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting scipy (from unstructured_inference)\n",
      "  Downloading scipy-1.17.0-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (62 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m62.1/62.1 kB\u001b[0m \u001b[31m15.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: torchvision in /opt/conda/lib/python3.11/site-packages (from timm->unstructured_inference) (0.16.2)\n",
      "INFO: pip is looking at multiple versions of contourpy to determine which version is compatible with other requirements. This could take a while.\n",
      "Collecting contourpy>=1.0.1 (from matplotlib->unstructured_inference)\n",
      "  Downloading contourpy-1.3.3-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (5.5 kB)\n",
      "Requirement already satisfied: cffi>=1.12 in /opt/conda/lib/python3.11/site-packages (from cryptography>=36.0.0->pdfminer-six->unstructured_inference) (1.16.0)\n",
      "Requirement already satisfied: six>=1.5 in /opt/conda/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib->unstructured_inference) (1.16.0)\n",
      "Requirement already satisfied: humanfriendly>=9.1 in /opt/conda/lib/python3.11/site-packages (from coloredlogs->onnxruntime>=1.18.0->unstructured_inference) (10.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /opt/conda/lib/python3.11/site-packages (from jinja2->torch->unstructured_inference) (2.1.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.11/site-packages (from requests->transformers>=4.25.1->unstructured_inference) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.11/site-packages (from requests->transformers>=4.25.1->unstructured_inference) (2.1.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.11/site-packages (from requests->transformers>=4.25.1->unstructured_inference) (2023.11.17)\n",
      "Requirement already satisfied: mpmath>=0.19 in /opt/conda/lib/python3.11/site-packages (from sympy->onnxruntime>=1.18.0->unstructured_inference) (1.3.0)\n",
      "Requirement already satisfied: pycparser in /opt/conda/lib/python3.11/site-packages (from cffi>=1.12->cryptography>=36.0.0->pdfminer-six->unstructured_inference) (2.21)\n",
      "Downloading unstructured_inference-1.1.4-py3-none-any.whl (47 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m47.9/47.9 kB\u001b[0m \u001b[31m12.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading onnxruntime-1.23.2-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (17.4 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m17.4/17.4 MB\u001b[0m \u001b[31m99.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading opencv_python-4.12.0.88-cp37-abi3-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (67.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m67.0/67.0 MB\u001b[0m \u001b[31m60.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading numpy-2.2.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m16.8/16.8 MB\u001b[0m \u001b[31m105.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading accelerate-1.12.0-py3-none-any.whl (380 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m380.9/380.9 kB\u001b[0m \u001b[31m80.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading huggingface_hub-0.36.0-py3-none-any.whl (566 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m566.1/566.1 kB\u001b[0m \u001b[31m82.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading matplotlib-3.10.8-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (8.7 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.7/8.7 MB\u001b[0m \u001b[31m101.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading pandas-2.3.3-cp311-cp311-manylinux_2_24_x86_64.manylinux_2_28_x86_64.whl (12.8 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.8/12.8 MB\u001b[0m \u001b[31m97.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m0:01\u001b[0m\n",
      "\u001b[?25hDownloading pdfminer_six-20260107-py3-none-any.whl (6.6 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.6/6.6 MB\u001b[0m \u001b[31m114.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading pypdfium2-5.3.0-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.0/3.0 MB\u001b[0m \u001b[31m158.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading python_multipart-0.0.21-py3-none-any.whl (24 kB)\n",
      "Downloading rapidfuzz-3.14.3-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m111.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading scipy-1.17.0-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (35.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m35.1/35.1 MB\u001b[0m \u001b[31m80.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hDownloading timm-1.0.24-py3-none-any.whl (2.6 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.6/2.6 MB\u001b[0m \u001b[31m144.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading contourpy-1.3.3-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (355 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m355.2/355.2 kB\u001b[0m \u001b[31m77.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading hf_xet-1.2.0-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m114.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading safetensors-0.7.0-cp38-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (507 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m507.2/507.2 kB\u001b[0m \u001b[31m53.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: safetensors, rapidfuzz, python-multipart, pypdfium2, numpy, hf-xet, scipy, pandas, opencv-python, onnxruntime, huggingface-hub, contourpy, pdfminer-six, matplotlib, accelerate, timm, unstructured_inference\n",
      "  Attempting uninstall: safetensors\n",
      "    Found existing installation: safetensors 0.4.2\n",
      "    Uninstalling safetensors-0.4.2:\n",
      "      Successfully uninstalled safetensors-0.4.2\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 1.26.4\n",
      "    Uninstalling numpy-1.26.4:\n",
      "      Successfully uninstalled numpy-1.26.4\n",
      "  Attempting uninstall: scipy\n",
      "    Found existing installation: scipy 1.10.1\n",
      "    Uninstalling scipy-1.10.1:\n",
      "      Successfully uninstalled scipy-1.10.1\n",
      "  Attempting uninstall: pandas\n",
      "    Found existing installation: pandas 2.2.0\n",
      "    Uninstalling pandas-2.2.0:\n",
      "      Successfully uninstalled pandas-2.2.0\n",
      "  Attempting uninstall: onnxruntime\n",
      "    Found existing installation: onnxruntime 1.16.3\n",
      "    Uninstalling onnxruntime-1.16.3:\n",
      "      Successfully uninstalled onnxruntime-1.16.3\n",
      "  Attempting uninstall: huggingface-hub\n",
      "    Found existing installation: huggingface-hub 0.20.3\n",
      "    Uninstalling huggingface-hub-0.20.3:\n",
      "      Successfully uninstalled huggingface-hub-0.20.3\n",
      "  Attempting uninstall: contourpy\n",
      "    Found existing installation: contourpy 1.2.0\n",
      "    Uninstalling contourpy-1.2.0:\n",
      "      Successfully uninstalled contourpy-1.2.0\n",
      "  Attempting uninstall: matplotlib\n",
      "    Found existing installation: matplotlib 3.8.2\n",
      "    Uninstalling matplotlib-3.8.2:\n",
      "      Successfully uninstalled matplotlib-3.8.2\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "scikit-learn 1.4.0 requires numpy<2.0,>=1.19.5, but you have numpy 2.2.6 which is incompatible.\n",
      "numba 0.58.1 requires numpy<1.27,>=1.22, but you have numpy 2.2.6 which is incompatible.\n",
      "statsmodels 0.14.1 requires numpy<2,>=1.18, but you have numpy 2.2.6 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed accelerate-1.12.0 contourpy-1.3.3 hf-xet-1.2.0 huggingface-hub-0.36.0 matplotlib-3.10.8 numpy-2.2.6 onnxruntime-1.23.2 opencv-python-4.12.0.88 pandas-2.3.3 pdfminer-six-20260107 pypdfium2-5.3.0 python-multipart-0.0.21 rapidfuzz-3.14.3 safetensors-0.7.0 scipy-1.17.0 timm-1.0.24 unstructured_inference-1.1.4\n",
      "Collecting langchain[community]\n",
      "  Downloading langchain-1.2.4-py3-none-any.whl.metadata (4.9 kB)\n",
      "Collecting langchain-core<2.0.0,>=1.2.1 (from langchain[community])\n",
      "  Downloading langchain_core-1.2.7-py3-none-any.whl.metadata (3.7 kB)\n",
      "Collecting langgraph<1.1.0,>=1.0.2 (from langchain[community])\n",
      "  Downloading langgraph-1.0.6-py3-none-any.whl.metadata (7.4 kB)\n",
      "Collecting pydantic<3.0.0,>=2.7.4 (from langchain[community])\n",
      "  Downloading pydantic-2.12.5-py3-none-any.whl.metadata (90 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m90.6/90.6 kB\u001b[0m \u001b[31m1.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting langchain-community (from langchain[community])\n",
      "  Downloading langchain_community-0.4.1-py3-none-any.whl.metadata (3.0 kB)\n",
      "Requirement already satisfied: jsonpatch<2.0.0,>=1.33.0 in /opt/conda/lib/python3.11/site-packages (from langchain-core<2.0.0,>=1.2.1->langchain[community]) (1.33)\n",
      "Collecting langsmith<1.0.0,>=0.3.45 (from langchain-core<2.0.0,>=1.2.1->langchain[community])\n",
      "  Downloading langsmith-0.6.4-py3-none-any.whl.metadata (15 kB)\n",
      "Requirement already satisfied: packaging<26.0.0,>=23.2.0 in /opt/conda/lib/python3.11/site-packages (from langchain-core<2.0.0,>=1.2.1->langchain[community]) (23.2)\n",
      "Requirement already satisfied: pyyaml<7.0.0,>=5.3.0 in /opt/conda/lib/python3.11/site-packages (from langchain-core<2.0.0,>=1.2.1->langchain[community]) (6.0.1)\n",
      "Collecting tenacity!=8.4.0,<10.0.0,>=8.1.0 (from langchain-core<2.0.0,>=1.2.1->langchain[community])\n",
      "  Downloading tenacity-9.1.2-py3-none-any.whl.metadata (1.2 kB)\n",
      "Requirement already satisfied: typing-extensions<5.0.0,>=4.7.0 in /opt/conda/lib/python3.11/site-packages (from langchain-core<2.0.0,>=1.2.1->langchain[community]) (4.9.0)\n",
      "Collecting uuid-utils<1.0,>=0.12.0 (from langchain-core<2.0.0,>=1.2.1->langchain[community])\n",
      "  Downloading uuid_utils-0.13.0-cp39-abi3-manylinux_2_24_x86_64.whl.metadata (5.4 kB)\n",
      "Collecting langgraph-checkpoint<5.0.0,>=2.1.0 (from langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading langgraph_checkpoint-4.0.0-py3-none-any.whl.metadata (4.9 kB)\n",
      "Collecting langgraph-prebuilt<1.1.0,>=1.0.2 (from langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading langgraph_prebuilt-1.0.6-py3-none-any.whl.metadata (5.2 kB)\n",
      "Collecting langgraph-sdk<0.4.0,>=0.3.0 (from langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading langgraph_sdk-0.3.3-py3-none-any.whl.metadata (1.6 kB)\n",
      "Collecting xxhash>=3.5.0 (from langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading xxhash-3.6.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
      "Collecting annotated-types>=0.6.0 (from pydantic<3.0.0,>=2.7.4->langchain[community])\n",
      "  Downloading annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)\n",
      "Collecting pydantic-core==2.41.5 (from pydantic<3.0.0,>=2.7.4->langchain[community])\n",
      "  Downloading pydantic_core-2.41.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (7.3 kB)\n",
      "Collecting typing-extensions<5.0.0,>=4.7.0 (from langchain-core<2.0.0,>=1.2.1->langchain[community])\n",
      "  Downloading typing_extensions-4.15.0-py3-none-any.whl.metadata (3.3 kB)\n",
      "Collecting typing-inspection>=0.4.2 (from pydantic<3.0.0,>=2.7.4->langchain[community])\n",
      "  Downloading typing_inspection-0.4.2-py3-none-any.whl.metadata (2.6 kB)\n",
      "Collecting langchain-classic<2.0.0,>=1.0.0 (from langchain-community->langchain[community])\n",
      "  Downloading langchain_classic-1.0.1-py3-none-any.whl.metadata (4.2 kB)\n",
      "Requirement already satisfied: SQLAlchemy<3.0.0,>=1.4.0 in /opt/conda/lib/python3.11/site-packages (from langchain-community->langchain[community]) (2.0.25)\n",
      "Collecting requests<3.0.0,>=2.32.5 (from langchain-community->langchain[community])\n",
      "  Downloading requests-2.32.5-py3-none-any.whl.metadata (4.9 kB)\n",
      "Collecting aiohttp<4.0.0,>=3.8.3 (from langchain-community->langchain[community])\n",
      "  Downloading aiohttp-3.13.3-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (8.1 kB)\n",
      "Collecting dataclasses-json<0.7.0,>=0.6.7 (from langchain-community->langchain[community])\n",
      "  Downloading dataclasses_json-0.6.7-py3-none-any.whl.metadata (25 kB)\n",
      "Collecting pydantic-settings<3.0.0,>=2.10.1 (from langchain-community->langchain[community])\n",
      "  Downloading pydantic_settings-2.12.0-py3-none-any.whl.metadata (3.4 kB)\n",
      "Collecting httpx-sse<1.0.0,>=0.4.0 (from langchain-community->langchain[community])\n",
      "  Downloading httpx_sse-0.4.3-py3-none-any.whl.metadata (9.7 kB)\n",
      "Requirement already satisfied: numpy>=1.26.2 in /opt/conda/lib/python3.11/site-packages (from langchain-community->langchain[community]) (2.2.6)\n",
      "Collecting aiohappyeyeballs>=2.5.0 (from aiohttp<4.0.0,>=3.8.3->langchain-community->langchain[community])\n",
      "  Downloading aiohappyeyeballs-2.6.1-py3-none-any.whl.metadata (5.9 kB)\n",
      "Collecting aiosignal>=1.4.0 (from aiohttp<4.0.0,>=3.8.3->langchain-community->langchain[community])\n",
      "  Downloading aiosignal-1.4.0-py3-none-any.whl.metadata (3.7 kB)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /opt/conda/lib/python3.11/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain-community->langchain[community]) (23.2.0)\n",
      "Collecting frozenlist>=1.1.1 (from aiohttp<4.0.0,>=3.8.3->langchain-community->langchain[community])\n",
      "  Downloading frozenlist-1.8.0-cp311-cp311-manylinux1_x86_64.manylinux_2_28_x86_64.manylinux_2_5_x86_64.whl.metadata (20 kB)\n",
      "Collecting multidict<7.0,>=4.5 (from aiohttp<4.0.0,>=3.8.3->langchain-community->langchain[community])\n",
      "  Downloading multidict-6.7.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (5.3 kB)\n",
      "Collecting propcache>=0.2.0 (from aiohttp<4.0.0,>=3.8.3->langchain-community->langchain[community])\n",
      "  Downloading propcache-0.4.1-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
      "Collecting yarl<2.0,>=1.17.0 (from aiohttp<4.0.0,>=3.8.3->langchain-community->langchain[community])\n",
      "  Downloading yarl-1.22.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (75 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m75.1/75.1 kB\u001b[0m \u001b[31m10.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting marshmallow<4.0.0,>=3.18.0 (from dataclasses-json<0.7.0,>=0.6.7->langchain-community->langchain[community])\n",
      "  Downloading marshmallow-3.26.2-py3-none-any.whl.metadata (7.3 kB)\n",
      "Collecting typing-inspect<1,>=0.4.0 (from dataclasses-json<0.7.0,>=0.6.7->langchain-community->langchain[community])\n",
      "  Downloading typing_inspect-0.9.0-py3-none-any.whl.metadata (1.5 kB)\n",
      "Requirement already satisfied: jsonpointer>=1.9 in /opt/conda/lib/python3.11/site-packages (from jsonpatch<2.0.0,>=1.33.0->langchain-core<2.0.0,>=1.2.1->langchain[community]) (2.4)\n",
      "Collecting langchain-text-splitters<2.0.0,>=1.1.0 (from langchain-classic<2.0.0,>=1.0.0->langchain-community->langchain[community])\n",
      "  Downloading langchain_text_splitters-1.1.0-py3-none-any.whl.metadata (2.7 kB)\n",
      "Collecting ormsgpack>=1.12.0 (from langgraph-checkpoint<5.0.0,>=2.1.0->langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading ormsgpack-1.12.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.2 kB)\n",
      "Collecting httpx>=0.25.2 (from langgraph-sdk<0.4.0,>=0.3.0->langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)\n",
      "Collecting orjson>=3.10.1 (from langgraph-sdk<0.4.0,>=0.3.0->langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading orjson-3.11.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (41 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m41.6/41.6 kB\u001b[0m \u001b[31m8.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting requests-toolbelt>=1.0.0 (from langsmith<1.0.0,>=0.3.45->langchain-core<2.0.0,>=1.2.1->langchain[community])\n",
      "  Downloading requests_toolbelt-1.0.0-py2.py3-none-any.whl.metadata (14 kB)\n",
      "Collecting zstandard>=0.23.0 (from langsmith<1.0.0,>=0.3.45->langchain-core<2.0.0,>=1.2.1->langchain[community])\n",
      "  Downloading zstandard-0.25.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (3.3 kB)\n",
      "Collecting python-dotenv>=0.21.0 (from pydantic-settings<3.0.0,>=2.10.1->langchain-community->langchain[community])\n",
      "  Downloading python_dotenv-1.2.1-py3-none-any.whl.metadata (25 kB)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /opt/conda/lib/python3.11/site-packages (from requests<3.0.0,>=2.32.5->langchain-community->langchain[community]) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.11/site-packages (from requests<3.0.0,>=2.32.5->langchain-community->langchain[community]) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.11/site-packages (from requests<3.0.0,>=2.32.5->langchain-community->langchain[community]) (2.1.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.11/site-packages (from requests<3.0.0,>=2.32.5->langchain-community->langchain[community]) (2023.11.17)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /opt/conda/lib/python3.11/site-packages (from SQLAlchemy<3.0.0,>=1.4.0->langchain-community->langchain[community]) (3.0.3)\n",
      "Requirement already satisfied: anyio in /opt/conda/lib/python3.11/site-packages (from httpx>=0.25.2->langgraph-sdk<0.4.0,>=0.3.0->langgraph<1.1.0,>=1.0.2->langchain[community]) (4.2.0)\n",
      "Collecting httpcore==1.* (from httpx>=0.25.2->langgraph-sdk<0.4.0,>=0.3.0->langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading httpcore-1.0.9-py3-none-any.whl.metadata (21 kB)\n",
      "Collecting h11>=0.16 (from httpcore==1.*->httpx>=0.25.2->langgraph-sdk<0.4.0,>=0.3.0->langgraph<1.1.0,>=1.0.2->langchain[community])\n",
      "  Downloading h11-0.16.0-py3-none-any.whl.metadata (8.3 kB)\n",
      "Collecting mypy-extensions>=0.3.0 (from typing-inspect<1,>=0.4.0->dataclasses-json<0.7.0,>=0.6.7->langchain-community->langchain[community])\n",
      "  Downloading mypy_extensions-1.1.0-py3-none-any.whl.metadata (1.1 kB)\n",
      "Requirement already satisfied: sniffio>=1.1 in /opt/conda/lib/python3.11/site-packages (from anyio->httpx>=0.25.2->langgraph-sdk<0.4.0,>=0.3.0->langgraph<1.1.0,>=1.0.2->langchain[community]) (1.3.0)\n",
      "Downloading langchain_core-1.2.7-py3-none-any.whl (490 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m490.2/490.2 kB\u001b[0m \u001b[31m9.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mta \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading langgraph-1.0.6-py3-none-any.whl (157 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m157.4/157.4 kB\u001b[0m \u001b[31m24.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pydantic-2.12.5-py3-none-any.whl (463 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m463.6/463.6 kB\u001b[0m \u001b[31m20.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pydantic_core-2.41.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.1 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m21.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading langchain-1.2.4-py3-none-any.whl (107 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m107.9/107.9 kB\u001b[0m \u001b[31m17.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading langchain_community-0.4.1-py3-none-any.whl (2.5 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.5/2.5 MB\u001b[0m \u001b[31m26.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading aiohttp-3.13.3-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (1.7 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.7/1.7 MB\u001b[0m \u001b[31m32.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading annotated_types-0.7.0-py3-none-any.whl (13 kB)\n",
      "Downloading dataclasses_json-0.6.7-py3-none-any.whl (28 kB)\n",
      "Downloading httpx_sse-0.4.3-py3-none-any.whl (9.0 kB)\n",
      "Downloading langchain_classic-1.0.1-py3-none-any.whl (1.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.0/1.0 MB\u001b[0m \u001b[31m33.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading langgraph_checkpoint-4.0.0-py3-none-any.whl (46 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.3/46.3 kB\u001b[0m \u001b[31m12.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading langgraph_prebuilt-1.0.6-py3-none-any.whl (35 kB)\n",
      "Downloading langgraph_sdk-0.3.3-py3-none-any.whl (67 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m67.0/67.0 kB\u001b[0m \u001b[31m11.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading langsmith-0.6.4-py3-none-any.whl (283 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m283.5/283.5 kB\u001b[0m \u001b[31m53.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading pydantic_settings-2.12.0-py3-none-any.whl (51 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m51.9/51.9 kB\u001b[0m \u001b[31m12.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading requests-2.32.5-py3-none-any.whl (64 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m64.7/64.7 kB\u001b[0m \u001b[31m16.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading tenacity-9.1.2-py3-none-any.whl (28 kB)\n",
      "Downloading typing_extensions-4.15.0-py3-none-any.whl (44 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m44.6/44.6 kB\u001b[0m \u001b[31m8.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading typing_inspection-0.4.2-py3-none-any.whl (14 kB)\n",
      "Downloading uuid_utils-0.13.0-cp39-abi3-manylinux_2_24_x86_64.whl (342 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m342.1/342.1 kB\u001b[0m \u001b[31m52.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading xxhash-3.6.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (193 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m193.9/193.9 kB\u001b[0m \u001b[31m43.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading aiohappyeyeballs-2.6.1-py3-none-any.whl (15 kB)\n",
      "Downloading aiosignal-1.4.0-py3-none-any.whl (7.5 kB)\n",
      "Downloading frozenlist-1.8.0-cp311-cp311-manylinux1_x86_64.manylinux_2_28_x86_64.manylinux_2_5_x86_64.whl (231 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m231.1/231.1 kB\u001b[0m \u001b[31m33.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading httpx-0.28.1-py3-none-any.whl (73 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m73.5/73.5 kB\u001b[0m \u001b[31m17.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading httpcore-1.0.9-py3-none-any.whl (78 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m78.8/78.8 kB\u001b[0m \u001b[31m19.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading langchain_text_splitters-1.1.0-py3-none-any.whl (34 kB)\n",
      "Downloading marshmallow-3.26.2-py3-none-any.whl (50 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m51.0/51.0 kB\u001b[0m \u001b[31m11.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading multidict-6.7.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (246 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m246.7/246.7 kB\u001b[0m \u001b[31m46.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading orjson-3.11.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (138 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m138.8/138.8 kB\u001b[0m \u001b[31m25.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading ormsgpack-1.12.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (211 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m211.3/211.3 kB\u001b[0m \u001b[31m40.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading propcache-0.4.1-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (210 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m210.0/210.0 kB\u001b[0m \u001b[31m50.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading python_dotenv-1.2.1-py3-none-any.whl (21 kB)\n",
      "Downloading requests_toolbelt-1.0.0-py2.py3-none-any.whl (54 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m54.5/54.5 kB\u001b[0m \u001b[31m12.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading typing_inspect-0.9.0-py3-none-any.whl (8.8 kB)\n",
      "Downloading yarl-1.22.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (365 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m365.8/365.8 kB\u001b[0m \u001b[31m49.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading zstandard-0.25.0-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (5.6 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.6/5.6 MB\u001b[0m \u001b[31m40.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading mypy_extensions-1.1.0-py3-none-any.whl (5.0 kB)\n",
      "Downloading h11-0.16.0-py3-none-any.whl (37 kB)\n",
      "Installing collected packages: zstandard, xxhash, uuid-utils, typing-extensions, tenacity, requests, python-dotenv, propcache, ormsgpack, orjson, mypy-extensions, multidict, marshmallow, httpx-sse, h11, frozenlist, annotated-types, aiohappyeyeballs, yarl, typing-inspection, typing-inspect, requests-toolbelt, pydantic-core, httpcore, aiosignal, pydantic, httpx, dataclasses-json, aiohttp, pydantic-settings, langsmith, langgraph-sdk, langchain-core, langgraph-checkpoint, langchain-text-splitters, langgraph-prebuilt, langchain-classic, langgraph, langchain-community, langchain\n",
      "  Attempting uninstall: zstandard\n",
      "    Found existing installation: zstandard 0.22.0\n",
      "    Uninstalling zstandard-0.22.0:\n",
      "      Successfully uninstalled zstandard-0.22.0\n",
      "  Attempting uninstall: typing-extensions\n",
      "    Found existing installation: typing_extensions 4.9.0\n",
      "    Uninstalling typing_extensions-4.9.0:\n",
      "      Successfully uninstalled typing_extensions-4.9.0\n",
      "  Attempting uninstall: requests\n",
      "    Found existing installation: requests 2.31.0\n",
      "    Uninstalling requests-2.31.0:\n",
      "      Successfully uninstalled requests-2.31.0\n",
      "Successfully installed aiohappyeyeballs-2.6.1 aiohttp-3.13.3 aiosignal-1.4.0 annotated-types-0.7.0 dataclasses-json-0.6.7 frozenlist-1.8.0 h11-0.16.0 httpcore-1.0.9 httpx-0.28.1 httpx-sse-0.4.3 langchain-1.2.4 langchain-classic-1.0.1 langchain-community-0.4.1 langchain-core-1.2.7 langchain-text-splitters-1.1.0 langgraph-1.0.6 langgraph-checkpoint-4.0.0 langgraph-prebuilt-1.0.6 langgraph-sdk-0.3.3 langsmith-0.6.4 marshmallow-3.26.2 multidict-6.7.0 mypy-extensions-1.1.0 orjson-3.11.5 ormsgpack-1.12.1 propcache-0.4.1 pydantic-2.12.5 pydantic-core-2.41.5 pydantic-settings-2.12.0 python-dotenv-1.2.1 requests-2.32.5 requests-toolbelt-1.0.0 tenacity-9.1.2 typing-extensions-4.15.0 typing-inspect-0.9.0 typing-inspection-0.4.2 uuid-utils-0.13.0 xxhash-3.6.0 yarl-1.22.0 zstandard-0.25.0\n",
      "Collecting unstructured[all]\n",
      "  Downloading unstructured-0.18.27-py3-none-any.whl.metadata (25 kB)\n",
      "\u001b[33mWARNING: unstructured 0.18.27 does not provide the extra 'all'\u001b[0m\u001b[33m\n",
      "\u001b[0mRequirement already satisfied: charset-normalizer in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (3.3.2)\n",
      "Collecting filetype (from unstructured[all])\n",
      "  Downloading filetype-1.2.0-py2.py3-none-any.whl.metadata (6.5 kB)\n",
      "Collecting python-magic (from unstructured[all])\n",
      "  Downloading python_magic-0.4.27-py2.py3-none-any.whl.metadata (5.8 kB)\n",
      "Collecting lxml (from unstructured[all])\n",
      "  Downloading lxml-6.0.2-cp311-cp311-manylinux_2_26_x86_64.manylinux_2_28_x86_64.whl.metadata (3.6 kB)\n",
      "Collecting nltk (from unstructured[all])\n",
      "  Downloading nltk-3.9.2-py3-none-any.whl.metadata (3.2 kB)\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (2.32.5)\n",
      "Requirement already satisfied: beautifulsoup4 in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (4.12.3)\n",
      "Collecting emoji (from unstructured[all])\n",
      "  Downloading emoji-2.15.0-py3-none-any.whl.metadata (5.7 kB)\n",
      "Requirement already satisfied: dataclasses-json in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (0.6.7)\n",
      "Collecting python-iso639 (from unstructured[all])\n",
      "  Downloading python_iso639-2025.11.16-py3-none-any.whl.metadata (15 kB)\n",
      "Collecting langdetect (from unstructured[all])\n",
      "  Downloading langdetect-1.0.9.tar.gz (981 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m981.5/981.5 kB\u001b[0m \u001b[31m11.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: numpy in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (2.2.6)\n",
      "Requirement already satisfied: rapidfuzz in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (3.14.3)\n",
      "Collecting backoff (from unstructured[all])\n",
      "  Downloading backoff-2.2.1-py3-none-any.whl.metadata (14 kB)\n",
      "Requirement already satisfied: typing-extensions in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (4.15.0)\n",
      "Collecting unstructured-client (from unstructured[all])\n",
      "  Downloading unstructured_client-0.42.8-py3-none-any.whl.metadata (23 kB)\n",
      "Collecting wrapt (from unstructured[all])\n",
      "  Downloading wrapt-2.0.1-cp311-cp311-manylinux1_x86_64.manylinux_2_28_x86_64.manylinux_2_5_x86_64.whl.metadata (9.0 kB)\n",
      "Requirement already satisfied: tqdm in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (4.66.1)\n",
      "Requirement already satisfied: psutil in /opt/conda/lib/python3.11/site-packages (from unstructured[all]) (5.9.8)\n",
      "Collecting python-oxmsg (from unstructured[all])\n",
      "  Downloading python_oxmsg-0.0.2-py3-none-any.whl.metadata (5.0 kB)\n",
      "Collecting html5lib (from unstructured[all])\n",
      "  Downloading html5lib-1.1-py2.py3-none-any.whl.metadata (16 kB)\n",
      "Requirement already satisfied: soupsieve>1.2 in /opt/conda/lib/python3.11/site-packages (from beautifulsoup4->unstructured[all]) (2.5)\n",
      "Requirement already satisfied: marshmallow<4.0.0,>=3.18.0 in /opt/conda/lib/python3.11/site-packages (from dataclasses-json->unstructured[all]) (3.26.2)\n",
      "Requirement already satisfied: typing-inspect<1,>=0.4.0 in /opt/conda/lib/python3.11/site-packages (from dataclasses-json->unstructured[all]) (0.9.0)\n",
      "Requirement already satisfied: six>=1.9 in /opt/conda/lib/python3.11/site-packages (from html5lib->unstructured[all]) (1.16.0)\n",
      "Requirement already satisfied: webencodings in /opt/conda/lib/python3.11/site-packages (from html5lib->unstructured[all]) (0.5.1)\n",
      "Requirement already satisfied: click in /opt/conda/lib/python3.11/site-packages (from nltk->unstructured[all]) (8.1.7)\n",
      "Requirement already satisfied: joblib in /opt/conda/lib/python3.11/site-packages (from nltk->unstructured[all]) (1.3.2)\n",
      "Requirement already satisfied: regex>=2021.8.3 in /opt/conda/lib/python3.11/site-packages (from nltk->unstructured[all]) (2023.12.25)\n",
      "Collecting olefile (from python-oxmsg->unstructured[all])\n",
      "  Downloading olefile-0.47-py2.py3-none-any.whl.metadata (9.7 kB)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.11/site-packages (from requests->unstructured[all]) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/conda/lib/python3.11/site-packages (from requests->unstructured[all]) (2.1.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.11/site-packages (from requests->unstructured[all]) (2023.11.17)\n",
      "Collecting aiofiles>=24.1.0 (from unstructured-client->unstructured[all])\n",
      "  Downloading aiofiles-25.1.0-py3-none-any.whl.metadata (6.3 kB)\n",
      "Requirement already satisfied: cryptography>=3.1 in /opt/conda/lib/python3.11/site-packages (from unstructured-client->unstructured[all]) (41.0.7)\n",
      "Requirement already satisfied: httpcore>=1.0.9 in /opt/conda/lib/python3.11/site-packages (from unstructured-client->unstructured[all]) (1.0.9)\n",
      "Requirement already satisfied: httpx>=0.27.0 in /opt/conda/lib/python3.11/site-packages (from unstructured-client->unstructured[all]) (0.28.1)\n",
      "Requirement already satisfied: pydantic>=2.11.2 in /opt/conda/lib/python3.11/site-packages (from unstructured-client->unstructured[all]) (2.12.5)\n",
      "Collecting pypdf>=6.2.0 (from unstructured-client->unstructured[all])\n",
      "  Downloading pypdf-6.6.0-py3-none-any.whl.metadata (7.1 kB)\n",
      "Requirement already satisfied: requests-toolbelt>=1.0.0 in /opt/conda/lib/python3.11/site-packages (from unstructured-client->unstructured[all]) (1.0.0)\n",
      "Requirement already satisfied: cffi>=1.12 in /opt/conda/lib/python3.11/site-packages (from cryptography>=3.1->unstructured-client->unstructured[all]) (1.16.0)\n",
      "Requirement already satisfied: h11>=0.16 in /opt/conda/lib/python3.11/site-packages (from httpcore>=1.0.9->unstructured-client->unstructured[all]) (0.16.0)\n",
      "Requirement already satisfied: anyio in /opt/conda/lib/python3.11/site-packages (from httpx>=0.27.0->unstructured-client->unstructured[all]) (4.2.0)\n",
      "Requirement already satisfied: packaging>=17.0 in /opt/conda/lib/python3.11/site-packages (from marshmallow<4.0.0,>=3.18.0->dataclasses-json->unstructured[all]) (23.2)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in /opt/conda/lib/python3.11/site-packages (from pydantic>=2.11.2->unstructured-client->unstructured[all]) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.41.5 in /opt/conda/lib/python3.11/site-packages (from pydantic>=2.11.2->unstructured-client->unstructured[all]) (2.41.5)\n",
      "Requirement already satisfied: typing-inspection>=0.4.2 in /opt/conda/lib/python3.11/site-packages (from pydantic>=2.11.2->unstructured-client->unstructured[all]) (0.4.2)\n",
      "Requirement already satisfied: mypy-extensions>=0.3.0 in /opt/conda/lib/python3.11/site-packages (from typing-inspect<1,>=0.4.0->dataclasses-json->unstructured[all]) (1.1.0)\n",
      "Requirement already satisfied: pycparser in /opt/conda/lib/python3.11/site-packages (from cffi>=1.12->cryptography>=3.1->unstructured-client->unstructured[all]) (2.21)\n",
      "Requirement already satisfied: sniffio>=1.1 in /opt/conda/lib/python3.11/site-packages (from anyio->httpx>=0.27.0->unstructured-client->unstructured[all]) (1.3.0)\n",
      "Downloading backoff-2.2.1-py3-none-any.whl (15 kB)\n",
      "Downloading emoji-2.15.0-py3-none-any.whl (608 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m608.4/608.4 kB\u001b[0m \u001b[31m89.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading filetype-1.2.0-py2.py3-none-any.whl (19 kB)\n",
      "Downloading html5lib-1.1-py2.py3-none-any.whl (112 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m112.2/112.2 kB\u001b[0m \u001b[31m28.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading lxml-6.0.2-cp311-cp311-manylinux_2_26_x86_64.manylinux_2_28_x86_64.whl (5.2 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.2/5.2 MB\u001b[0m \u001b[31m97.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mta \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading nltk-3.9.2-py3-none-any.whl (1.5 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.5/1.5 MB\u001b[0m \u001b[31m141.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading python_iso639-2025.11.16-py3-none-any.whl (167 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m167.8/167.8 kB\u001b[0m \u001b[31m37.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading python_magic-0.4.27-py2.py3-none-any.whl (13 kB)\n",
      "Downloading python_oxmsg-0.0.2-py3-none-any.whl (31 kB)\n",
      "Downloading unstructured-0.18.27-py3-none-any.whl (1.8 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.8/1.8 MB\u001b[0m \u001b[31m117.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading unstructured_client-0.42.8-py3-none-any.whl (219 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m220.0/220.0 kB\u001b[0m \u001b[31m51.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading wrapt-2.0.1-cp311-cp311-manylinux1_x86_64.manylinux_2_28_x86_64.manylinux_2_5_x86_64.whl (114 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m114.1/114.1 kB\u001b[0m \u001b[31m27.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading aiofiles-25.1.0-py3-none-any.whl (14 kB)\n",
      "Downloading pypdf-6.6.0-py3-none-any.whl (328 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m329.0/329.0 kB\u001b[0m \u001b[31m53.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading olefile-0.47-py2.py3-none-any.whl (114 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m114.6/114.6 kB\u001b[0m \u001b[31m27.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hBuilding wheels for collected packages: langdetect\n",
      "  Building wheel for langdetect (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993225 sha256=6a5e30c60da39b03a1814690e0f1bcf467429941f10c6f408cb3d9dfafbfb1c6\n",
      "  Stored in directory: /home/jovyan/.cache/pip/wheels/0a/f2/b2/e5ca405801e05eb7c8ed5b3b4bcf1fcabcd6272c167640072e\n",
      "Successfully built langdetect\n",
      "Installing collected packages: filetype, wrapt, python-magic, python-iso639, pypdf, olefile, nltk, lxml, langdetect, html5lib, emoji, backoff, aiofiles, python-oxmsg, unstructured-client, unstructured\n",
      "Successfully installed aiofiles-25.1.0 backoff-2.2.1 emoji-2.15.0 filetype-1.2.0 html5lib-1.1 langdetect-1.0.9 lxml-6.0.2 nltk-3.9.2 olefile-0.47 pypdf-6.6.0 python-iso639-2025.11.16 python-magic-0.4.27 python-oxmsg-0.0.2 unstructured-0.18.27 unstructured-client-0.42.8 wrapt-2.0.1\n",
      "Collecting pi-heif\n",
      "  Downloading pi_heif-1.1.1-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (6.5 kB)\n",
      "Collecting pillow>=11.1.0 (from pi-heif)\n",
      "  Downloading pillow-12.1.0-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (8.8 kB)\n",
      "Downloading pi_heif-1.1.1-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (1.4 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.4/1.4 MB\u001b[0m \u001b[31m11.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hDownloading pillow-12.1.0-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (7.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.0/7.0 MB\u001b[0m \u001b[31m116.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: pillow, pi-heif\n",
      "  Attempting uninstall: pillow\n",
      "    Found existing installation: pillow 10.2.0\n",
      "    Uninstalling pillow-10.2.0:\n",
      "      Successfully uninstalled pillow-10.2.0\n",
      "Successfully installed pi-heif-1.1.1 pillow-12.1.0\n"
     ]
    }
   ],
   "source": [
    "!pip uninstall -y numpy ml_dtypes onnx onnxruntime\n",
    "!pip install numpy==1.26.4\n",
    "!pip install ml_dtypes==0.2.0\n",
    "!pip install onnx==1.14.1\n",
    "!pip install onnxruntime==1.16.3\n",
    "!pip uninstall -y scipy\n",
    "!pip install scipy==1.10.1\n",
    "!pip install 'pdf2image'\n",
    "!pip install unstructured_inference\n",
    "!pip install langchain[community]\n",
    "!pip install \"unstructured[all]\"\n",
    "!pip install pi-heif\n",
    "!pip install unstructured-pytesseract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a603021d-08e8-45c6-accf-be67686d0bce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The `max_size` parameter is deprecated and will be removed in v4.26. Please specify in `size['longest_edge'] instead`.\n"
     ]
    }
   ],
   "source": [
    "## Loading a document\n",
    "from langchain_community.document_loaders import UnstructuredPDFLoader\n",
    "import os\n",
    "os.environ[\"TESSDATA_PREFIX\"] = \"/opt/conda/share/tessdata\"\n",
    "file_name = \"/home/jovyan/work/BioRAG/Data/pdfs/cells-11-02650-v2.pdf\"\n",
    "loader = UnstructuredPDFLoader(file_name,\n",
    "                               languages = ['eng'],\n",
    "                               mode = \"elements\",\n",
    "                               strategy = \"hi_res\",\n",
    "                               infer_table_structure= True,\n",
    "                              )\n",
    "docs = loader.load()"
   ]
  },
  {
   "cell_type": "raw",
   "id": "59979896-3c9a-497d-8e65-cd1c7f506055",
   "metadata": {},
   "source": [
    "import pickle\n",
    "import os\n",
    "\n",
    "with open('raw_docs.pkl', 'wb') as f:\n",
    "    pickle.dump(docs, f)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "4f037b11-de41-472b-b97c-a862ccd1617e",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "import pickle\n",
    "with open('/home/jovyan/work/CellSense/X20251016-CellSense-VB/pkl_files/raw_docs.pkl', 'rb') as f:\n",
    "     docs = pickle.load(f)\n",
    "\n",
    "# Observe the document objbects\n",
    "for d in docs:\n",
    "    print(d.page_content)\n",
    "    print(\"\\n\"+\"---\"*20+\"\\n\" + str(d.metadata)+\"\\n\" +\"---\"*20 +\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0d121013-b804-4a13-ab30-760284e6b51a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NarrativeText', 'Header', 'Title', 'ListItem', 'Image', 'Table', 'FigureCaption', 'UncategorizedText', 'Formula'}\n"
     ]
    }
   ],
   "source": [
    "cats = set()\n",
    "for do in docs:\n",
    "    cats.add(do.metadata.get('category'))\n",
    "print(cats)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9549474",
   "metadata": {},
   "source": [
    "### Key Notes:\n",
    "So, keeping only these four categories: Title, NarrativeText, Table, FigureCaption"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e43f9487-2b69-45c2-a135-6423c99f6f4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of filtered docs: 140\n"
     ]
    }
   ],
   "source": [
    "Docs = docs.copy()\n",
    "Docs = [doc for doc in Docs if doc.metadata.get('category') not in ['Image', 'Header', 'ListItem', 'UncategorizedText']]\n",
    "print(f\"Number of filtered docs: {len(Docs)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f3e0d5f-5afb-456b-97d4-e71461b9ba45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NarrativeText', 'Title', 'FigureCaption', 'Table', 'Formula'}\n"
     ]
    }
   ],
   "source": [
    "cats = set()\n",
    "for do in Docs:\n",
    "    cats.add(do.metadata.get('category'))\n",
    "print(cats)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "527c0440-ccc7-42ad-8785-0129ddf716dc",
   "metadata": {},
   "source": [
    "## Creating functions to clean texts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d5837da8-98d5-43df-89b2-15304aec08cd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Article\n",
      "\n",
      "Academic Editor: Ohad Medalia\n",
      "\n",
      "\n",
      "Received: 28 July 2022 Accepted: 23 August 2022 Published: 25 August 2022\n",
      "Publisher’s Note: MDPI stays neutral with regard to jurisdictional claims in published maps and institutional affil- iations.\n",
      "\n",
      "\n",
      "Cells 2022, 11, 2650. https: / /doi.org /10.3390/cells11172650\n",
      "https: //www.mdpi.com/journal /cells\n",
      "\n",
      "\n",
      "2.1. Electrical Stimulation (ES) Chamber C-Pace EM 100\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "2.2. Calculation and Simulation of ES\n",
      "\n",
      "\n",
      "2.3. Current Measurement\n",
      "\n",
      "2.4. Characterization of AC-Stimulated Liquid\n",
      "\n",
      "\n",
      "For all Controls, the electrodes were placed in the medium, and the power switch was not returned on (no ES).\n",
      "2.5. Cell Culture\n",
      "\n",
      "2.6. AC-Activated Medium and Online Monitoring of Cell Adhesion\n",
      "\n",
      "\n",
      "2.7. AC stimulation of Cells and Initial Adhesion\n",
      "\n",
      "2.8. Cell Morphology and Spreading\n",
      "\n",
      "\n",
      "\n",
      "2.9. Calcium Mobilization and AC stimulation\n",
      "\n",
      "\n",
      "2.10. Cellular Reactive Oxygen Species (ROS)\n",
      "\n",
      "\n",
      "\n",
      "2.11. Statistics\n",
      "\n",
      "\n",
      "3.1. Characterization of AC Electrical Stimulation (ES) Parameter\n",
      "\n",
      "\n",
      "8 of 22\n",
      "Table 1. Individual ES values after measurement and simulation. To each setup, appropriate colors were dedicated (Code).\n",
      "* Own calculation of electric field from simulation. ** Own measurement of electric current in DMEM with 10% FCS using Tektronix TDS2014B.\n",
      "3.2. Acellular Medium Characteristics—AC-Stimulated Liquid\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "3.3, AC-Stimulated Liquids and Long-Time Cell Adhesion\n",
      "\n",
      "10 of 22\n",
      "\n",
      "3.4. AC-Stimulated Cells\n",
      "3.4.1. Initial Cell Adhesion\n",
      "\n",
      "3.4.2. Intracellular Calcium Ion (Ca?*) Level during AC stimulation\n",
      "\n",
      "3.4.3. Calcium Ion Mobilization in AC-Pre-Stimulated Cells\n",
      "\n",
      "\n",
      "\n",
      "13 of 22\n",
      "3.4.5. Reactive Oxygen Species (ROS) Production\n",
      "\n",
      "14 of 22\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "16 of 22\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "18 of 22\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Electrical stimulation has the potential for novel treatments in medical applications.\n",
      "\n",
      "\n",
      "Institutional Review Board Statement: Not applicable.\n",
      "Informed Consent Statement: Not applicable.\n",
      "\n",
      "Conflicts of Interest: The authors declare no conflict of interest.\n",
      "19 of 22\n",
      "Cells 2022, 11, 2650\n",
      "20 of 22\n",
      "Cells 2022, 11, 2650\n"
     ]
    }
   ],
   "source": [
    "# serach for unnecessary narrative texts like 18 of 22 etc\n",
    "s= set()\n",
    "for do in Docs:\n",
    "    if do.metadata.get('category') == 'NarrativeText':\n",
    "        s.add(len(do.page_content))\n",
    "        print(do.page_content if len(do.page_content)<150 else \"\", end=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0b494767-010f-44e8-9f2a-e36646a2ef75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove boilerplate narrative texts\n",
    "import re\n",
    "from langchain_core.documents import Document\n",
    "BOILERPLATE_PATTERN = r\"\"\"\n",
    "\\b\\d+\\s*of\\s*\\d+\\b | # Matches '18 of 22'\n",
    "\\b[A-Za-z]+\\s+\\d{4},\\s*\\d+,\\s*\\d+\\b # Matches 'Cells 2022, 11, 2650'\n",
    "\"\"\"\n",
    "\n",
    "boilerplate_regex = re.compile(BOILERPLATE_PATTERN, re.VERBOSE)\n",
    "def remove_boilerplate(Docs: Document) -> list[Document]:\n",
    "    for doc in Docs:\n",
    "        if (doc.metadata.get('category') == 'NarrativeText'):\n",
    "            if boilerplate_regex.fullmatch(doc.page_content.strip()):\n",
    "                Docs.remove(doc)\n",
    "    return Docs\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0457a353-e0e3-44c2-a179-4306f873b938",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove unnnecessary narattivetext like citation publisher info\n",
    "\n",
    "def clean_narrative_text(Docs: Document) -> list[Document]:\n",
    "    text_start_with = ['Cells 2022, 11, 2650', 'Article', 'Academic Editor','Received:', 'Citation',\n",
    "                       'Accepted:' , 'Published:',\"Publisher’s Note:\", \"G check for updates\", \"Susanne Staehlke 1\",\n",
    "                       \"Institutional Review Board Statement: Not applic\",'Informed Consent Statement: Not applicable',\n",
    "                       'Conﬂicts of Interest:', 'https:', 'Copyright: © ', 'Acknowledgments:', 'Funding:','Data Availability Statement','References']\n",
    "    docs = [doc for doc in Docs if not any(doc.page_content.strip().startswith(text) for text in text_start_with)]\n",
    "    return docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "49c375ac-fc23-4c62-90cf-d286b9d9ca56",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling\n",
      "\n",
      "\n",
      "1. Introduction\n",
      "\n",
      "\n",
      "\n",
      "2. Materials and Methods\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "3. Results\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "3.4.4. Cell Morphology\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "4. Discussion\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "5. Conclusions\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Docs = remove_boilerplate(Docs)\n",
    "Docs = clean_narrative_text(Docs)\n",
    "len(Docs)\n",
    "metadata_keys = set()\n",
    "for doc in Docs:\n",
    "    print(doc.page_content if doc.metadata.get('category') == 'Title' else \"\",)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8a4440cf-4241-4581-953c-3e25c5fc6c1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_huggingface.llms import HuggingFacePipeline\n",
    "from transformers import AutoTokenizer, pipeline, AutoModelForCausalLM\n",
    "from dotenv import load_dotenv\n",
    "import torch\n",
    "\n",
    "# defining environment variable\n",
    "load_dotenv()\n",
    "\n",
    "# Defining Model\n",
    "model_name = \"meta-llama/Llama-3.1-8B-Instruct\"\n",
    "\n",
    "def open_souce_model(model_name, max_new_tokens: int):\n",
    "    tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "    \n",
    "    pipe = pipeline(\n",
    "        \"text-generation\",\n",
    "        model=model_name,\n",
    "        tokenizer=tokenizer,\n",
    "        device_map=\"auto\",\n",
    "        torch_dtype=torch.bfloat16, \n",
    "        max_new_tokens=max_new_tokens,   \n",
    "        do_sample=False,          \n",
    "        temperature=0.0,           \n",
    "        repetition_penalty=1.2,\n",
    "        return_full_text=False     \n",
    "    )\n",
    "    \n",
    "    model_pipe = HuggingFacePipeline(pipeline=pipe)\n",
    "    \n",
    "    # Binding parameters to the runnable\n",
    "    model = model_pipe.bind(skip_prompt=True) \n",
    "    \n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "88f191f2-5d78-4676-b6b7-2f6440b6fafc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "from langchain_core.output_parsers import PydanticOutputParser, StrOutputParser\n",
    "\n",
    "def generate_table_summary(txt_as_html: str, caption: str) -> str:\n",
    "\n",
    "    # defining a role for model\n",
    "    Role = \"\"\"<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n",
    "\n",
    "You are an expert biomedical research assistant specializing in analyzing scientific tables. \n",
    "Your ONLY task is to generate a written summary of the provided table in exactly one paragraph of approximately 100 words.\n",
    "\n",
    "**Strict Directives:**\n",
    "1. **Focus on Comparisons:** Describe the relationship between the experimental groups (e.g., 1V vs 5V) and the control. \n",
    "2. **Handle Footnotes Safely:** Do not use math formulas (like E=U/d) to calculate new values. Mention that theoretical and measured values were compared if applicable, but focus on the resulting data.\n",
    "3. **Data Precision:** For mean ± s.e.m. data (e.g., 319 ± 5.1), report only the mean (319) to maintain narrative flow.\n",
    "4. **Professional Tone:** Use high-level scientific terminology (e.g., \"osteoblastic proliferation,\" \"amplitude-dependent response,\" \"temporal increase\").\n",
    "5. **No Filler:** Start immediately with the summary. Do not use headers, bolding, or introductory phrases.\n",
    "\n",
    "**Constraint:** If the summary is under 80 words, expand on the temporal trends (how data changed over time) or the impact of voltage/frequency variations.<|eot_id|><|start_header_id|>user<|end_header_id|>\n",
    "\n",
    "Table Caption: [INSERT_CAPTION_HERE]\n",
    "HTML Table: [INSERT_HTML_HERE]<|eot_id|><|start_header_id|>assistant<|end_header_id|>\"\"\"\n",
    "\n",
    "    # setting up the parser. The summary will be in string only so, using string output parser\n",
    "    #parser = PydanticOutputParser(pydantic_object=tablesummary)\n",
    "\n",
    "    parser = StrOutputParser()\n",
    "    # defining prompt\n",
    "    table_template = ChatPromptTemplate([\n",
    "        (\"system\", Role),\n",
    "        (\"human\", \"Please generate summary, Table content: \\n{table_content}\\nCaption: {table_caption}\")\n",
    "    ],\n",
    "    )\n",
    "\n",
    "    \n",
    "    # defining a chain\n",
    "    chain = table_template | model | parser\n",
    "    response = chain.invoke({\"table_content\": txt_as_html, \"table_caption\": caption})\n",
    "    return response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f81c71d5-1084-4c72-810d-a8ec0266b277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dd16be8695264be1bd94edf5281a882a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Device set to use cuda:0\n",
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n",
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The asterisk indicates calculated parameters based on given settings. \n",
      "\n",
      "Summary:\n",
      "The results show significant differences in electric current measurements across various setups when comparing the control group to those exposed to electrical stimulation at different frequencies and voltages. Notably, increasing the frequency from 7.9 Hz to 20 Hz resulted in higher measured currents for both 1V and 5V conditions, indicating an amplitude-dependent response. In contrast, varying the pulse duration had minimal effect on the measured current levels. These findings suggest that changes in either frequency or voltage significantly influence the magnitude of the induced electric field, leading to increased osteoblastic activity as observed through the measured electric current.\n",
      "The results show that all stimulation conditions significantly increased cell area when compared to the unstimulated controls at each respective time point. The amplitude-dependent response was observed as higher voltages resulted in greater increases in cell area; however, this effect diminished by 24 hours. A notable exception occurred for the 5V_20Hz group which showed a significant decrease from its peak value at 4 hours.\n"
     ]
    }
   ],
   "source": [
    "model = open_souce_model(model_name, max_new_tokens=300)\n",
    "\n",
    "current_caption = None  # track latest caption\n",
    "\n",
    "for doc in Docs:\n",
    "    content = doc.page_content\n",
    "    doc_category = doc.metadata.get(\"category\", \"\")\n",
    "    metadata_copy = doc.metadata.copy()\n",
    "\n",
    "    # Step 1: Capture table caption\n",
    "    if doc_category in {\"NarrativeText\", \"FigureCaption\"} and content.strip().startswith(\"Table\"):\n",
    "        current_caption = content.strip()\n",
    "        continue\n",
    "\n",
    "    # Step 2: Process table\n",
    "    if doc_category == \"Table\" and metadata_copy.get(\"text_as_html\"):\n",
    "        table_html = metadata_copy[\"text_as_html\"]\n",
    "\n",
    "        summary = generate_table_summary(\n",
    "            txt_as_html=table_html,\n",
    "            caption=current_caption,\n",
    "            \n",
    "        )\n",
    "\n",
    "        # Clean model artifacts\n",
    "        summary = summary.replace(\"Assistant:\", \"\").strip()\n",
    "\n",
    "        print(summary)\n",
    "\n",
    "        # Reset caption after use\n",
    "        current_caption = None\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dde18354",
   "metadata": {},
   "source": [
    "### Key Notes: \n",
    "The above two functions will clean the complete NarativeText and Title category."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f090377a",
   "metadata": {},
   "source": [
    "## Strategy 1: Parent-child chunking for Parent Document Retriever"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c615b8d4-35d7-4d48-afa1-2237688127c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'text_as_html', 'parent_id', 'coordinates', 'detection_class_prob', 'page_number', 'languages', 'last_modified', 'file_directory', 'element_id', 'source', 'filename', 'filetype', 'category'}\n"
     ]
    }
   ],
   "source": [
    "# First check the what metadata keys are present\n",
    "metadata_keys = set()\n",
    "for doc in Docs:\n",
    "    for key in doc.metadata.keys():\n",
    "        metadata_keys.add(key)\n",
    "print(metadata_keys)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "01ac4e47-bb9d-4038-bb87-5d573b8c57f0",
   "metadata": {},
   "source": [
    "# Filtering the metadtata keys\n",
    "def filter_metadata_keys(Docs: list[Document], not_allowed_keys: list[str]) -> list[Document]:\n",
    "    filtered_docs = []\n",
    "    for doc in Docs:\n",
    "        filtered_metadata = {key: value for key, value in doc.metadata.items() if key not in not_allowed_keys}\n",
    "        filtered_doc = Document(page_content=doc.page_content, metadata=filtered_metadata)\n",
    "        filtered_docs.append(filtered_doc)\n",
    "    return filtered_docs\n",
    "\n",
    "filter_docs = filter_metadata_keys(Docs, not_allowed_keys=['coordinates', 'languages', 'element_id', 'last_modified'])\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "476b8b31-e23a-44ed-9f77-53a88b1d9a4d",
   "metadata": {},
   "source": [
    "# First check the what metadata keys are present\n",
    "metadata_keys = set()\n",
    "for doc in filter_docs:\n",
    "    for key in doc.metadata.keys():\n",
    "        metadata_keys.add(key)\n",
    "print(metadata_keys)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b97dc33",
   "metadata": {},
   "source": [
    "### Generating parent chunks\n",
    "> Generating parent chunks by diving document into sub-section wise chunks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5e1aeae0-eea3-4bf4-b0d7-d6e86b2e92e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creating sub-section wise chunks from cleaned documents\n",
    "from uuid import uuid4\n",
    "import re\n",
    "\n",
    "def Create_parent_chunks(Docs: list[Document]) -> list[Document]:\n",
    "    new_docs = []\n",
    "    new_content = \"\"\n",
    "    section_header = \"\"\n",
    "    subsection_header = \"\"\n",
    "    Paper_title = \"Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling\"\n",
    "    metadata = {}\n",
    "    current_chunk_page = None\n",
    "    section_pattern = r\"^(\\d+\\.)+\\s*(.+)$\"\n",
    "    subsection_pattern = r\"`\\s*\\d+(?:\\.\\d+)*\\s+[^\\n]+`\"\n",
    "    for doc in Docs:\n",
    "        metadata_copy = doc.metadata.copy()\n",
    "        \n",
    "        content = doc.page_content\n",
    "        current_page = metadata_copy.get('page_number', None)\n",
    "        doc_category = doc.metadata.get('category', '')\n",
    "        is_section_header = (doc_category == 'Title')\n",
    "\n",
    "        if is_section_header:\n",
    "            if new_content:\n",
    "                chunk_metadata = doc.metadata.copy()\n",
    "                chunk_metadata['section_header'] = section_header\n",
    "                \n",
    "                if current_chunk_page is not None:\n",
    "                    chunk_metadata['start_page'] = current_chunk_page\n",
    "                chunk_metadata['end_page'] = current_page\n",
    "                chunk_metadata['paper_title'] = Paper_title\n",
    "                document_id = str(uuid4())\n",
    "                chunk_metadata['document_id'] = document_id\n",
    "                new_docs.append(Document(page_content=new_content.strip(), metadata=chunk_metadata))\n",
    "\n",
    "            if content.strip() != Paper_title:\n",
    "                section_header = content.strip()\n",
    "                new_content = \"\"\n",
    "                metadata = metadata_copy\n",
    "                current_chunk_page = current_page\n",
    "            continue\n",
    "\n",
    "        # Now we will add the category wise content, unless it is a section header\n",
    "        if doc_category in ['NarrativeText', 'FigureCaption']:\n",
    "            current_page = doc.metadata.get('page_number', None)\n",
    "            chunk_metadata = doc.metadata.copy()\n",
    "            chunk_metadata['section_header'] = section_header\n",
    "            chunk_metadata['subsection_header'] = subsection_header\n",
    "            if re.match(section_pattern, content.strip()):\n",
    "                subsection_header = content.strip()\n",
    "            new_content += \"\\n\" + content.strip()+ \"\\n\"\n",
    "\n",
    "        if (doc_category == \"NarrativeText\" and content.startswith(\"Table\")) or (doc_category == \"FigureCaption\" and content.startswith(\"Table\")):\n",
    "                caption = content.strip()\n",
    "\n",
    "        if doc_category == \"Table\" and metadata_copy.get('text_as_html'):\n",
    "                txt_html = metadata_copy.get('text_as_html')\n",
    "                summary = generate_table_summary(txt_html, caption)\n",
    "                new_content += \"\\n Table Summary: \\n\" + summary.replace(\"Assistant: \", \"\") + \"\\n\"\n",
    "                \n",
    "        metadata.update(chunk_metadata)\n",
    "\n",
    "    # After processing all documents, check if there's remaining content to be added\n",
    "    if new_content:\n",
    "        chunk_metadata = metadata.copy()\n",
    "        chunk_metadata['section_header'] = section_header\n",
    "        if current_chunk_page is not None:\n",
    "            chunk_metadata['start_page'] = current_chunk_page\n",
    "        chunk_metadata['end_page'] = current_page\n",
    "        chunk_metadata['paper_title'] = Paper_title\n",
    "        document_id = str(uuid4())\n",
    "        chunk_metadata['document_id'] = document_id\n",
    "        new_docs.append(Document(page_content=new_content.strip(), metadata=chunk_metadata))\n",
    "    return new_docs\n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c310643b-acc7-4277-a759-1e9b0b681f88",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "model_name = \"Qwen/Qwen2.5-7B-Instruct\"\n",
    "model = open_souce_model(model_name,max_new_tokens=500)\n",
    "new_docs = Create_parent_chunks(filter_docs)\n",
    "for doc in new_docs:\n",
    "    print(doc.page_content)\n",
    "    print(\"\\n\"+\"---\"*20+\"\\n\" + str(doc.metadata)+\"\\n\" +\"---\"*20 +\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f83a8a46-04c8-45e9-8ea1-5c5a3b533c8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 1568\n",
      "1 3625\n",
      "2 17195\n",
      "3 12349\n",
      "4 5193\n",
      "5 15386\n",
      "6 1989\n"
     ]
    }
   ],
   "source": [
    "for i, doc in enumerate(new_docs):\n",
    "    print(i, len(doc.page_content))"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a6046602-d731-4443-92ea-e1a7f2019db1",
   "metadata": {},
   "source": [
    "# Create Parent chunk doc pickle\n",
    "import pickle\n",
    "\n",
    "with open(\"pkl_files/parent_docs.pkl\", \"wb\") as f:\n",
    "    pickle.dump(new_docs, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2b9ca4ae-417b-4043-a369-2ac1e7f59552",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Abstract: An extensive research ﬁeld in regenerative medicine is electrical stimulation (ES) and its impact on tissue and cells. The mechanism of action of ES, particularly the role of electrical parameters like intensity, frequency, and duration of the electric ﬁeld, is not yet fully understood. Human MG-63 osteoblasts were electrically stimulated for 10 min with a commercially available multi-channel system (IonOptix). We generated alternating current (AC) electrical ﬁelds with a voltage of 1 or 5 V and frequencies of 7.9 or 20 Hz, respectively. To exclude liquid-mediated effects, we characterized the AC-stimulated culture medium. AC stimulation did not change the medium’s pH, temperature, and oxygen content. The H2O2 level was comparable with the unstimulated samples except at 5 V_7.9 Hz, where a signiﬁcant increase in H2O2 was found within the ﬁrst 30 min. Pulsed electrical stimulation was beneﬁcial for the process of attachment and initial adhesion of suspended osteoblasts. At the same time, the intracellular Ca2+ level was enhanced and highest for 20 Hz stimulated cells with 1 and 5 V, respectively. In addition, increased Ca2+ mobilization after an additional trigger (ATP) was detected at these parameters. New knowledge was provided on why electrical stimulation contributes to cell activation in bone tissue regeneration.\n",
      "\n",
      "Keywords: electrical stimulation; electric ﬁeld strength; ﬁeld simulation; AC-stimulated liquid; osteoblasts adhesion; spreading; calcium ions; reactive oxygen species; confocal microscopy; scanning electron microscopy\n",
      "\n",
      "------------------------------------------------------------\n",
      "{'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.8415048718452454, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 1, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'category': 'Title', 'element_id': '296a68ec9895666a85773a65935440ac', 'section_header': '', 'end_page': 1, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': '0dd19c2d-1b2c-4cca-822a-df636d4b0b78'}\n",
      "------------------------------------------------------------\n",
      "\n",
      "A widespread research ﬁeld in regenerative medicine is electrical stimulation (ES), and its impact on tissue and cells, such as bone [1]. In 1957 the piezoelectric properties of bone were described [2]. As the bone healing process takes place under the mechanical strain of the bone [3], an electric ﬁeld is generated in the bone in vitro and in vivo [4]. Externally applied electric ﬁelds were shown to contribute to bone deposition and osteoblast differentiation and proliferation [5]. The living cell has a membrane potential that indicates the electrical potential difference between the intracellular and extracellular space, which is generated by the transport of ions through ion channels and ion transporters [6]. The membrane potential is always negative because of the higher negative ion concentration inside the cell [7,8]. Primary osteoblasts revealed a membrane potential of −60 mV [9], MG- 63 osteoblasts demonstrated a negatively charged cell surface of −15.6 mV [10]. Due to the appearance of ions and polar or charged molecules, cells can build up electrical ﬁelds or can respond to them [11,12]. If injuries appear on the bone, a wound potential, which should lead precursor cells to differentiation, will be generated [13]. A possible approach for the\n",
      "\n",
      "therapy is through electrical stimulation to accelerate bone healing. In some publications, it is described that ES can manipulate different cell processes, for example, cell adhesion, proliferation, apoptosis, cell migration, and differentiation [8,14]. Widespread methods of the ES for bone tissue are invasive methods like direct contact stimulation [14], in vitro, the electrically conductive electrodes are placed directly in the culture medium. The applied current streams from the anode to the cathode [15]. The modiﬁcation of the cell membrane potential through ES can activate different intracellular pathways. External electrical ﬁelds can generate an intracellular calcium ion (Ca2+) ﬂow and consequently activate different pathways [16], such as wnt [17]. Ca2+ dynamics were identiﬁed as a sensitive parameter for cell physiological processes. Furthermore, the ES can stimulate the production of reactive oxygen species (ROS), which can control different cellular processes like proliferation and modulated signaling pathways [18,19]. When the equilibrium is destroyed, and the concentration of ROS is relatively high, cells suffer under oxidative stress. Moreover, Ca2+ signaling was shown to contribute to ROS generation through calmodulin [20], as well as ROS contributing to increased intracellular Ca2+ concentrations by altering calcium channels [21].\n",
      "\n",
      "It was described that the electric ﬁeld could modulate cell functions in order to accelerate wound healing and bone regeneration [5,12,14]. In contrast to previous studies, we are focusing on the inﬂuence of a short alternating current (AC)-stimulation time on settling cells via a commercial device (IonOptix) with the following parameters: rectangular waveform, voltages of ±1 or ±5 V, frequency of 7.9 Hz with pulse duration of 10 ms, or 20 Hz with pulse duration of 3.6 ms. In vitro, cell processes such as intracellular Ca2+ level and mobilization, adhesion, and morphology of human osteoblasts within 24 h were studied to understand the effect of ES on osteoblast behavior. In addition, we wanted to shed light on the question of whether ES affects the cells via liquids, i.e., AC-pre-conditioned medium. For these cell experiments, solely the cell culture medium was stimulated in the IonOptix chamber and analyzed for pH, temperature, oxygen, and hydrogen peroxide content.\n",
      "\n",
      "------------------------------------------------------------\n",
      "{'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.8418117165565491, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 2, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'parent_id': '5c72fa6a292e875d7cf7227febce387c', 'category': 'Title', 'element_id': '83ad84162f789ab47bfee56a766fdcac', 'section_header': '1. Introduction', 'start_page': 1, 'end_page': 2, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': '44aba20e-bd51-4964-950a-69aaf97680e9'}\n",
      "------------------------------------------------------------\n",
      "\n",
      "2.1. Electrical Stimulation (ES) Chamber C-Pace EM 100\n",
      "\n",
      "For electrical stimulation (ES), we used a multi-channel electrical stimulator, a voltage generator, and a 12-well C (culture)-Dish (IonOptix, Milton, MA, USA) (Figure 1a,c).\n",
      "\n",
      "IonOptix devices were used for the electrical and mechanical in vitro stimulation of cells, e.g., stem cells, cardiogenic and osteoblasts, or for tissues [12,14].\n",
      "\n",
      "The stimulation chamber C-Pace EM 100 is a multi-channel stimulator that enables the alternating current (AC)-stimulation of cells in an incubator. The electrodes are designed for a 12-well plate (Greiner bio-one, Frickenhausen, Germany), and reach down to the bottom of the wells. This way, all cells can be electrically stimulated. The electrodes consist of carbon in the graphite modiﬁcation and are shaped in a plate format (capacitor). The advantages of graphite are its excellent electrical conductivity and the high resistance against corrosion. The protein adsorption on the surface is problematic as it affects the electrical ﬁeld´s homogeneity. Therefore, a regular cleaning after use is necessary [22]. The electrodes can be sterilized by treatment with ethanol and ultraviolet light (UV). The space between both plate electrodes amounts to 1.1 cm, and the wetted area in a well of the 12-well plate totals 0.42 cm2 when 1 mL ﬂuid is in the well (Figure 1b). To always establish identical conditions of the experiments, coverslips (Menzel™ Microscope Coverslips, Ø 18 mm, Thermo Fisher Scientiﬁc, Waltham, MA, USA) were placed on each bottom of the 12-well plate.\n",
      "\n",
      "Figure 1. Scheme of the AC electrical stimulation system. (a) IonOptix stimulation chamber C-Pace EM 100—multi-channel stimulator, (b) graphite electrodes for a 12-well plate format reaching the bottom, and (c) simplified scheme of the 12 well C-Dish. For the current measurement, each column of the 12-well plate (background) is connected to the IonOptix controller via a shunt resistor. The corresponding voltage drop is amplified by an instrumentation amplifier which can be connected to any of the columns (not shown). In (d) illustration of rectangular electronic bipolar pulses of the AC stimulation with 20 Hz, and (e) 7.9 Hz. The pulse duration covers both positive and negative half pulses.\n",
      "\n",
      "This ES-chamber generates rectangular electronic biphasic pulses for electrical stim- ulation. The parameters can be varied: the applied voltage up to ±40 V, the frequencies from 0.01 to 99 Hz, and the pulse duration from 0.4 to 10 ms [12,14]. Every well is gated in a deﬁnite sequence.\n",
      "\n",
      "In general, we used 1 and 5 V for the experiments, and a frequency of 20 Hz (3.6 ms double duration time, ± V) (1 V_20 Hz, 5 V_20 Hz) (Figure 1d) or 7.9 Hz (10 ms double duration time, ± V) (1 V_7.9 Hz, 5 V_7.9 Hz) (Figure 1e). After seeding, the cells in the 12-well plates were directly stimulated for 10 min. For the controls we worked with the C-Dish electrodes but without electrical stimulation (system still switched off) (Control).\n",
      "\n",
      "2.2. Calculation and Simulation of ES\n",
      "\n",
      "The applied electric ﬁeld strength was evaluated by employing numerical simulations. Under the assumption that the dielectric parameters of the cell culture medium are not frequency-dependent, the electrostatic Laplace equation is solved [23]. This equation reads\n",
      "\n",
      "This yields the electric potential (φ). Subsequently, the electric ﬁeld can be computed by taking the gradient of the potential. As the solution depends linearly on the voltage difference between the two plates, we evaluated the electric ﬁeld at a voltage difference of 1 V. In the case of a different voltage difference, the solution can be linearly scaled. We used the ﬁnite element method for our computations, implemented in the open-source package EMStimTools (https://github.com/j-zimmermann/EMStimTools/releases/tag/ v0.1.4.dev0; accessed on 27 July 2022) [24]. The geometry of a single well was used and represented by a 2D model. The 2D representation corresponds to a slice through the geometry at a certain height and the top view, respectively.\n",
      "\n",
      "2.3. Current Measurement\n",
      "\n",
      "The electrodes of the 12-well plate are arranged in a 4 × 3 matrix such that individual wells can be addressed sequentially by applying the stimulation voltage to the respective row and column, see Figure 1c. Current measurement was accomplished via a shunt resistor (R = 1 Ω) inserted into each column, enabling the sequential measurement of all four wells in the same column. An instrumentation ampliﬁer with variable gain and high bandwidth (up to 500 kHz) was used to precisely convert the differential shunt voltage drop to a single-ended output voltage. The stated currents corresponded to the peak values of the time-dependent signal.\n",
      "\n",
      "2.4. Characterization of AC-Stimulated Liquid\n",
      "\n",
      "To eliminate the possibility that the electric ﬁeld inﬂuences the medium alone and thus inﬂuences the cell physiology, the medium parameters temperature, pH, oxygen, and hydrogen peroxide (H2O2) content were analyzed. For temperature, pH, and oxygen, fresh Dulbecco’s Modiﬁed Eagle’s Medium (DMEM; Thermo Fisher Scientiﬁc, Waltham, MA, USA) with 10% fetal calf serum (FCS; Biochrom FCS Superior, Merck KGaA, Darmstadt, Germany) and 1% gentamicin (Ratiopharm GmbH, Ulm, Germany) was exposed to ES for 10 min in a 12-well plate (1 mL per well). These parameters were measured in the middle of the liquid volume in the incubator at 37 ◦C, and 5% CO2 in triplicate: (i) temperature with the device Mini-K (Dostmann, with temperature gauger for Type K probes by testo), (ii) pH with a pH meter (Sentron SI series pH meter, SI400-010; Sentron Europe BV, Leek, The Netherlands), and (iii) oxygen with a needle-shaped optical oxygen microsensor (Oxygen micro-optode, type PSt1; Presens) connected to the Microx TX3 (Presens).\n",
      "\n",
      "The H2O2 concentration was determined using the Fluorimetric Hydrogen Peroxidase Assay kit (Sigma Aldrich, Merck KGaA, Darmstadt, Germany). For H2O2 measurement, DMEM (without pyruvate; Life technologies, Thermo Fisher Scientiﬁc) with 10% FCS was 10 min AC-stimulated. H2O2 content was analyzed at time points 0, 0.5, 4, and 24 h, and the probes were kept at 37 ◦C.\n",
      "\n",
      "For all Controls, the electrodes were placed in the medium, and the power switch was not returned on (no ES).\n",
      "\n",
      "2.5. Cell Culture\n",
      "\n",
      "The cell biological investigations were performed with human osteoblast-like cells MG-63 (passage number 5 to 30, ATCC® CRL-1427™, Manassas, VA, USA) [25]. The MG- 63s represent a well-established osteoblast model due to the similar characteristics in terms of cell spreading, adhesion, and signaling properties as primary human osteoblasts [25–27]. The MG-63s were cultivated in complete DMEM (with pyruvate, Thermo Fisher; 10% FCS; see Section 2.4.) at 37 ◦C in an incubator with 5% CO2. To obtain suspended cells for the experiments, cells were trypsinated (0.05% trypsin-EDTA, ethylenediaminetetraacetic acid; PAA Laboratories, Pasching, Australia), collected in a complete medium and centrifuged (13,000× g, 4 ◦C, 5 min; Eppendorf 5702R; Hamburg, Germany).\n",
      "\n",
      "2.6. AC-Activated Medium and Online Monitoring of Cell Adhesion\n",
      "\n",
      "To solve the question of whether the electrostimulation effects on cells are mediated via the liquid, the medium alone was AC-activated. For long-time cell adhesion and spreading, the xCELLigence RTCA (Real-time Cell Analysis) S16 instrument (ACEA Biosciences, San Diego, CA, USA) was used [28]. The impedance-based assay determines the ion environment at the microelectrode/solution interface. The system displays the impedance in an arbitrary unit—Cell Index, which is determined by the cell-microelectrode impedance minus the baseline-impedance (medium without cells).\n",
      "\n",
      "The E-plate VIEW 16 (ACEA Biosciences, similar to 96-well format) was pre-incubated in an incubator at 37 ◦C for 2 h before the experiments started [29]. The complete DMEM was ﬁrst exposed to AC electric ﬁelds for 10 min. Then, MG-63s (10,000 cells/well) were seeded into the E-plate in 200 µL of the appropriate AC-stimulated medium. The impedance change caused by cell adhesion, and spreading was measured continuously at 37 ◦C and 5% CO2 up to 24 h every 8 min. Data were analyzed using RTCA Data Analysis Software 1.0. The software calculated the slope of the cell adhesion over time, the Cell Index value after 24 h (Max Cell Index), and the adhesion rate (Cell Index at the time point when 50% of the cells adhere). MG-63s long-time adhesion was presented for ﬁve independent experiments.\n",
      "\n",
      "2.7. AC stimulation of Cells and Initial Adhesion\n",
      "\n",
      "The question arose whether cells in suspension are already activated by AC and thus their adhesion behavior is altered. To hold the cells (1 × 106) in suspension, MG-63s were shaken (ThermoMixer C; Eppendorf AG, Hamburg, Germany) in a complete medium at 300 rpm at 37 ◦C. For the initial adhesion approach, 40 µL of suspended MG-63s was seeded into one well (12-well plate) containing a coverslip, electrodes were placed, and ES was started. After 10 min, 300 µL of the non-adherent MG-63s in the supernatant were counted by the ﬂow cytometer FACSCalibur (BD Biosciences, San Jose, CA, USA) using CellQuest Pro 4.0.1 (BD) software for data acquisition. The number of adherent osteoblasts was calculated in percent relative to the cell number at 0 min (suspended cells without adhesion on glasses). A new seeding of MG-63s with different ES parameters was performed every 10 min within a series of approaches with the same cell vial. Initial adhesion assays were performed seven times in independent experiments.\n",
      "\n",
      "2.8. Cell Morphology and Spreading\n",
      "\n",
      "At 0.5, 4, and 24 h after electrical stimulation (ES), the medium was removed, and the cells on coverslips in the wells were washed with PBS and ﬁxed with 2.5% glutardialdehyde (Merck KGaA, Darmstadt, Germany) solution in PBS. The samples were stored until further processing at 4 ◦C. The ﬁxed samples were washed with 0.1 M sodium–phosphate buffer (Merck KGaA), then they were dehydrated in an increasing ethanol solution and dried with the critical point dryer Emitech K850 (Quorum Technologies Ltd., East Sussex, UK) [29]. The dried samples were afﬁxed on sample holders with the aid of conducting carbon tape and\n",
      "\n",
      "sputtered with a thin gold layer (10–15 nm; EM SCD 004, BAL-TEC, Balzers, Liechtenstein). The samples were analyzed by scanning electron microscopy (FE-SEM, MERLIN® VP Compact, Carl Zeiss, Oberkochen, Germany) with the following instrumentation: HE-SE2 detector, acceleration voltage 5 kV, and working distance 5–7 mm.\n",
      "\n",
      "TheFE-SEMimageswereprocessedwiththeprogramImageJ(NationalInstitutesofHealth, Bethesda, USA) to calculate the cell area. Therefore, 40 cells per treatment were measured.\n",
      "\n",
      "2.9. Calcium Mobilization and AC stimulation\n",
      "\n",
      "The membrane-permeable calcium indicator ﬂuo-3/AM (Life Technologies Corpora- tion, Eugene, OR, USA) was used [30], to detect if AC mobilizes calcium ions (Ca2+) in the cytosol. Fluo-3/AM consists of molecules with two free binding sites that display an increase in ﬂuorescence in the presence of free Ca2+.\n",
      "\n",
      "MG-63s in a near conﬂuent state (70–80%) were washed with PBS, trypsinized, and centrifuged (Centrifuge 5810 R, Eppendorf AG). Next, 2 × 106 cells/2 mL were then resuspended in complete medium DMEM with shaking at 37 ◦C (ThermoMixer C). For every experiment, 2.5 × 105 cells from this pool were rewashed and centrifuged, and the cell sediment was stained with ﬂuo-3/AM (5 µM) in slightly hypotonic 4-(2-hydroxyethyl)- 1-piperazineethanesulfonic acid (HEPES) buffer [31], with shaking at 37 ◦C. After 40 min, cells were centrifuged and resuspended in 250 µL of isotonic HEPES (formulation described in [32]). Next, 750 µL of isotonic HEPES was added to a 12-well plate (glass bottom, Cellvis, Mountain View, CA, USA), the stained suspended cells were added, and stimulation was started immediately. After 10 min, AC stimulation was stopped, and electrodes were removed. Both, during stimulation and after the 10 min AC stimulation, time series visualized the calcium signal of MG-63s using an LSM780 inverted microscope (Carl Zeiss, Jena, Germany, argon ion laser at 488 nm) with a C Apochromat 40× water immersion objective (Carl Zeiss, 1.20 W Korr M27). The acquisition of the Ca2+ signal was made using Zen2011 (black edition, Carl Zeiss, Jenna, Germany). The ﬁrst time series during stimulation included 150 cycles every 2 s, whereby the Ca2+ signal was recorded upon stimulation after a 5 min settling phase of the cells. After the 10 min AC stimulation, the electrodes were removed, and the second time series was started with 240 cycles at 2 s intervals, whereby after the 90th cycle, 10 mM adenosine-50-triphosphate (ATP; SERVA Electrophoresis GmbH, Heidelberg, Germany; dissolved in PBS) was added. The mean ﬂuorescence intensity (MFI) analysis was evaluated with Zen2012 software (blue edition, version 2.0.0.0, Carl Zeiss). The mean ROI for 10 deﬁned areas of individual cells for each time point (240 cycles = 240 time points) was determined. In preliminary experiments, the cell areas went out of focus after approximately 60 s AC stimulation (not in Controls), so only the ﬁrst 30 cycles could be evaluated. At least 6 independent experiments were analyzed to see the inﬂuence of AC stimulation on calcium–ion dynamics (Figure 2). In all experiments, the Ca2+ signal was recorded under the same settings (gain, digital offset)\n",
      "\n",
      "2.10. Cellular Reactive Oxygen Species (ROS)\n",
      "\n",
      "Intracellular reactive oxygen species (ROS) were detected via dichloro-dihydro-fluorescein diacetate (DCFH-DA) assay (Abcam Inc, Cambridge, MA, USA) to verify whether the initial 10 min AC stimulation leads to oxidative stress. The non-ﬂuorescent substance DCFH- DA is incorporated by cells through the cell membrane. Inside the cell, the substance is deacetylated by esterases and cannot pass the cell membrane anymore. DCFH reacts with ROS like hydroxyl or peroxyl. Through the oxidation process 2, 7-dichloroﬂuorescein (DCF), which is highly ﬂuorescent, is generated [34].\n",
      "\n",
      "Figure 2. Scheme of experimental approaches of AC stimulation and calcium ions (Ca2+)-imaging in confocal microscopy. At ﬁrst, we stained the suspended MG-63s with ﬂuo3/AM for 40 min. Stained cells were then seeded into the 12-well plate under the microscope (LSM780, Carl Zeiss). Next, the electrodes were placed, and the electrical stimulation started. After a short settling phase of cells, a time series was done to measure the basal Ca2+ activity during ES (60 s). After 10 min ES, the electrodes were removed, and a second time series was recorded immediately to analyze the Ca2+ release and reaction after ATP addition (480 s; 240 cycles each 2 s, after 180 s ATP addition). (Parts of the ﬁgure were drawn by using pictures from Servier Medical Art. Servier Medical Art by Servier is licensed under a Creative Commons Attribution 3.0 Unported License (https://creativecommons. org/licenses/by/3.0/, accessed on 21 July 2022)).\n",
      "\n",
      "MG-63sweretrypsinized,centrifuged,andwashedwithPBS.Thecellpellet(1 × 106 cells) was resuspended in DCFH-DA staining solution (20 µM) and incubated using ThermoMixer C (Eppendorf; 300 rpm at 37 ◦C in the dark) for 30 min. Afterward, the stained MG-63s were washed, centrifuged, and resuspended in phenol-red free DMEM (without pyruvate, Life technologies, Thermo Fisher Scientific) with 10% FCS and gentamycin. 40 µL of stained cells were seeded on coverslips in the 12-well plates, electrodes were placed on them, and AC stimulation started, or not at Control samples. After ES, 5% hydrogen peroxide (H2O2; Sigma-Aldrich, Milan, Italy) was pipetted in one Control sample as a positive check.\n",
      "\n",
      "Immediately after stimulation, the DCF ﬂuorescence was recorded by the plate reader inﬁnite M200 (Tecan, Tecan i-control 1.9.17.0; Grödig, Austria) at an excitation of 485 nm and an emission of 535 nm. Subsequently, all samples were further cultivated in the incubator at 37 ◦C and removed at 0.5, 4, and 24 h for the ROS measurement. Four individual experi- ments with two respective technical replicates were done. The blank, i.e., the ﬂuorescence intensity of the phenol-red free medium, was subtracted from the sample values.\n",
      "\n",
      "2.11. Statistics\n",
      "\n",
      "All in vitro experiments were repeated at least three times (unless otherwise stated) with independently passaged cells. The GraphPad Prism 7 software for Windows (GraphPad Software Inc., La Jolla, CA, USA) was used to evaluate statistical differences in in vitro experiments. The data are illustrated in a boxplot showing the median, the minima, and the maxima (median ± min/max). In the ﬁrst step, we tested for normal distribution with the Shapiro–Wilk normality test (normal distribution: p > 0.05). Depending on the normal distribution, we used the parametric tests one-way ANOVA posthoc Bonferroni (paired values; medium characteristics, adhesion), one-way ANOVA posthoc uncorrected Fisher´s LSD (xCELLigence measurement, spreading, ROS production), or multiple t-tests (Ca2+ mobilization course).\n",
      "\n",
      "For non-parametric procedures (no normal distribution), we applied the Friedman test (paired values) posthoc uncorrected Dunn’s test (Ca2+-level). We set the signiﬁcance limit of p < 0.05 [35].\n",
      "\n",
      "------------------------------------------------------------\n",
      "{'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.8067933917045593, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 8, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'parent_id': '1dd6fd1918c016205b276e806647cbfb', 'category': 'Title', 'element_id': 'a5959447d22444d02f7b97fc38f5b372', 'section_header': '2. Materials and Methods', 'start_page': 2, 'end_page': 8, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': '7c7131c0-da93-4fc1-94e8-cd50f4417922'}\n",
      "------------------------------------------------------------\n",
      "\n",
      "3.1. Characterization of AC Electrical Stimulation (ES) Parameter\n",
      "\n",
      "The electric ﬁeld strength and its magnitude between the electrodes can be seen in Figure 3. Between the electrodes, where the cells are placed, is a large region with a more or less homogeneous ﬁeld. Only close to the electrode edges a ﬁeld enhancement can be observed (Figure 3a). However, the area with an enhanced ﬁeld is small in comparison to the area with a more or less constant ﬁeld strength. Along the centerline of the well, the ﬁeld decreases from 91.7 V/m at the electrodes to 90.2 V/m precisely at the midpoint between the electrodes (Figure 3b). Under the assumption of inﬁnitely wide capacitor plates, the ﬁeld could also be analytically estimated. In this case, it would be equal to the voltage difference over the electrode spacing. At a voltage difference of 1 V and a spacing of 1.1 cm, this corresponds to 90.91 V/m. It is of great interest to validate the numerical model. Since neither the electric ﬁeld nor the electric potential in the well can be measured easily, the current through the electrode needs to be compared. The current equals the conductivity times the normal component of the electric ﬁeld integrated over the electrode surface, yielding 1.75 A/m for the 2D geometry. With a 1 mL cell culture medium, a ﬁll level of 3.4 mm is reached. It should be noted that capillary effects are not considered here. Then, the current is 5.95 mA for 1 S/m conductivity and scales linearly with the conductivity. Consequently, it would be 11.9 mA for 2 S/m and so on.\n",
      "\n",
      "Figure 3. Simulation data revealed (a) the electric ﬁeld distribution and strength (red arrows) between the two electrodes at a voltage difference of 1 V. The grey isolines highlight the regions where the ﬁeld reaches 80, 90, and 100 V/m (the latter in the vicinity of the electrodes’ sharp corners), respectively. Furthermore, (b) shows the electric ﬁeld strength along the centerline between the two electrodes. The result indicates that the cells are exerted to a mostly homogeneous ﬁeld with a magnitude of about 90 V/m. Note that for different voltage differences, the ﬁeld strength changes linearly w.r.t. the voltage.\n",
      "\n",
      "The individual simulated and measured electric parameters are summarized in Table 1. The measured currents are peak values during the pulse. See Chapter 4 for a comparative discussion of these values.\n",
      "\n",
      "Table 1. Individual ES values after measurement and simulation. To each setup, appropriate colors were dedicated (Code).\n",
      "\n",
      " Table Summary: \n",
      " * Theoretical values calculated from the applied voltage and pulse duration. ** Measured values obtained from the experiment.\n",
      "\n",
      "This table compares theoretical and measured electric current values for different setups involving electrical stimulation. At 1V, a 20 Hz frequency with a 3.6 ms pulse duration resulted in a theoretical current of 90 A and a measured current of 0.012 A. Increasing both voltage to 5V and frequency to 20 Hz maintained the same pulse duration, leading to a theoretical current of 450 A and a measured current of 0.067 A. Similarly, at 5V but with a 7.9 Hz frequency and 10 ms pulse duration, the theoretical and measured currents were 450 A and 0.069 A, respectively. These results suggest that higher voltages and frequencies generally lead to higher currents, though measured values are consistently lower than theoretical predictions. Notably, the control setup lacks specific data.\n",
      "\n",
      "* Own calculation of electric ﬁeld from simulation. ** Own measurement of electric current in DMEM with 10% FCS using Tektronix TDS2014B.\n",
      "\n",
      "3.2. Acellular Medium Characteristics—AC-Stimulated Liquid\n",
      "\n",
      "Analyzing the effect of AC stimulation on the culture medium is a crucial step in order to exclude changes in liquid properties, such as pH value, as a reason for the resulting cellular adaption/behavior.\n",
      "\n",
      "The temperature of the AC-stimulated medium was signiﬁcantly changed vs. the Control only for the 5 V setups with 20 Hz and 7.9 Hz with 34.8 ± 0.3 ◦C and 34.6 ± 0.2 ◦C (mean ± s.e.m.), respectively. The mean value of the Control was 33.2 ± 0.3 ◦C (Figure 4a).\n",
      "\n",
      "Figure 4. Properties of AC-stimulated liquids, i.e., the complete DMEM, after 10 min ES: (a) tempera- ture, (b) pH, and (c) oxygen content. Note that only the temperature changed for the 5 V settings. (median ± min/max, n = 5–6, one-way ANOVA posthoc Bonferroni, * p < 0.05).\n",
      "\n",
      "The pH values in the medium after ES did not change for 1 and 5 V in combination with 20 Hz and 7.9 Hz (e.g., Control pH 7.40 ± 0.03, 5 V_20 Hz pH 7.42 ± 0.02; mean ± s.e.m.) (Figure 4b).\n",
      "\n",
      "The oxygen content in the medium was not changed after the 10 min AC stimulation (e.g., Control 20.2 ± 0.1%, 5 V_20 Hz 20.1 ± 0.09%; mean ± s.e.m.) (Figure 4c).\n",
      "\n",
      "During electrical stimulation, faradic byproducts such as H2O2 can be generated. H2O2 in the media can inﬂuence cellular fate and may contribute to ES-induced effects. In our experiments, the concentration of H2O2 was only increased in media stimulated with 5 V_7.9 Hz compared to the Control (Figure 5). The mean concentration increased from 1.48 µM ± 0.02 in Control samples to 2.52 µM ± 0.32 in 5 V_7.9 Hz samples directly after 10 min ES. The H2O2 concentration declined over time in all samples, with 5 V_ 7.9 Hz showing elevated levels also after 0.5 h (mean 2.05 ± 0.28 µM). After 4 h and 24 h, the concentration was similar in all samples and leveled off at about 0.38 µM (Figure 5a,b). The overall decline was independent of the O2 concentration in the media (data not shown).\n",
      "\n",
      "Figure 5. H2O2 content of AC-stimulated liquids, i.e., the DMEM w/o pyruvate, after 10 min ES, (a) for all AC-stimulation parameters as BoxPlot graph, (b) time course of H2O2 concentration. Note that 10 min AC with 1 V settings did not change the H2O2 concentration in contrast to 5 V_7.9 Hz with a signiﬁcant increase in H2O2 up to 0.5 h. (Median ± min/max, n = 5–6, two-way ANOVA posthoc Bonferroni, * p < 0.05 compared to all).\n",
      "\n",
      "3.3. AC-Stimulated Liquids and Long-Time Cell Adhesion\n",
      "\n",
      "In order to analyze whether AC stimulation of liquids alone (rather than the cells) affected cell adhesion; we recorded the impedance with xCELLigence RTCA over 24 h. The adhering cells interfere with the electrode structures on the chip, thus impeding the electron response in the medium. The impedance depends on the number of cells, the cell size, and the strength of adhering cells. As shown in Figure 6, cells were not inﬂuenced by the AC- stimulated liquid, and the cell adhesion was approximately the same as the unstimulated Control. Solely a slight increase was detectable with 1 V_20 Hz AC-stimulated liquids\n",
      "\n",
      "(Figure 6a). After 24 h, the max cell index (i.e., the 24 h endpoint value) in MG-63s was comparable in all liquids indicating no changes of the liquid DMEM due to AC stimulation (Control: 2.5 ± 0.1, 1 V_20 Hz: 2.7 ± 0.1, 1 V_7.9 Hz: 2.4 ± 0.1, 5 V_20 Hz: 2.5 ± 0.1, 5 V_7.9 Hz: 2.4 ± 0.1; mean ± s.e.m.; Figure 6b). In addition, the adhesion rate (time point when 50% of the cells adhere) was the same in all samples and showed no inﬂuence of the AC-stimulated liquid (Control, 1 V_20 Hz, 1 V_7.9 Hz, 5 V_7.9 Hz: all 0.5; 5 V_20 Hz: 0.4; mean; Figure 6c).\n",
      "\n",
      "Figure 6. AC-stimulated liquid DMEM and inﬂuence on a long-time MG-63 cell adhesion via impedance measurement. (a) Cell adhesion and spreading expressed by the “Cell Index” parameter over 24 h. (xCELLigence RTCA; 5 independent experiments, mean ± s.e.m.; multiple t-tests, n.s.) (b) Maximal Cell Index (values after 24 h), and (c) corresponding cell adhesion rate (Cell Index at the time point when 50% of the cells adhere). Note that AC-stimulated liquids did not inﬂuence the cell’s adhesion capacity. (xCELLigence, 5 independent experiments; median ± min/max; one-way ANOVA posthoc uncorrected Fisher’s LSD, n.s.).\n",
      "\n",
      "3.4. AC-Stimulated Cells\n",
      "\n",
      "3.4.1. Initial Cell Adhesion\n",
      "\n",
      "In the initial phase of a few minutes, the cellular adhesion of suspended cells starts with cell attachment of still rounded cells on the sample surface. In our approach, ﬂow cytometry was used to count the non-adhered cells, and then attached cells after 10 min AC stimulation were calculated in % of the adherent cells. In Figure 7, the inﬂuence of different electric ﬁeld parameters on initial cell adhesion is shown. It is clearly to be seen, that the cells were activated, and adhesion (in %) was accelerated under AC (1 V_20 Hz: 67.7 ± 5.9, 1 V_7.9 Hz: 67.8 ± 3.9, 5 V_20 Hz: 71 ± 4.3, 5 V_7.9 Hz: 61.9 ± 4.6, Control: 40.9 ± 5.7; mean ± s.e.m.) (Figure 7). Interestingly, the activation of cell adhesion is independent of the ﬁeld amplitude in the case of ES.\n",
      "\n",
      "Figure 7. The adhesion of MG-63s after 10 min AC stimulation of the suspended cells. Note that after AC, the attachment of osteoblasts was signiﬁcantly increased. (Cell count via FACSCalibur, median ± min/max, n = 7 independent experiments, one-way ANOVA posthoc Bonferroni, * p < 0.05 compared to Control).\n",
      "\n",
      "3.4.2. Intracellular Calcium Ion (Ca2+) Level during AC stimulation\n",
      "\n",
      "The intracellular Ca2+ signaling has been suggested to play a critical role in the impact of electric ﬁelds on cells. Therefore, we ﬁrst analyzed the immediate effect of AC stimulation on the Ca2+ level of ﬂuo-3-stained MG-63s. The mean ﬂuorescence intensity (MFI of Ca2+- signal) within 10 min AC stimulation of MG-63s is given in Figure 8. The AC electric ﬁeld activated the cells, and the level of calcium ions increased signiﬁcantly independent of the stimulation parameters used, but obviously with 20 Hz (1 V: 85.8 ± 0.2; 5 V: 90.1 ± 0.2; MFI mean ± s.e.m.) compared to the Control (78.5 ± 0.2; MFI mean ± s.e.m.).\n",
      "\n",
      "Figure 8. The level of intracellular calcium ions (Ca2+) in MG-63s within 10 min AC stimulation. Note that electrical stimulation activates the cells resulting in signiﬁcantly higher Ca2+ levels in cells compared to the unstimulated Control. (LSM780, Carl Zeiss; time series of 30 cycles each 2 s started after 5 min settling phase of MG-63s; 7 independent experiments of deﬁned areas of 10 cells, i.e., 70 cells per time point; median ± min/max; Friedman test posthoc uncorrected Dunn’s test, * p < 0.05 compared to Control).\n",
      "\n",
      "3.4.3. Calcium Ion Mobilization in AC-Pre-Stimulated Cells\n",
      "\n",
      "It is obvious from Figure 8 that cells are activated during the 10 min AC stimulation. Figure 9 presents the time course of the intracellular Ca2+ signal of MG-63s immediately after ES. The basal intracellular Ca2+ ion level (time frame 0–179 s) is enhanced 1.2 times in cells that were pre-stimulated with the higher frequency of 20 Hz (Figure 9a). After an additional stimulus with the energy-rich molecule ATP at 180 s, the Ca2+ ions were mobilized signiﬁcantly in these cells with a strong slope and a plateau phase (Figure 9a). The Ca2+ level after ATP increased signiﬁcantly for AC pre-stimulated cells with the settings 5 V_20 Hz, 1 V_20 Hz, 5 V_7.9 Hz with MFI values of 164.9 ± 10.23, 178.2 ± 9.5,\n",
      "\n",
      "and 162.5 ± 9.9, respectively; compared to Controls (MFI: 131 ± 9.1), and 1 V_7.9 Hz (MFI: 119.5 ± 8.9) (Figure 9b). These experiments were carried out in a calcium-containing medium. To solve the question of whether ES directly mobilizes the Ca2+ ions from internal stores during the period 0–179 s or if we observe an inﬂux of Ca2+ via the cell membrane, we used a calcium-free medium [36]. The additional experiments revealed that the basal Ca2+ level is also enhanced at least two times after ES (data not shown). Therefore, calcium ions are mobilized from intracellular stores.\n",
      "\n",
      "Figure 9. The intracellular calcium ion (Ca2+) signal of ﬂuo-3-loaded MG-63s after 10 min electrical stimulation. (a) The time course of the Ca2+ signal (0–480 s) was immediately started after the 10 min AC stimulation (0 s). The adenosine triphosphate (ATP) addition was at time point 180 s. Note that electrical stimulation with 20 Hz (1 and 5 V) and 5 V_7.9 Hz pre-activates the cells resulting in a signiﬁcantly increased mobilization of calcium ions after additional ATP. (Confocal microscope LSM780, Carl Zeiss; 6 independent experiments, deﬁned areas of 10 cells per time point; polygon line as mean; multiple t-test, * p < 0.05 compared to Control). (b) Adequate ﬂuorescence images of intracellular Ca2+ levels after ATP in electrically pre-stimulated cells (time point: 240 s). (LSM780, scale bars = 50 µm, 40× water objective).\n",
      "\n",
      "------------------------------------------------------------\n",
      "{'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.7357485890388489, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 13, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'category': 'Title', 'element_id': '55d05466ab31cd967316cbb1702b4345', 'section_header': '3. Results', 'start_page': 8, 'end_page': 13, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': 'e986d6bb-50eb-40e6-b096-f643bd1b4f6c'}\n",
      "------------------------------------------------------------\n",
      "\n",
      "Physical factors can inﬂuence the cell morphology, e.g., physical plasma can shorten the microvilli on the cell membrane or inhibit the cell area [37]. Therefore, the inﬂuence of AC electrical stimulation on cell morphology and spreading behavior was observed. Cells were 10 min AC-stimulated, and then the cell areas were analyzed after 0.5, 4, and 24 h (Figure 10). With our AC stimulation using different parameter settings, we could not identify any changes in cell morphology compared to unstimulated cells (Control) (Figure 10a). The ES neither hampered nor signiﬁcantly increased the spreading behavior (Figure 10b). A slight tendency of an increased cell area can be seen in 1 V_20 Hz cells after 4 and 24 h compared to Controls (Figure 10b). After 24 h the tendency of higher cell spreading can also be observed in 1 V_7.9 Hz stimulated cells (Table 2).\n",
      "\n",
      "14 of 22\n",
      "\n",
      "Figure 10. Inﬂuence of AC electrical stimulation (ES, 10 min) on MG-63 cell morphology and spreading at 0.5, 4, and 24 h. (a) Scanning electron microscopy of cells. Note that ES does not inﬂuence the overall cell shape. Cells start to spread at 0.5 h, which is typical for MG-63s. (FE-SEM Merlin compact, Carl Zeiss, 500×, scale bars = 10 µm). (b) Cell area analysis. Note that the different AC-stimulation settings have no signiﬁcant impact on the time-dependent spreading compared to unstimulated control cells (Control). (ImageJ, n = 40 cells, median ± min/max; one-way ANOVA posthoc uncorrected Fisher’s LSD, * p < 0.05).\n",
      "\n",
      "Table 2. Cell area values of 10 min AC-stimulated MG-63 osteoblastic cells (mean ± s.e.m.).\n",
      "\n",
      " Table Summary: \n",
      " The stimulation parameters were as follows: 1V_20 Hz (1 Volt, 20 Hertz), 1V_7.9 Hz (1 Volt, 7.9 Hertz), 5 V_20 Hz (5 Volts, 20 Hertz), 5 V_7.9 Hz (5 Volts, 7.9 Hertz). Control represents unstimulated cells.\n",
      "\n",
      "This table presents cell area measurements of MG-63 osteoblastic cells after various time points following AC stimulation. At 5 hours post-stimulation, the 5V_7.9 Hz condition showed the largest increase in cell area compared to other conditions and the control. By 24 hours, all stimulated groups had larger areas than the control, with the 5V_20 Hz and 5V_7.9 Hz conditions showing similar but slightly smaller increases compared to the 1V_20 Hz and 1V_7.9 Hz conditions. The 1V_7.9 Hz condition consistently resulted in the smallest increase in cell area across all time points. These results suggest that higher voltage and frequency combinations may enhance cell area expansion more effectively than lower settings.\n",
      "\n",
      "3.4.5. Reactive Oxygen Species (ROS) Production\n",
      "\n",
      "The stress level of MG-63s due to the AC stimulation was quantiﬁed using DCF-DA ROS detection assay over time: 0, 0.5, 4, and 24 h after ES. Figure 11a shows a signiﬁ- cant increase of the ROS level directly after 5 V_7.9 Hz stimulation (mean ﬂuorescence: 1412 ± 273) compared to the Control (mean ﬂuorescence: 935 ± 151) as well as all other AC-stimulated MG-63s. This increased ROS production at 5 V_7.9 Hz was still signiﬁcant after 0.5 h (mean ﬂuorescence 0.5 h: 955 ± 92) (Figure 11b) but negligible after 4 h (median ﬂuorescence 4 h: 1277 ± 267) compared to Control (mean ﬂuorescence 4 h: 964 ± 147) (Figure 11c). An increase of one ROS species—H2O2—had also been detectable in the liquid stimulated with 5 V_7.9 Hz within the ﬁrst minutes (see Section 3.2, Figure 5). After 4 h, however, the H2O2 concentrations in the media were approximately the same.\n",
      "\n",
      "Figure 11. Reactive oxygen species (ROS) production in MG-63s after AC stimulation (10 min) over time (a) 0 h, (b) 0.5 h, (c) 4 h, and (d) 24 h. Note that 1 V settings did not change the intracellular ROS up to 0.5 h. In contrast, an increased ROS production could be observed after 5 V_7.9 Hz stimulation up to 0.5 h compared to Controls. At 4 h, a signiﬁcant increase in ROS is then detectable in 1 V_20 Hz stimulated cells. However, after 24 h, all ROS levels were comparable to the Control. Note, the insert at 24 h plot (d) shows the maximum ROS production of cells treated with H2O2 (positive check, 5%). (DCF-DA test, Tecan, n = 4, median ± min/max; one-way ANOVA posthoc uncorrected Fisher’s LSD, * p < 0.05).\n",
      "\n",
      "Nevertheless, from the 4 h time point, a signiﬁcant increase in ROS production in MG-63s was detectable with 10 min of 1 V_20 Hz stimulation (mean ﬂuorescence 4 h: 1381 ± 96) (Figure 11c). However, after 24 h, all ROS levels were comparable to the Control (mean ﬂuorescence 24 h: 2029 ± 223), only among each other, a signiﬁcant increase in ROS level in MG-63s under 1 V_20 Hz stimulation (mean ﬂuorescence 24 h: 2496 ± 145) comparable to all other AC stimulations could be observed (Figure 11d). The fact that the overall intracellular ROS level is 2-fold higher after 24 h compared to the earlier time points can be explained by the doubling of cell numbers [38].\n",
      "\n",
      "Increased production of intracellular ROS could be induced within 4 h in 5 V_7.9 Hz stimulated MG-63s and after 4 h in 1 V_ 20 Hz stimulated cells. Treatment with H2O2, as a positive check, resulted in a pronounced production of ROS (6.4-fold to Control after 24 h) and thus cellular stress, which ﬁnally led to the detachment of MG-63s (data not shown).\n",
      "\n",
      "------------------------------------------------------------\n",
      "{'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.8448848724365234, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 15, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'category': 'Title', 'element_id': 'd76741b0c9220a7cc752d09d2e32e883', 'section_header': '3.4.4. Cell Morphology', 'start_page': 13, 'end_page': 15, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': '56e66a30-b5c3-4bd5-b0dc-809be6b86b87'}\n",
      "------------------------------------------------------------\n",
      "\n",
      "Electrical stimulation (ES) plays a role in physiological processes in various tissues such as nerves, heart, cartilage, muscles [39], and bone [1,2,5,38]. Therefore, exogenous ES is a popular and successful adjunct therapy in vitro as well as in vivo for bone regen- eration [39–45]. In the present study, we investigated the response of human osteoblasts (MG-63s) to 10 min ES by the multi-channel electrical stimulator of IonOptix (C-PACE EM) when seeding directly in the 12-well C-Dish chamber. The purpose of this study is threefold: (i) characterizing the electric ﬁelds; (ii) the inﬂuence of ES (AC) on liquid characteristics (DMEM); and (iii) inﬂuence of ES (AC) on the initial cell adhesion and calcium ion (Ca2+) signaling of MG-63s. ES generated by IonOptix is described in studies with, e.g., myoblasts [46], or osteoblasts [12,14]. Ercan et al. [12], showed a positive effect\n",
      "\n",
      "of IonOptix AC stimulation on osteoblast (hFOB 1.19) proliferation, but they technically modiﬁed the system with potentiometers to achieve a physiological current. In general, the comparability and interpretation of studies are very challenging due to the use of different systems (mostly homemade), type of ES, electrode conﬁguration, experimental timeline, stimulation parameters, and hence the resulting electric ﬁelds, or the absence of essential data [5,24,40,47]. Budde et al. [24], published a study to draw attention to these misunderstandings and to explain which documentation is necessary for in vitro experiments, such as detailed information about electrical devices, electrode material, and experimental setups like waveform or exposure duration. Simulations could then be used to identify and explain parameters that affect the physiology of the cells [24]. However, the optimal ES parameters and the underlying mechanism are still unclear concerning the impact of cell behavior [5].\n",
      "\n",
      "In the ﬁrst part, we characterized the C-Dish chamber and the generated electric ﬁelds. The electric ﬁeld is generated between two parallel graphite plates (i.e., a plate capacitor) 1.1 cm apart and in direct contact with the cells. We utilized a commercially available device from IonOptix (voltage controlled), which has been described earlier [12,14,46]. The IonOptix stimulator generates AC electric ﬁelds with constant biphasic rectangular pulses at a frequency of 7.9 or 20 Hz and voltage of 1 and 5 V, respectively.\n",
      "\n",
      "The electric ﬁeld (distribution and strength) depends on the electrode shape, its material, the surrounding media, and the type of ES. The simulation reveals a homogeneous ﬁeld strength of 90 V/m at 1 V with our C-Dish chamber. The system can be assumed to be linear. Thus, the ﬁeld should be about 450 V/m at 5 V. To prove this assumption, we measured the current at different voltages. Indeed, the current scales by about 5.6 when the voltage is increased by 5. Although this is not perfectly linear, it shows that the assumption of linearity is applicable.\n",
      "\n",
      "The current through the medium can be considered to validate the simulation and hence the correctness of the estimated ﬁeld [23]. We measured a current of 12 mA at 1V input amplitude. However, the simulations predicted a current of 7.7 mA using the measured conductivity of the cell culture medium (1.3 S/m). This deviation is unexpected but requires further analysis to identify possible reasons. For example, charging and discharging effects due to the capacitive electrode–electrolyte interface could change the shape of the current pulse. Possibilities to quantify the capacitive behavior are, for example, described in [23], but have not yet been implemented for the considered system. Other reasons could be measurement errors of the conductivity or the capillary effect in the dish, which has not been considered in the model.\n",
      "\n",
      "Moreover, direct current (DC) can damage host tissues due to electrochemical reactions such as local pH increase [43], oxygen reduction, hydrogen peroxide (H2O2) formation, as well as the formation of free radicals and reactive oxygen species (ROS; such as hydrogen peroxide, hydroxyl) [5,15,44,48]. Electrochemical reactions such as corrosion phenomena due to the electrode material can also be prevented with graphene electrodes [15,42,49]. Furthermore, the use of AC, instead of DC, and further the pulsed ES prevents ion and protein accumulation near the electrodes [5,50], as well as the indirect effects like a pH increase [5,14,42,51].\n",
      "\n",
      "For optimal growth of cells in vitro, we monitored pH, temperature, O2−, and H2O2 concentration of an acellular medium under ES in the present study. We could not detect any effects on pH and oxygen concentration in our AC-stimulated DMEM. Likewise, Ercan et al. [14], also found no signiﬁcant change in the medium, both pH and temperature, after 1 h of ES (IonOptix, 1 V_20 Hz). However, we detected a slight temperature increase of 1.4 ◦C at 7.9 Hz and 1.6 ◦C at 20 Hz in the medium during AC stimulation at 5 V. This is consistent with a simple estimate based on the electric energy deposited into each well during stimulation (~10 J), particularly when considering thermal losses to the environment. For 1 V AC stimulation, the medium temperature increased continuously by about 0.6 ◦C every 5 min. Reissis et al. [52], demonstrated that mesenchymal stem cells (hMSCs) retain their morphology, metabolic activity, and viability when exposed to temperatures ranging\n",
      "\n",
      "up to 48 ◦C for at least 150 s. With our temperature increase up to 1.6 ◦C of the medium by the ES, we can assume no cell damage. Our in vitro results conﬁrmed this assumption, as no morphological and growth impairment of MG-63s was detectable for all parameters after 10 min AC stimulation.\n",
      "\n",
      "The H2O2 concentration in the medium increased signiﬁcantly solely after stimulation with 5 V_7.9 Hz (2.52 µM). Srirussamee et al. [53], found H2O2 concentrations of acellular medium to be approximately 5 µM after two hours of DC stimulation. They applied the stimulated media to their cell culture, and the metabolic activity decreased in macrophages but not in MC3T3-E1 osteoblasts compared to the unstimulated Control. When ES was applied directly to the cells, all stimulated cells showed decreased metabolic activity [53]. An extracellular H2O2 concentration above 10 µM was described to lead to oxidative stress and cellular damage [54]. As we did not measure such high concentrations after ES, adverse effects through H2O2 are unlikely. Srirussamee et al. [53], suggested that the inﬂuence of other faradic byproducts in the media contributes to decreased metabolic activity after incubation with the stimulated medium. We hypothesized that the overall decline of H2O2 over 24 h may occur due to incubation at a 5% CO2 atmosphere in the incubator and consequently changed pH. The pH values of freshly opened medium and medium incubated for 24 h at standard culture conditions were unaltered. Only medium which had been open for several days at atmospheric CO2 pressure showed increased pH values which might correlate to increased H2O2 concentrations before incubation. The oxygen content in the media did not inﬂuence the H2O2 concentration (data not shown).\n",
      "\n",
      "Similar to Srirussamee et al. [53], we applied the 10 min AC-stimulated acellular medium to our cell culture and found no effect on osteoblast adhesion, spreading, and growth within the ﬁrst 24 h. We expected an effect because, e.g., another physical modiﬁcation—a 60 s physical plasma treatment (Argon-plasma jet kINPen09) of the liquid DMEM—signiﬁcantly affected cell vitality, microvilli, and tight junction formation [37]. Due to only minor changes in the temperature, but not for pH and oxygen content in the AC-stimulated medium, we can exclude this as a reason for the altered cell behavior. Balint et al. [43], also postulated that an inﬂuence of the surrounding medium on cell behavior can be excluded and that cells respond directly to the electric ﬁeld. Electric ﬁelds are apparently sensed through the same pathways observed in mechanotransduction and chemotaxis [43].\n",
      "\n",
      "Different cell types (like hMSC and osteoblasts) have been found to respond to various experimental ES conditions with altered cellular behaviors such as adhesion, proliferation, differentiation, and directed migration due to effects on signaling pathways [8,39,55]. A positive impact can only be achieved if the type of ES parameters such as intensity, frequency, and duration of electric ﬁeld were well chosen [5,39,40,42]. Our ES parameters were chosen from the literature, where positive inﬂuence on osteoblasts has already been reported [12,14,56]. Ercan et al. [12], found that an optimal voltage window for maximizing osteoblast density on nanotubular titanium is 1 V to 15 V and 20 Hz (IonOptix). In the study of Aaron et al. [50], electric ﬁeld amplitudes in the range of 10–1000 V/m were identiﬁed as sufﬁcient to achieve an effect without cell damage; we are also within this range.\n",
      "\n",
      "In our third part of the study, we demonstrated that cell adhesion was increased by 10 min ES, independent of AC-stimulation parameters. Enhanced cell attachment to surfaces due to ES has already been described in some studies [57–59]. Dubey et al. [60], showed that a 2 V pulsed ES strongly promoted the adhesion of L929 ﬁbroblasts, whereas, at 25 V, the cells died. It is also described that, among others, ES inﬂuences the arrangement and density of adhesion receptors—integrins [41,61,62], or focal adhesion [63] such as vinculin [58,64]. In the subsequent studies, we also plan to examine integrin activation, such as ß1 integrin (9EG7), to conﬁrm cell adhesion receptors’ involvement.\n",
      "\n",
      "Cell attachment could inﬂuence cell behavior like cell signaling. We detected increased basal Ca2+ activity for settling cells under ES. However, the effect of ES on Ca2+ mobiliza- tion was dependent on frequencies and voltages: while at 1 V_7.9 Hz, reduced tendency was evident, all other ES parameters showed a signiﬁcantly increased Ca2+ response after adenosine triphosphate (ATP) addition. Note that trypsinization of cells affects the calcium\n",
      "\n",
      "response after ATP addition [65], but we accept the systematic error in analyzing calcium signaling during and immediately after 10 min AC stimulation. A 10 min stimulation of 24 h adherent cells indicated no effect on calcium response, either basal or after ATP addition (data not shown). Many studies report that cells are sensitive to ES through activation of ion channels in the cell membrane, and release of Ca2+ from intracellular calcium stores like endoplasmic reticulum (ER) [8,13,15,39,42,43,45,57,66,67]. For exam- ple, Khatib et al. [68] described an increase in intracellular Ca2+ concentration in human osteoblasts after 20 min of DC stimulation at 2 V/cm due to an inﬂux of extracellular Ca2+ through stretch-activated cation channels (SACC), a release from intracellular stores, as well as activation of phospholipase C (PLC). We primarily suspect an enhanced release of Ca2+ from intracellular stores, as we observe a strong response of cells by the addition of ATP, which activates the PLC pathway via G-protein coupled receptor, and ﬁnally re- leases Ca2+ from the ER [30]. Extracellular Ca2+ inﬂux via ion channels is of secondary importance in our short 10 min stimulation, as analyses without extracellular Ca2+ sources from the HEPES buffer [36], also demonstrated an increased basal Ca2+ level in MG-63s in all ES parameters (data not shown). Zhou et al. [69], demonstrated that the intake of extracellular Ca2+ plays a decisive role in the increased calcium mobilization in MG-63s during nanosecond pulsed electric ﬁeld stimulation.\n",
      "\n",
      "In addition to Ca2+ signaling, many other cellular mechanisms, e.g., redox stress, are discussed to contribute to the effects of ES. Since direct ES stimulation at 5 V_7.9 Hz resulted in the generation of H2O2 in the acellular medium, we found the production of intracellular ROS in MG-63s within the ﬁrst minutes to rise accordingly, which was then equalized to the level of the unstimulated Control after 4 h. Wu et al. [70] described a linear relationship between ﬁeld strength and intracellular ROS production. However, the ﬁeld strength alone could not be causative in our case, as an increase in intracellular ROS was detectable at 5 V_7.9 Hz but not at 5 V_20 Hz. However, from 4 h, there was an increase in intracellular ROS level at 1 V_20 Hz, suggesting that a signaling pathway was turned on by ES, like Ca2+ signaling. At 1 V_20 Hz, we also had the strongest Ca2+ release in response to ATP addition. Thus, Görlach et al. [21] postulated that Ca2+ and ROS signaling are inter-regulated and crucial for cell signaling. Diaz-Vegas et al. [71] found the production of intracellular ROS after ES in muscle ﬁbers to be dependent on NOX2 activity. They found extracellular ATP to contribute to ROS production via P2Y1 receptors, also crosslinking Ca- and ROS-signaling. Studies indicated that intracellular ROS production could be increased by ES, acting as a signal transducer in physiological as well as pathological processes [15,57,72]. A 1.3-fold increase of ROS level in MG-63s at 1 V_20 Hz AC stimulation, compared to the Control (24 h), could be suggestive of a physiological level, according to our results, which triggers Ca2+ signaling. In contrast, the positive check (H2O2) showed a 6.4-fold increase in ROS production, leading to cell detachment and thus to a pathological process.\n",
      "\n",
      "Sun et al. [67] demonstrated that alteration of ion gradients or cell surface charges by an applied electric ﬁeld resulted in changes in Ca2+ signaling and subsequently increased gene expression of Col I & ALP but had no effect on hMSC growth rate or morphology. We also demonstrated the unaltered morphology of MG-63s under 10 min ES in FE-SEM images within 24 h. Furthermore, Sahm et al. [73] revealed in human primary osteoblast no altered proliferation and viability within 7 d AC stimulation on Ti6Al4V surfaces. Many studies describe improved cell survival and promote osseointegration (such synthesis of Collagen I) by ES [5,39,42,43,51,55,56,64,67]. Stimulation of osteoblasts with pulsed ES increased the calcium deposition after 3 weeks [14]. Furthermore, pulsed ES increased collagen production in pre-osteoblasts [5], demonstrating that pulsed stimulation can increase bone matrix deposition.\n",
      "\n",
      "In this study, we have analyzed and validated a method for ES stimulation of os- teoblasts in vitro based on direct contact of AC electric ﬁelds. The ﬁrst results show that the cells are sensitive to AC stimulation and react with altered cell adhesion, redox status, and Ca2+ signaling, which are not due to changes in the liquid properties. In various studies, including ours, a positive effect on cell physiology was found using ES, through\n",
      "\n",
      "the activation of signaling pathways [42], such as Ca2+ or PI3K/Akt pathway [5,41,55,74]. Nevertheless, the duration of stimulation, as well as the parameters, need to be further optimized and automatically documented to allow the best possible adaptation. Thus, Pettersen et al. [5] reviewed short-term studies lasting between 2 h and 5 days.\n",
      "\n",
      "In addition to the characterization and validation of the electrical stimulation system, the understanding of the molecular mechanisms of the electric ﬁelds is essential to optimize the stimulation parameters in bone healing.\n",
      "\n",
      "------------------------------------------------------------\n",
      "{'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.8590059876441956, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 19, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'parent_id': '5cff98ef86c9449c40d746607388de3f', 'category': 'Title', 'element_id': 'fa631b702d35563385205e6e468c8500', 'section_header': '4. Discussion', 'start_page': 15, 'end_page': 19, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': 'bcdc5b75-85d5-4543-9bab-7813f52b022d'}\n",
      "------------------------------------------------------------\n",
      "\n",
      "Precisely characterizing the electric ﬁelds and their inﬂuence on the environment is crucial to comparing and evaluating in vitro stimulation studies. The liquid DMEM electrically stimulated with 1 V (7.9 and 20 Hz) did not change their pH, temperature, H2O2, and O2 content characteristics. Therefore, the AC-stimulated liquid alone did not inﬂuence the long-time MG-63 cell adhesion measured via impedance.\n",
      "\n",
      "In contrast, the AC stimulation of suspended cells positively inﬂuences initial cell adhesion processes. Due to ES, the cells were more active, and a higher basal level of Ca2+ ions was detectable. Furthermore, an increase in voltage and pulse duration raises the Ca2+ mobilization and ROS production, which could trigger further signaling pathways.\n",
      "\n",
      "Electrical stimulation has the potential for novel treatments in medical applications.\n",
      "\n",
      "Author Contributions: Conceptualization, S.S. (Susanne Staehlke), B.N., S.S. (Sylvia Speller) and U.V.R.; methodology, S.S. (Susanne Staehlke), M.G., M.B., I.B., J.Z. and B.N.; software, I.B., J.Z. and S.S. (Susanne Staehlke); validation, S.S. (Susanne Staehlke), M.G., M.B., I.B., J.Z. and T.F.; for- mal analysis, S.S. (Susanne Staehlke), M.G., M.B., I.B., J.Z. and T.F.; investigation, S.S. (Susanne Staehlke), M.G., M.B., I.B., J.Z. and T.F.; resources, U.V.R., S.S. (Sylvia Speller) and B.N.; data cu- ration, S.S. (Susanne Staehlke), M.B., J.Z., I.B. and M.G.; writing—original draft preparation, S.S. (Susanne Staehlke), I.B., J.Z., M.B. and B.N.; writing—review and editing, S.S. (Susanne Staehlke), I.B., J.Z., M.B., S.S. (Sylvia Speller), M.G., U.V.R. and B.N.; visualization, S.S. (Susanne Staehlke), I.B., J.Z., M.B., S.S. (Sylvia Speller) and M.G.; supervision, S.S. (Susanne Staehlke), B.N., S.S. (Sylvia Speller) and U.V.R.; project administration, B.N., S.S. (Sylvia Speller) and U.V.R.; funding acquisition, B.N., S.S. (Sylvia Speller) and U.V.R. All authors have read and agreed to the published version of the manuscript.\n",
      "\n",
      "------------------------------------------------------------\n",
      "{'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.9470230937004089, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 19, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'parent_id': 'fa631b702d35563385205e6e468c8500', 'category': 'NarrativeText', 'element_id': 'c0f4541538a2cec1d73446f49db8584c', 'section_header': '5. Conclusions', 'subsection_header': '3.4.5. Reactive Oxygen Species (ROS) Production', 'start_page': 19, 'end_page': 19, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': '26acc14b-3182-4be2-9818-4bc9ba8d0438'}\n",
      "------------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Load new_docs\n",
    "\n",
    "import pickle\n",
    "\n",
    "with open(\"pkl_files/parent_docs.pkl\", \"rb\") as f:\n",
    "    new_docs=pickle.load(f)\n",
    "\n",
    "for doc in new_docs:\n",
    "    print(doc.page_content)\n",
    "    print(\"\\n\"+\"---\"*20+\"\\n\" + str(doc.metadata)+\"\\n\" +\"---\"*20 +\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2db6a6ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0dd19c2d-1b2c-4cca-822a-df636d4b0b78\n",
      "44aba20e-bd51-4964-950a-69aaf97680e9\n",
      "7c7131c0-da93-4fc1-94e8-cd50f4417922\n",
      "e986d6bb-50eb-40e6-b096-f643bd1b4f6c\n",
      "56e66a30-b5c3-4bd5-b0dc-809be6b86b87\n",
      "bcdc5b75-85d5-4543-9bab-7813f52b022d\n",
      "26acc14b-3182-4be2-9818-4bc9ba8d0438\n"
     ]
    }
   ],
   "source": [
    "for doc in new_docs:\n",
    "    print(doc.metadata.get('document_id'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0fbacd7",
   "metadata": {},
   "source": [
    "### Crating parent and child chunks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9144b175-6e0c-4c01-90bc-010bde6f910a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoTokenizer\n",
    "model_name=\"nvidia/llama-embed-nemotron-8b\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "\n",
    "def token_length(text: str) -> int:\n",
    "    return len(tokenizer.encode(text, add_special_tokens=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b64331d3-7c54-403b-8be6-f5f3849f9c3d",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'new_docs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, doc \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[43mnew_docs\u001b[49m):\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;28mprint\u001b[39m(i, token_length(doc\u001b[38;5;241m.\u001b[39mpage_content))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'new_docs' is not defined"
     ]
    }
   ],
   "source": [
    "for i, doc in enumerate(new_docs):\n",
    "    print(i, token_length(doc.page_content))"
   ]
  },
  {
   "cell_type": "raw",
   "id": "0b4bfa26-7a49-467e-bd55-32578a35e67f",
   "metadata": {},
   "source": [
    "!pip install langchain-chroma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6ff4854d-e214-4c18-a561-372e7b045f5d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pydantic in /opt/conda/lib/python3.11/site-packages (1.10.9)\n",
      "Collecting pydantic\n",
      "  Using cached pydantic-2.12.5-py3-none-any.whl.metadata (90 kB)\n",
      "Requirement already satisfied: chromadb in /opt/conda/lib/python3.11/site-packages (1.4.1)\n",
      "Requirement already satisfied: langchain-chroma in /opt/conda/lib/python3.11/site-packages (1.1.0)\n",
      "Requirement already satisfied: langchain in /opt/conda/lib/python3.11/site-packages (1.2.6)\n",
      "Requirement already satisfied: annotated-types>=0.6.0 in /opt/conda/lib/python3.11/site-packages (from pydantic) (0.7.0)\n",
      "Requirement already satisfied: pydantic-core==2.41.5 in /opt/conda/lib/python3.11/site-packages (from pydantic) (2.41.5)\n",
      "Requirement already satisfied: typing-extensions>=4.14.1 in /opt/conda/lib/python3.11/site-packages (from pydantic) (4.15.0)\n",
      "Requirement already satisfied: typing-inspection>=0.4.2 in /opt/conda/lib/python3.11/site-packages (from pydantic) (0.4.2)\n",
      "Requirement already satisfied: build>=1.0.3 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.4.0)\n",
      "Requirement already satisfied: pybase64>=1.4.1 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.4.3)\n",
      "Requirement already satisfied: uvicorn>=0.18.3 in /opt/conda/lib/python3.11/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (0.40.0)\n",
      "Requirement already satisfied: numpy>=1.22.5 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.26.4)\n",
      "Requirement already satisfied: posthog<6.0.0,>=2.4.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (5.4.0)\n",
      "Requirement already satisfied: onnxruntime>=1.14.1 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.16.3)\n",
      "Requirement already satisfied: opentelemetry-api>=1.2.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.39.1)\n",
      "Requirement already satisfied: opentelemetry-exporter-otlp-proto-grpc>=1.2.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.39.1)\n",
      "Requirement already satisfied: opentelemetry-sdk>=1.2.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.39.1)\n",
      "Requirement already satisfied: tokenizers>=0.13.2 in /opt/conda/lib/python3.11/site-packages (from chromadb) (0.22.2)\n",
      "Requirement already satisfied: pypika>=0.48.9 in /opt/conda/lib/python3.11/site-packages (from chromadb) (0.50.0)\n",
      "Requirement already satisfied: tqdm>=4.65.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (4.66.1)\n",
      "Requirement already satisfied: overrides>=7.3.1 in /opt/conda/lib/python3.11/site-packages (from chromadb) (7.6.0)\n",
      "Requirement already satisfied: importlib-resources in /opt/conda/lib/python3.11/site-packages (from chromadb) (6.1.1)\n",
      "Requirement already satisfied: grpcio>=1.58.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (1.76.0)\n",
      "Requirement already satisfied: bcrypt>=4.0.1 in /opt/conda/lib/python3.11/site-packages (from chromadb) (5.0.0)\n",
      "Requirement already satisfied: typer>=0.9.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (0.21.1)\n",
      "Requirement already satisfied: kubernetes>=28.1.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (35.0.0)\n",
      "Requirement already satisfied: tenacity>=8.2.3 in /opt/conda/lib/python3.11/site-packages (from chromadb) (8.5.0)\n",
      "Requirement already satisfied: pyyaml>=6.0.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (6.0.1)\n",
      "Requirement already satisfied: mmh3>=4.0.1 in /opt/conda/lib/python3.11/site-packages (from chromadb) (5.2.0)\n",
      "Requirement already satisfied: orjson>=3.9.12 in /opt/conda/lib/python3.11/site-packages (from chromadb) (3.11.5)\n",
      "Requirement already satisfied: httpx>=0.27.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (0.28.1)\n",
      "Requirement already satisfied: rich>=10.11.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (14.2.0)\n",
      "Requirement already satisfied: jsonschema>=4.19.0 in /opt/conda/lib/python3.11/site-packages (from chromadb) (4.21.1)\n",
      "Requirement already satisfied: langchain-core<2.0.0,>=1.1.3 in /opt/conda/lib/python3.11/site-packages (from langchain-chroma) (1.2.7)\n",
      "Requirement already satisfied: langgraph<1.1.0,>=1.0.2 in /opt/conda/lib/python3.11/site-packages (from langchain) (1.0.6)\n",
      "Requirement already satisfied: packaging>=24.0 in /opt/conda/lib/python3.11/site-packages (from build>=1.0.3->chromadb) (25.0)\n",
      "Requirement already satisfied: pyproject_hooks in /opt/conda/lib/python3.11/site-packages (from build>=1.0.3->chromadb) (1.2.0)\n",
      "Requirement already satisfied: anyio in /opt/conda/lib/python3.11/site-packages (from httpx>=0.27.0->chromadb) (4.2.0)\n",
      "Requirement already satisfied: certifi in /opt/conda/lib/python3.11/site-packages (from httpx>=0.27.0->chromadb) (2026.1.4)\n",
      "Requirement already satisfied: httpcore==1.* in /opt/conda/lib/python3.11/site-packages (from httpx>=0.27.0->chromadb) (1.0.9)\n",
      "Requirement already satisfied: idna in /opt/conda/lib/python3.11/site-packages (from httpx>=0.27.0->chromadb) (3.6)\n",
      "Requirement already satisfied: h11>=0.16 in /opt/conda/lib/python3.11/site-packages (from httpcore==1.*->httpx>=0.27.0->chromadb) (0.16.0)\n",
      "Requirement already satisfied: attrs>=22.2.0 in /opt/conda/lib/python3.11/site-packages (from jsonschema>=4.19.0->chromadb) (23.2.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /opt/conda/lib/python3.11/site-packages (from jsonschema>=4.19.0->chromadb) (2023.12.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /opt/conda/lib/python3.11/site-packages (from jsonschema>=4.19.0->chromadb) (0.32.1)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /opt/conda/lib/python3.11/site-packages (from jsonschema>=4.19.0->chromadb) (0.17.1)\n",
      "Requirement already satisfied: six>=1.9.0 in /opt/conda/lib/python3.11/site-packages (from kubernetes>=28.1.0->chromadb) (1.16.0)\n",
      "Requirement already satisfied: python-dateutil>=2.5.3 in /opt/conda/lib/python3.11/site-packages (from kubernetes>=28.1.0->chromadb) (2.8.2)\n",
      "Requirement already satisfied: websocket-client!=0.40.0,!=0.41.*,!=0.42.*,>=0.32.0 in /opt/conda/lib/python3.11/site-packages (from kubernetes>=28.1.0->chromadb) (1.7.0)\n",
      "Requirement already satisfied: requests in /opt/conda/lib/python3.11/site-packages (from kubernetes>=28.1.0->chromadb) (2.32.5)\n",
      "Requirement already satisfied: requests-oauthlib in /opt/conda/lib/python3.11/site-packages (from kubernetes>=28.1.0->chromadb) (2.0.0)\n",
      "Requirement already satisfied: urllib3!=2.6.0,>=1.24.2 in /opt/conda/lib/python3.11/site-packages (from kubernetes>=28.1.0->chromadb) (2.1.0)\n",
      "Requirement already satisfied: durationpy>=0.7 in /opt/conda/lib/python3.11/site-packages (from kubernetes>=28.1.0->chromadb) (0.10)\n",
      "Requirement already satisfied: jsonpatch<2.0.0,>=1.33.0 in /opt/conda/lib/python3.11/site-packages (from langchain-core<2.0.0,>=1.1.3->langchain-chroma) (1.33)\n",
      "Requirement already satisfied: langsmith<1.0.0,>=0.3.45 in /opt/conda/lib/python3.11/site-packages (from langchain-core<2.0.0,>=1.1.3->langchain-chroma) (0.6.4)\n",
      "Requirement already satisfied: uuid-utils<1.0,>=0.12.0 in /opt/conda/lib/python3.11/site-packages (from langchain-core<2.0.0,>=1.1.3->langchain-chroma) (0.13.0)\n",
      "Requirement already satisfied: langgraph-checkpoint<5.0.0,>=2.1.0 in /opt/conda/lib/python3.11/site-packages (from langgraph<1.1.0,>=1.0.2->langchain) (4.0.0)\n",
      "Requirement already satisfied: langgraph-prebuilt<1.1.0,>=1.0.2 in /opt/conda/lib/python3.11/site-packages (from langgraph<1.1.0,>=1.0.2->langchain) (1.0.6)\n",
      "Requirement already satisfied: langgraph-sdk<0.4.0,>=0.3.0 in /opt/conda/lib/python3.11/site-packages (from langgraph<1.1.0,>=1.0.2->langchain) (0.3.3)\n",
      "Requirement already satisfied: xxhash>=3.5.0 in /opt/conda/lib/python3.11/site-packages (from langgraph<1.1.0,>=1.0.2->langchain) (3.6.0)\n",
      "Requirement already satisfied: coloredlogs in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.14.1->chromadb) (15.0.1)\n",
      "Requirement already satisfied: flatbuffers in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.14.1->chromadb) (25.12.19)\n",
      "Requirement already satisfied: protobuf in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.14.1->chromadb) (6.33.4)\n",
      "Requirement already satisfied: sympy in /opt/conda/lib/python3.11/site-packages (from onnxruntime>=1.14.1->chromadb) (1.14.0)\n",
      "Requirement already satisfied: importlib-metadata<8.8.0,>=6.0 in /opt/conda/lib/python3.11/site-packages (from opentelemetry-api>=1.2.0->chromadb) (7.0.1)\n",
      "Requirement already satisfied: googleapis-common-protos~=1.57 in /opt/conda/lib/python3.11/site-packages (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb) (1.72.0)\n",
      "Requirement already satisfied: opentelemetry-exporter-otlp-proto-common==1.39.1 in /opt/conda/lib/python3.11/site-packages (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb) (1.39.1)\n",
      "Requirement already satisfied: opentelemetry-proto==1.39.1 in /opt/conda/lib/python3.11/site-packages (from opentelemetry-exporter-otlp-proto-grpc>=1.2.0->chromadb) (1.39.1)\n",
      "Requirement already satisfied: opentelemetry-semantic-conventions==0.60b1 in /opt/conda/lib/python3.11/site-packages (from opentelemetry-sdk>=1.2.0->chromadb) (0.60b1)\n",
      "Requirement already satisfied: backoff>=1.10.0 in /opt/conda/lib/python3.11/site-packages (from posthog<6.0.0,>=2.4.0->chromadb) (2.2.1)\n",
      "Requirement already satisfied: distro>=1.5.0 in /opt/conda/lib/python3.11/site-packages (from posthog<6.0.0,>=2.4.0->chromadb) (1.9.0)\n",
      "Requirement already satisfied: markdown-it-py>=2.2.0 in /opt/conda/lib/python3.11/site-packages (from rich>=10.11.0->chromadb) (4.0.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /opt/conda/lib/python3.11/site-packages (from rich>=10.11.0->chromadb) (2.17.2)\n",
      "Requirement already satisfied: huggingface-hub<2.0,>=0.16.4 in /opt/conda/lib/python3.11/site-packages (from tokenizers>=0.13.2->chromadb) (0.36.0)\n",
      "Requirement already satisfied: click>=8.0.0 in /opt/conda/lib/python3.11/site-packages (from typer>=0.9.0->chromadb) (8.1.7)\n",
      "Requirement already satisfied: shellingham>=1.3.0 in /opt/conda/lib/python3.11/site-packages (from typer>=0.9.0->chromadb) (1.5.4)\n",
      "Requirement already satisfied: httptools>=0.6.3 in /opt/conda/lib/python3.11/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (0.7.1)\n",
      "Requirement already satisfied: python-dotenv>=0.13 in /opt/conda/lib/python3.11/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (1.2.1)\n",
      "Requirement already satisfied: uvloop>=0.15.1 in /opt/conda/lib/python3.11/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (0.22.1)\n",
      "Requirement already satisfied: watchfiles>=0.13 in /opt/conda/lib/python3.11/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (1.1.1)\n",
      "Requirement already satisfied: websockets>=10.4 in /opt/conda/lib/python3.11/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (16.0)\n",
      "Requirement already satisfied: filelock in /opt/conda/lib/python3.11/site-packages (from huggingface-hub<2.0,>=0.16.4->tokenizers>=0.13.2->chromadb) (3.13.1)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in /opt/conda/lib/python3.11/site-packages (from huggingface-hub<2.0,>=0.16.4->tokenizers>=0.13.2->chromadb) (2023.12.2)\n",
      "Requirement already satisfied: hf-xet<2.0.0,>=1.1.3 in /opt/conda/lib/python3.11/site-packages (from huggingface-hub<2.0,>=0.16.4->tokenizers>=0.13.2->chromadb) (1.2.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /opt/conda/lib/python3.11/site-packages (from importlib-metadata<8.8.0,>=6.0->opentelemetry-api>=1.2.0->chromadb) (3.17.0)\n",
      "Requirement already satisfied: jsonpointer>=1.9 in /opt/conda/lib/python3.11/site-packages (from jsonpatch<2.0.0,>=1.33.0->langchain-core<2.0.0,>=1.1.3->langchain-chroma) (2.4)\n",
      "Requirement already satisfied: ormsgpack>=1.12.0 in /opt/conda/lib/python3.11/site-packages (from langgraph-checkpoint<5.0.0,>=2.1.0->langgraph<1.1.0,>=1.0.2->langchain) (1.12.1)\n",
      "Requirement already satisfied: requests-toolbelt>=1.0.0 in /opt/conda/lib/python3.11/site-packages (from langsmith<1.0.0,>=0.3.45->langchain-core<2.0.0,>=1.1.3->langchain-chroma) (1.0.0)\n",
      "Requirement already satisfied: zstandard>=0.23.0 in /opt/conda/lib/python3.11/site-packages (from langsmith<1.0.0,>=0.3.45->langchain-core<2.0.0,>=1.1.3->langchain-chroma) (0.25.0)\n",
      "Requirement already satisfied: mdurl~=0.1 in /opt/conda/lib/python3.11/site-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->chromadb) (0.1.2)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in /opt/conda/lib/python3.11/site-packages (from requests->kubernetes>=28.1.0->chromadb) (3.3.2)\n",
      "Requirement already satisfied: sniffio>=1.1 in /opt/conda/lib/python3.11/site-packages (from anyio->httpx>=0.27.0->chromadb) (1.3.0)\n",
      "Requirement already satisfied: humanfriendly>=9.1 in /opt/conda/lib/python3.11/site-packages (from coloredlogs->onnxruntime>=1.14.1->chromadb) (10.0)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /opt/conda/lib/python3.11/site-packages (from requests-oauthlib->kubernetes>=28.1.0->chromadb) (3.2.2)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /opt/conda/lib/python3.11/site-packages (from sympy->onnxruntime>=1.14.1->chromadb) (1.3.0)\n",
      "Using cached pydantic-2.12.5-py3-none-any.whl (463 kB)\n",
      "Installing collected packages: pydantic\n",
      "  Attempting uninstall: pydantic\n",
      "    Found existing installation: pydantic 1.10.9\n",
      "    Uninstalling pydantic-1.10.9:\n",
      "      Successfully uninstalled pydantic-1.10.9\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "langchain-classic 1.0.1 requires langchain-text-splitters<2.0.0,>=1.1.0, but you have langchain-text-splitters 0.2.4 which is incompatible.\n",
      "langchain-huggingface 0.0.3 requires langchain-core<0.3,>=0.1.52, but you have langchain-core 1.2.7 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed pydantic-2.12.5\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install --upgrade pydantic chromadb langchain-chroma langchain\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "98b43ba7-d651-4cec-aba0-9086b1927f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip freeze > requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a2b41ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_chroma import Chroma\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langgraph.store.memory import InMemoryStore\n",
    "from langchain_classic.retrievers import ParentDocumentRetriever\n",
    "from langchain_core.documents import Document\n",
    "\n",
    "parent_splitter = RecursiveCharacterTextSplitter(chunk_size = 2000, \n",
    "                                                 chunk_overlap =100,\n",
    "                                                 separators=[\"\\n\\n\", \"\\n\", \" \", \"\"],\n",
    "                                                 length_function = token_length,\n",
    "                                                 add_start_index = True)\n",
    "\n",
    "child_splitter = RecursiveCharacterTextSplitter(chunk_size = 400, \n",
    "                                                chunk_overlap = 20,\n",
    "                                                length_function=token_length,\n",
    "                                                add_start_index = True)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c62abdd8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "347 0dd19c2d-1b2c-4cca-822a-df636d4b0b78\n",
      "762 44aba20e-bd51-4964-950a-69aaf97680e9\n",
      "1952 7c7131c0-da93-4fc1-94e8-cd50f4417922\n",
      "1909 7c7131c0-da93-4fc1-94e8-cd50f4417922\n",
      "593 7c7131c0-da93-4fc1-94e8-cd50f4417922\n",
      "1956 e986d6bb-50eb-40e6-b096-f643bd1b4f6c\n",
      "1483 e986d6bb-50eb-40e6-b096-f643bd1b4f6c\n",
      "1501 56e66a30-b5c3-4bd5-b0dc-809be6b86b87\n",
      "1897 bcdc5b75-85d5-4543-9bab-7813f52b022d\n",
      "1751 bcdc5b75-85d5-4543-9bab-7813f52b022d\n",
      "609 26acc14b-3182-4be2-9818-4bc9ba8d0438\n"
     ]
    }
   ],
   "source": [
    "parent_chunks = parent_splitter.split_documents(new_docs)\n",
    "child_chunks = child_splitter.split_documents(new_docs)\n",
    "for p in parent_chunks:\n",
    "    print(token_length(p.page_content), p.metadata.get('document_id'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "61b03baf-87b1-4509-b665-345803fd687b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "347\n",
      "271\n",
      "266\n",
      "224\n",
      "384\n",
      "380\n",
      "384\n",
      "304\n",
      "394\n",
      "372\n",
      "278\n",
      "314\n",
      "94\n",
      "399\n",
      "250\n",
      "395\n",
      "356\n",
      "242\n",
      "375\n",
      "228\n",
      "360\n",
      "227\n",
      "388\n",
      "394\n",
      "191\n",
      "351\n",
      "366\n",
      "384\n",
      "237\n",
      "215\n",
      "307\n",
      "171\n",
      "307\n",
      "225\n",
      "293\n",
      "248\n",
      "328\n",
      "300\n",
      "388\n",
      "78\n",
      "336\n",
      "217\n",
      "221\n",
      "285\n",
      "373\n",
      "399\n",
      "64\n",
      "391\n",
      "37\n",
      "185\n",
      "398\n",
      "43\n",
      "total child chunks=  52\n"
     ]
    }
   ],
   "source": [
    "for p in child_chunks:\n",
    "    print(token_length(p.page_content))\n",
    "print(\"total child chunks= \",len(child_chunks))"
   ]
  },
  {
   "cell_type": "raw",
   "id": "918713c2-eccc-41de-92ac-cbdbd0a2eb55",
   "metadata": {},
   "source": [
    "!pip install sentence-transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0b892d1d-8dd2-47c7-93f7-1228aba127c9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "A new version of the following files was downloaded from https://huggingface.co/nvidia/llama-embed-nemotron-8b:\n",
      "- llama_bidirectional_model.py\n",
      ". Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.\n",
      "Fetching 4 files: 100%|██████████| 4/4 [00:25<00:00,  6.40s/it]\n",
      "Loading checkpoint shards: 100%|██████████| 4/4 [00:01<00:00,  2.37it/s]\n"
     ]
    }
   ],
   "source": [
    "from langchain_huggingface import HuggingFaceEmbeddings\n",
    "\n",
    "model_name=\"nvidia/llama-embed-nemotron-8b\"\n",
    "\n",
    "embeddings = HuggingFaceEmbeddings(model_name=model_name,\n",
    "                                      model_kwargs={'device': 0, \"trust_remote_code\": True} )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ed41ccd5-3e48-4f30-8786-7fe29d068dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "vector_store = Chroma(embedding_function=embeddings,\n",
    "                      collection_name=\"biorag_indexed_docs\",  # Giving vectore store a name\n",
    "                      persist_directory=\"./chroma_db/\",  # Creeating a dirtectory to store the vecotr store in the computer\n",
    "                      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ca00b4ee-4e0c-4f2e-bd4d-28aa38327011",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining the orchestrator to manage parent-child relationship between documents\n",
    "from langchain_community.docstore import InMemoryDocstore\n",
    "from langchain_core.stores import InMemoryByteStore\n",
    "# Two different stores for two jobs\n",
    "byte_store = InMemoryByteStore()    # Embeddings → Fast search\n",
    "docstore = InMemoryDocstore()       # Full docs → Rich context\n",
    "\n",
    "retriever = ParentDocumentRetriever(\n",
    "    vectorstore=vector_store,    # Uses byte_store internally\n",
    "    docstore=byte_store,           # Full documents\n",
    "    child_splitter=child_splitter,\n",
    "    parent_splitter=parent_splitter,\n",
    "    search_kwargs={\"k\": 3}\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8bd17f34-0c11-4913-b9d7-6d195ccb16a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "langchain_classic.retrievers.parent_document_retriever.ParentDocumentRetriever"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(retriever)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e558d307-3266-4820-90fc-4ddbdf359a4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "retriever.add_documents(new_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ac5545e4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "3. Results\n",
      "4. Discussion\n"
     ]
    }
   ],
   "source": [
    "# Retrieve Documents relevent to query\n",
    "\n",
    "query = \"Compare the simulated electric field outcomes with the measured electric current.\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f10132d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "3.4.4. Cell Morphology\n",
      "4. Discussion\n",
      "2. Materials and Methods\n"
     ]
    }
   ],
   "source": [
    "# Retrieve Documents relevent to query\n",
    "\n",
    "query = \"What is Cell Morphology and its significance in the study?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "444a3e61-43f9-416b-a8be-fb876f5d0422",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "5. Conclusions\n",
      "4. Discussion\n"
     ]
    }
   ],
   "source": [
    "# Retrieve Documents relevent to query\n",
    "# Target: Abstract/ Conclusion\n",
    "\n",
    "query = \"What is the key findings of the paper?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c1fa99cc-5964-4d17-b22a-d94956f57b2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "4. Discussion\n",
      "4. Discussion\n",
      "1. Introduction\n"
     ]
    }
   ],
   "source": [
    "# Target: Abstract\n",
    "query = \"What was the main objective of this study regarding pulsed electrical stimulation and osteoblast behavior?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a4748f1a-c4d4-4a49-8286-3eda9096e158",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "4. Discussion\n",
      "1. Introduction\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Target: Introduction\n",
    "query = \"Why is electrical stimulation considered relevant for bone regeneration and osteoblast activity?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9e4f0fbb-eb00-4e5c-8f30-a52ebc33dfce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "2. Materials and Methods\n",
      "4. Discussion\n"
     ]
    }
   ],
   "source": [
    "# Target: Section 2.1 – Electrical Stimulation Chamber section 2\n",
    "query = \"In the Materials and Methods section, how is the IonOptix electrical stimulation chamber described, including electrode material, geometry, and well configuration?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c0c64977-1adf-4532-87f1-2b3cabeea443",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "3. Results\n",
      "2. Materials and Methods\n"
     ]
    }
   ],
   "source": [
    "# Target : Section 2.2 + 3.1\n",
    "query = \"How was the electric field strength between the electrodes calculated and validated in this study?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5fb48ac8-fbf8-43f8-a3e4-1020c308dbae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "4. Discussion\n",
      "3. Results\n",
      "5. Conclusions\n"
     ]
    }
   ],
   "source": [
    "## The retrievd chunk should be from section 2 2.4 current measurements\n",
    "\n",
    "query = \"How was the electrical current measured during stimulation, and why was this measurement important?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "60846b67-aae2-49e9-8d6c-c84ce94677c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "3. Results\n",
      "5. Conclusions\n",
      "4. Discussion\n"
     ]
    }
   ],
   "source": [
    "# Target: Section 2.4 + 3.2\n",
    "query = \"Which physicochemical properties of the culture medium were analyzed after AC stimulation, and what changes were observed?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b8c50a09-853b-4686-a47a-63a8f457c406",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "2. Materials and Methods\n",
      "2. Materials and Methods\n"
     ]
    }
   ],
   "source": [
    "# Target: Section 2.5 – Cell Culture\n",
    "\n",
    "query = \"Why were MG-63 osteoblast-like cells chosen, and how were they prepared for the experiments?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6b38c4ad-b455-4a30-8ea2-4cb5dbb3ee7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "2. Materials and Methods\n",
      "3. Results\n",
      "4. Discussion\n"
     ]
    }
   ],
   "source": [
    "#Target: Sections 3.4.2 & 3.4.3- Calcium signaling results\n",
    "\n",
    "query = \"How did pulsed electrical stimulation affect the initial adhesion of suspended osteoblasts?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "375f8a1a-e45e-480e-bad4-2fff1198ae39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "3. Results\n",
      "4. Discussion\n"
     ]
    }
   ],
   "source": [
    "#target: Section 3.4.5 – ROS Production\n",
    "\n",
    "query = \"What effect did AC electrical stimulation have on intracellular calcium levels and ATP-induced calcium mobilization?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "eba38e9f-a38c-47fc-bf50-bb1faae02d40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "1. Introduction\n",
      "4. Discussion\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Target:  Discussion section\n",
    "\n",
    "query = \"How do the authors explain the relationship between electrical stimulation, calcium signaling, and osteoblast activation?\"\n",
    "relevent_docs = retriever.invoke(query)\n",
    "print(len(relevent_docs))\n",
    "for i in relevent_docs:\n",
    "    print(i.metadata.get('section_header'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5548ce4f-0e0d-406f-a27e-f765fc9aba91",
   "metadata": {},
   "source": [
    "#### In above evaluation, i was cheking if the retriever gets the most relevent chunks from the document or not. Here is what I observed:\n",
    "* The discussion section contains information for materials section as well as result section.\n",
    "* Even when I ask question related to 2. section it retrieves discussion section.\n",
    "* Most favorable approach here is to use advanced RAG techniques called query re-writting.\n",
    "* Below in few experiments I will try to do few shot and zero shot query re-writting(prompting technique)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e486d57d-4095-4cf1-886b-015df4dea2cb",
   "metadata": {},
   "source": [
    "## Strategy- 2 : Rewrite-Retrieve-Read method"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d089340-c583-4c4d-b9d5-facf4bc0381d",
   "metadata": {},
   "source": [
    "### Zero shot- Rewriting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3a53a5a9-68de-42d0-b048-18e2d4f705bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9a08f65a9898472e8e3b922fc821b4d7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Device set to use cuda:0\n",
      "The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.\n"
     ]
    }
   ],
   "source": [
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "from langchain_core.output_parsers import StrOutputParser\n",
    "from langchain_huggingface.llms import HuggingFacePipeline\n",
    "from transformers import AutoTokenizer, pipeline, AutoModelForCausalLM\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "# defining environment variable\n",
    "load_dotenv()\n",
    "\n",
    "model_name = \"meta-llama/Llama-3.1-8B-Instruct\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "#model = AutoModelForCausalLM.from_pretrained(model_name)\n",
    "pipe = pipeline(\"text-generation\",\n",
    "                tokenizer=tokenizer,\n",
    "               model= model_name,\n",
    "                device_map = \"auto\",\n",
    "                dtype = \"auto\",\n",
    "                max_new_tokens=300,   \n",
    "                do_sample=False,\n",
    "                temperature = 0.4,\n",
    "                repetition_penalty=1.2\n",
    "               )\n",
    "\n",
    "model_pipe = HuggingFacePipeline(pipeline=pipe)\n",
    "model = model_pipe.bind(skip_prompt = True,\n",
    "                       stop=[\"\\n\\n\\n\", \"Note:\", \"Query 4:\", \"Input:\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ac3aaca2-6634-4318-87d7-74318f9a09ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_rewriting_system_prompt = \"\"\"\n",
    "\n",
    "You are a document-aligned technical query generator for the Bioelectromagnetics and Cell Biology domain.\n",
    "\n",
    "Task:\n",
    "Your task is to rewrite the user query for better search retrieval. Do not change the core subject or intent.\n",
    "\n",
    "Primary objective:\n",
    "The generated queries must help retrieve the SAME document sections as the original query.\n",
    "Do NOT expand, generalize, or biologically enrich the query beyond the document’s scope.\n",
    "\n",
    "Alignment constraints (MANDATORY):\n",
    "- Preserve the original intent and level of abstraction of the input query.\n",
    "- Do NOT introduce new cell types, organisms, tissues, or biological models not explicitly present or clearly implied in the input.\n",
    "- Do NOT replace electrical stimulation with electromagnetic, magnetic, or clinical modalities unless explicitly stated.\n",
    "- Do NOT shift from background or mechanistic questions to outcome-based, clinical, or experimental-result questions unless explicitly requested.\n",
    "- Do NOT introduce new devices, procedures, or application contexts.\n",
    "\n",
    "Semantic constraints:\n",
    "- Queries must remain conservative and document-grounded.\n",
    "- Prefer general mechanistic or explanatory phrasing over specific experimental outcomes.\n",
    "- Do not speculate beyond concepts present in the input.\n",
    "\n",
    "Terminology constraints:\n",
    "- Do NOT rephrase, expand, or reinterpret technical acronyms.\n",
    "- Use only terminology that is present or directly implied by the input query.\n",
    "\n",
    "Output constraints:\n",
    "- Output ONLY the queries.\n",
    "- Exactly 3 queries.\n",
    "- One query per line.\n",
    "- No numbering, no bullets.\n",
    "- No explanations, notes, headers, or additional text.\n",
    "\n",
    "CRITICAL: You must return ONLY a JSON object. \n",
    "Do not include any text before or after the JSON.\n",
    "Example format:\n",
    "{{\"queries\": [\"query 1\", \"query 2\", \"query 3\"]}}\n",
    "\n",
    "Failure condition:\n",
    "If any text other than exactly three standalone queries is produced, the output is invalid.\n",
    "\n",
    "\n",
    "\n",
    "Input Query: {input_query}\n",
    "\n",
    "\n",
    "\n",
    "STRICT RULE:\n",
    "Return only the three queries, each on its own line.\n",
    "\n",
    "\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a61f45a2-5e0b-4f0b-85d5-7279e4a454fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.output_parsers import PydanticOutputParser\n",
    "from pydantic import BaseModel, Field\n",
    "from typing import List\n",
    "\n",
    "class query_parser(BaseModel):\n",
    "    queries: List[str] = Field(description=\"A list of exactly 3 rewritten technical queries.\")\n",
    "\n",
    "QueryParser = PydanticOutputParser(pydantic_object = query_parser)\n",
    "query_prompt = ChatPromptTemplate.from_messages(\n",
    "    [\n",
    "        (\"system\", query_rewriting_system_prompt),\n",
    "    ]\n",
    ").partial(format_instructions=QueryParser.get_format_instructions())\n",
    "\n",
    "\n",
    "query_chain = query_prompt | model | QueryParser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "7d89ff4f-5adb-4c99-955e-aad1651a9929",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "queries=['How was the electric field strength between the electrodes determined?', 'The calculation method used for electric field strength between electrodes', 'What validation methods were employed for electrode spacing?']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "reponse = query_chain.invoke({\n",
    "    \"input_query\": \"How was the electric field strength between the electrodes calculated and validated in this study?\"\n",
    "})\n",
    "\n",
    "print(reponse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a740cb70-4b36-4fac-b90b-6018ed44d70d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "queries=['Why does electrical stimulation influence bone regeneration?', 'What role does electrical stimulation play in promoting osteoblast activity?', 'How can electrical stimulation impact osteogenesis?']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "reponse = query_chain.invoke({\n",
    "    \"input_query\": \"Why is electrical stimulation considered relevant for bone regeneration and osteoblast activity?\"\n",
    "})\n",
    "\n",
    "print(reponse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "8bfc0534-46dd-4fd1-9bdc-2b0e5bd1f9fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['What physicochemical properties of the culture medium were measured post-AC stimulation?',\n",
       " 'The effects of AC stimulation on which culture medium properties have been studied?',\n",
       " \"What alterations occurred in the culture medium's physical/chemical characteristics following AC exposure?\"]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"Which physicochemical properties of the culture medium were analyzed after AC stimulation, and what changes were observed?\"\n",
    "reponse = query_chain.invoke({\n",
    "    \"input_query\":query\n",
    "})\n",
    "reponse.queries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5c68193-f6bb-4f63-ad1f-4ef7d4fc5357",
   "metadata": {},
   "source": [
    "### Query rewritting is successfully done. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94dda46d-9469-4088-81b9-8424f99b7293",
   "metadata": {},
   "source": [
    "## Step-2 Retrieve"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "345c3c51-bf22-4f90-bc05-55bc792d09be",
   "metadata": {},
   "source": [
    "* Parallel Search: Run each query against your vector database independently.\n",
    "\n",
    "* Aggregation: Collect all the document chunks returned by the three searches.\n",
    "\n",
    "* Deduplication: Remove duplicate chunks (since different queries might find the same paragraph)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "d3a4bdb2-2a59-44e2-9c95-d296c874b838",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['e986d6bb-50eb-40e6-b096-f643bd1b4f6c', 'bcdc5b75-85d5-4543-9bab-7813f52b022d', '7c7131c0-da93-4fc1-94e8-cd50f4417922'])\n",
      "3. Results\n",
      "4. Discussion\n",
      "2. Materials and Methods\n"
     ]
    }
   ],
   "source": [
    "## Parallel serach and Aggregation\n",
    "\n",
    "re_chunks= []\n",
    "for query in reponse.queries:\n",
    "    retrieved_docs = retriever.invoke(query)\n",
    "    re_chunks.extend(retrieved_docs)\n",
    "\n",
    "## Deduplication\n",
    "d=dict()\n",
    "for cnks in re_chunks:\n",
    "    docs_id = cnks.metadata.get('document_id')\n",
    "    d[docs_id] = cnks\n",
    "    \n",
    "print(d.keys())\n",
    "for k in d.keys():\n",
    "    doc = d.get(k).metadata.get('section_header')\n",
    "    print(doc)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45a33166-b231-4220-82d2-0966aee46cd7",
   "metadata": {},
   "source": [
    "* Here, All those queries have retrieved 5 parent documents for one user query\n",
    "* Each parent document has almost 2000 token length.\n",
    "* Model's contrxt window is limited and all these documnets combines with approximately 10000 tokens.\n",
    "* The issue, stuffing 10,000 tokens of dense technical data into a single prompt, can lead to \"Lost in the Middle\" syndrome.\n",
    "* The model ignores the middle documents and only focuses on the beginning and end."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c81f4182-6d19-442a-b0ce-c29018f973a4",
   "metadata": {},
   "source": [
    "## Strategy-3: Map-Reduce\n",
    "\n",
    "### Mapping\n",
    "* First, we take the all document and generate it's summaries too 400 tokens so all document's summary will be 2000 tokens.\n",
    "* All the summary will relate to user query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "24b393ba-7ab1-4093-95ad-f65d4ee518a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "\n",
    "map_sys_msg = \"\"\"\n",
    "You are a Scientific Data Extraction Engine.\n",
    "\n",
    "TASK:\n",
    "Analyze the provided document text and determine whether it contains information relevant to the user query.\n",
    "If relevant, extract and summarize the experimental findings in a single, detailed, informative, well-formed paragraph.\n",
    "\n",
    "STRICT OUTPUT RULES (MANDATORY):\n",
    "1. Output ONLY a single valid JSON object.\n",
    "2. Output MUST start with '{{' and end with '}}'.\n",
    "3. The JSON object must contain exactly two fields: \"relevant\" and \"summary\".\n",
    "4. The \"summary\" field MUST be a SINGLE STRING containing ONE coherent paragraph.\n",
    "5. The summary MUST describe all findings, observations, or understanding, and MUST NOT contain objects, dictionaries, lists, or key–value formatting inside it.\n",
    "6. The summary should be up to 400–500 words.\n",
    "7. Do NOT include explanations, notes, markdown, code, or any extra text outside the JSON.\n",
    "8. If no relevant information is found, return EXACTLY:\n",
    "   {{\"relevant\": false, \"summary\": \"No relevant data found.\"}}\n",
    "\n",
    "SCHEMA:\n",
    "{{\n",
    "  \"relevant\": boolean,\n",
    "  \"summary\": string\n",
    "}}\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a66eba00-d780-4f27-a81e-a39b54688e51",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b2d3542cff9c40febcc67c5fc3d6c870",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Device set to use cuda:0\n"
     ]
    }
   ],
   "source": [
    "from langchain_core.output_parsers import StrOutputParser\n",
    "from langchain_huggingface.llms import HuggingFacePipeline\n",
    "from transformers import AutoTokenizer, pipeline, AutoModelForCausalLM\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "# defining environment variable\n",
    "load_dotenv()\n",
    "\n",
    "model_name = \"meta-llama/Llama-3.1-8B-Instruct\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "#model = AutoModelForCausalLM.from_pretrained(model_name)\n",
    "pipe = pipeline(\"text-generation\",\n",
    "                tokenizer=tokenizer,\n",
    "                   model= model_name,\n",
    "                device_map = \"auto\",\n",
    "                dtype = \"auto\",\n",
    "                max_new_tokens=512,   \n",
    "                do_sample=False,\n",
    "                temperature = 0.4,\n",
    "                repetition_penalty=1.2,\n",
    "                return_full_text= True\n",
    "               )\n",
    "\n",
    "model_pipe = HuggingFacePipeline(pipeline=pipe)\n",
    "map_model = model_pipe.bind(skip_prompt = True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "dda3252f-5134-43c7-aba8-8d859fffac1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.runnables import RunnableLambda\n",
    "import json\n",
    "\n",
    "def normalize_llm_output(raw: str) -> str:\n",
    "    \"\"\"\n",
    "    Extracts the FIRST valid JSON object from LLM output\n",
    "    and returns it as a JSON STRING (LangChain-safe).\n",
    "    \"\"\"\n",
    "\n",
    "    if not raw or not isinstance(raw, str):\n",
    "        raise ValueError(\"LLM output is empty or not a string\")\n",
    "\n",
    "    text = raw.strip()\n",
    "    text = text.replace(\"```json\", \"\").replace(\"```\", \"\")\n",
    "\n",
    "    start = None\n",
    "    brace_count = 0\n",
    "\n",
    "    for i, ch in enumerate(text):\n",
    "        if ch == \"{\":\n",
    "            if start is None:\n",
    "                start = i\n",
    "            brace_count += 1\n",
    "        elif ch == \"}\":\n",
    "            brace_count -= 1\n",
    "            if brace_count == 0 and start is not None:\n",
    "                json_text = text[start:i+1]\n",
    "                break\n",
    "    else:\n",
    "        raise ValueError(\"No JSON object found in LLM output\")\n",
    "\n",
    "    try:\n",
    "        data = json.loads(json_text)\n",
    "    except json.JSONDecodeError as e:\n",
    "        raise ValueError(f\"JSON parsing failed: {e}\\nExtracted:\\n{json_text}\")\n",
    "\n",
    "    # Schema hardening\n",
    "    data.setdefault(\"relevant\", False)\n",
    "    data.setdefault(\"summary\", \"\")\n",
    "\n",
    "    if isinstance(data[\"summary\"], dict):\n",
    "        data[\"summary\"] = \" \".join(f\"{k}: {v}\" for k, v in data[\"summary\"].items())\n",
    "\n",
    "    data[\"relevant\"] = bool(data[\"relevant\"])\n",
    "    data[\"summary\"] = str(data[\"summary\"])\n",
    "\n",
    "\n",
    "    return json.dumps(data, ensure_ascii=False)\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "normalizer = RunnableLambda(normalize_llm_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "a9dbf7b3-14a1-4f3c-8d73-9f90b2b122ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydantic import BaseModel, Field\n",
    "\n",
    "class Summariser(BaseModel):\n",
    "    relevant: bool = Field(description=\"Whether the text contains info about the physicochemical properties or biological effects of the stimulation.\")\n",
    "    summary: str = Field(\n",
    "        description=\"A single-paragraph summary of 500-700 words containing only the information from the text that directly addresses the user query.\"\n",
    "    )\n",
    "\n",
    "\n",
    "# This parser creates the format instructions that we inject into the prompt\n",
    "map_parser = PydanticOutputParser(pydantic_object=Summariser)\n",
    "\n",
    "\n",
    "map_prompt = ChatPromptTemplate.from_messages(\n",
    "    [\n",
    "        (\"system\", map_sys_msg),\n",
    "        (\"user\", \"Query: {user_query}\\nText: {text_chunk}\")\n",
    "    ]\n",
    ").partial(format_instructions=map_parser.get_format_instructions())\n",
    "map_chain = map_prompt | map_model | normalizer | map_parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "1adba4ee-4408-4335-9378-2a158179ae0e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Document(metadata={'source': 'Documents/cells-11-02650-v2.pdf', 'detection_class_prob': 0.7357485890388489, 'is_extracted': 'true', 'last_modified': '2025-12-02T10:31:49', 'filetype': 'application/pdf', 'page_number': 13, 'file_directory': 'Documents', 'filename': 'cells-11-02650-v2.pdf', 'category': 'Title', 'element_id': '55d05466ab31cd967316cbb1702b4345', 'section_header': '3. Results', 'start_page': 8, 'end_page': 13, 'paper_title': 'Pulsed Electrical Stimulation Affects Osteoblast Adhesion and Calcium Ion Signaling', 'document_id': 'e986d6bb-50eb-40e6-b096-f643bd1b4f6c', 'start_index': 0}, page_content='3.1. Characterization of AC Electrical Stimulation (ES) Parameter\\n\\nThe electric ﬁeld strength and its magnitude between the electrodes can be seen in Figure 3. Between the electrodes, where the cells are placed, is a large region with a more or less homogeneous ﬁeld. Only close to the electrode edges a ﬁeld enhancement can be observed (Figure 3a). However, the area with an enhanced ﬁeld is small in comparison to the area with a more or less constant ﬁeld strength. Along the centerline of the well, the ﬁeld decreases from 91.7 V/m at the electrodes to 90.2 V/m precisely at the midpoint between the electrodes (Figure 3b). Under the assumption of inﬁnitely wide capacitor plates, the ﬁeld could also be analytically estimated. In this case, it would be equal to the voltage difference over the electrode spacing. At a voltage difference of 1 V and a spacing of 1.1 cm, this corresponds to 90.91 V/m. It is of great interest to validate the numerical model. Since neither the electric ﬁeld nor the electric potential in the well can be measured easily, the current through the electrode needs to be compared. The current equals the conductivity times the normal component of the electric ﬁeld integrated over the electrode surface, yielding 1.75 A/m for the 2D geometry. With a 1 mL cell culture medium, a ﬁll level of 3.4 mm is reached. It should be noted that capillary effects are not considered here. Then, the current is 5.95 mA for 1 S/m conductivity and scales linearly with the conductivity. Consequently, it would be 11.9 mA for 2 S/m and so on.\\n\\nFigure 3. Simulation data revealed (a) the electric ﬁeld distribution and strength (red arrows) between the two electrodes at a voltage difference of 1 V. The grey isolines highlight the regions where the ﬁeld reaches 80, 90, and 100 V/m (the latter in the vicinity of the electrodes’ sharp corners), respectively. Furthermore, (b) shows the electric ﬁeld strength along the centerline between the two electrodes. The result indicates that the cells are exerted to a mostly homogeneous ﬁeld with a magnitude of about 90 V/m. Note that for different voltage differences, the ﬁeld strength changes linearly w.r.t. the voltage.\\n\\nThe individual simulated and measured electric parameters are summarized in Table 1. The measured currents are peak values during the pulse. See Chapter 4 for a comparative discussion of these values.\\n\\nTable 1. Individual ES values after measurement and simulation. To each setup, appropriate colors were dedicated (Code).\\n\\n Table Summary: \\n * Theoretical values calculated from the applied voltage and pulse duration. ** Measured values obtained from the experiment.\\n\\nThis table compares theoretical and measured electric current values for different setups involving electrical stimulation. At 1V, a 20 Hz frequency with a 3.6 ms pulse duration resulted in a theoretical current of 90 A and a measured current of 0.012 A. Increasing both voltage to 5V and frequency to 20 Hz maintained the same pulse duration, leading to a theoretical current of 450 A and a measured current of 0.067 A. Similarly, at 5V but with a 7.9 Hz frequency and 10 ms pulse duration, the theoretical and measured currents were 450 A and 0.069 A, respectively. These results suggest that higher voltages and frequencies generally lead to higher currents, though measured values are consistently lower than theoretical predictions. Notably, the control setup lacks specific data.\\n\\n* Own calculation of electric ﬁeld from simulation. ** Own measurement of electric current in DMEM with 10% FCS using Tektronix TDS2014B.\\n\\n3.2. Acellular Medium Characteristics—AC-Stimulated Liquid\\n\\nAnalyzing the effect of AC stimulation on the culture medium is a crucial step in order to exclude changes in liquid properties, such as pH value, as a reason for the resulting cellular adaption/behavior.\\n\\nThe temperature of the AC-stimulated medium was signiﬁcantly changed vs. the Control only for the 5 V setups with 20 Hz and 7.9 Hz with 34.8 ± 0.3 ◦C and 34.6 ± 0.2 ◦C (mean ± s.e.m.), respectively. The mean value of the Control was 33.2 ± 0.3 ◦C (Figure 4a).\\n\\nFigure 4. Properties of AC-stimulated liquids, i.e., the complete DMEM, after 10 min ES: (a) tempera- ture, (b) pH, and (c) oxygen content. Note that only the temperature changed for the 5 V settings. (median ± min/max, n = 5–6, one-way ANOVA posthoc Bonferroni, * p < 0.05).\\n\\nThe pH values in the medium after ES did not change for 1 and 5 V in combination with 20 Hz and 7.9 Hz (e.g., Control pH 7.40 ± 0.03, 5 V_20 Hz pH 7.42 ± 0.02; mean ± s.e.m.) (Figure 4b).\\n\\nThe oxygen content in the medium was not changed after the 10 min AC stimulation (e.g., Control 20.2 ± 0.1%, 5 V_20 Hz 20.1 ± 0.09%; mean ± s.e.m.) (Figure 4c).\\n\\nDuring electrical stimulation, faradic byproducts such as H2O2 can be generated. H2O2 in the media can inﬂuence cellular fate and may contribute to ES-induced effects. In our experiments, the concentration of H2O2 was only increased in media stimulated with 5 V_7.9 Hz compared to the Control (Figure 5). The mean concentration increased from 1.48 µM ± 0.02 in Control samples to 2.52 µM ± 0.32 in 5 V_7.9 Hz samples directly after 10 min ES. The H2O2 concentration declined over time in all samples, with 5 V_ 7.9 Hz showing elevated levels also after 0.5 h (mean 2.05 ± 0.28 µM). After 4 h and 24 h, the concentration was similar in all samples and leveled off at about 0.38 µM (Figure 5a,b). The overall decline was independent of the O2 concentration in the media (data not shown).\\n\\nFigure 5. H2O2 content of AC-stimulated liquids, i.e., the DMEM w/o pyruvate, after 10 min ES, (a) for all AC-stimulation parameters as BoxPlot graph, (b) time course of H2O2 concentration. Note that 10 min AC with 1 V settings did not change the H2O2 concentration in contrast to 5 V_7.9 Hz with a signiﬁcant increase in H2O2 up to 0.5 h. (Median ± min/max, n = 5–6, two-way ANOVA posthoc Bonferroni, * p < 0.05 compared to all).\\n\\n3.3. AC-Stimulated Liquids and Long-Time Cell Adhesion\\n\\nIn order to analyze whether AC stimulation of liquids alone (rather than the cells) affected cell adhesion; we recorded the impedance with xCELLigence RTCA over 24 h. The adhering cells interfere with the electrode structures on the chip, thus impeding the electron response in the medium. The impedance depends on the number of cells, the cell size, and the strength of adhering cells. As shown in Figure 6, cells were not inﬂuenced by the AC- stimulated liquid, and the cell adhesion was approximately the same as the unstimulated Control. Solely a slight increase was detectable with 1 V_20 Hz AC-stimulated liquids\\n\\n(Figure 6a). After 24 h, the max cell index (i.e., the 24 h endpoint value) in MG-63s was comparable in all liquids indicating no changes of the liquid DMEM due to AC stimulation (Control: 2.5 ± 0.1, 1 V_20 Hz: 2.7 ± 0.1, 1 V_7.9 Hz: 2.4 ± 0.1, 5 V_20 Hz: 2.5 ± 0.1, 5 V_7.9 Hz: 2.4 ± 0.1; mean ± s.e.m.; Figure 6b). In addition, the adhesion rate (time point when 50% of the cells adhere) was the same in all samples and showed no inﬂuence of the AC-stimulated liquid (Control, 1 V_20 Hz, 1 V_7.9 Hz, 5 V_7.9 Hz: all 0.5; 5 V_20 Hz: 0.4; mean; Figure 6c).')"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for x in list(d.keys())[:1]:\n",
    "    do = d.get(x)\n",
    "do"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "0c583c92-3abd-472c-9ca8-4ba35ef19fdd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    }
   ],
   "source": [
    "query = \"How was the electric field strength between the electrodes calculated and validated in this study\"\n",
    "map_response = map_chain.invoke({\"user_query\":query,\n",
    "                        \"text_chunk\": do})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "31ff5872-417b-4fcd-a246-8fc656516477",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "To calculate the electric field strength between the electrodes, simulations were performed assuming infinitely wide capacitor plates, which yielded a field strength of 90.91 V/m under a voltage difference of 1 V and a spacing of 1.1 cm. This analytical estimation matched closely with the numerically modeled field strength of around 90 V/m along the centerline between the electrodes. Validation of the numerical model involved comparing the simulated electric field with the measured current through the electrode, which equaled the conductivity times the normal component of the electric field integrated over the electrode surface. The results indicated that the cells experienced a mostly homogeneous electric field with a magnitude of about 90 V/m. Additionally, measurements confirmed that increasing the voltage and frequency led to higher currents, although the measured values were consistently lower than the theoretically predicted ones. Further analysis focused on characterizing the acellular medium characteristics following AC stimulation, revealing significant changes in temperature for certain setups, while pH and oxygen content remained unaffected. Moreover, hydrogen peroxide concentrations were significantly increased in some cases, particularly after 10 minutes of stimulation with 5 V and 7.9 Hz. Lastly, long-term cell adhesion studies demonstrated that AC-stimulated liquids had minimal impact on cell attachment, with slight increases detected in some scenarios.\n"
     ]
    }
   ],
   "source": [
    "print(map_response.summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "a6cf9a44-92be-47a2-a57c-3ec112dfebdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "256"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_length(map_response.summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29fc9480-7702-48b9-80f8-8639a60976b9",
   "metadata": {},
   "source": [
    "### Writing functions to automate the mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "155948f4-11da-4935-b733-f6a7a712f286",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    }
   ],
   "source": [
    "# function for query rewriting\n",
    "query = \"How was the electric field strength between the electrodes calculated and validated in this study?\"\n",
    "def rewrite_query(user_query:str) -> List[str]:\n",
    "\n",
    "    # Defining chain\n",
    "    query_chain = query_prompt | model | QueryParser\n",
    "\n",
    "    # invoking the chain\n",
    "    queries = query_chain.invoke({\"input_query\": user_query})\n",
    "    return queries\n",
    "queries = rewrite_query(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "3530326e-6bfd-48e9-b522-210164d37aa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for collecting rlevent document based on re written queries\n",
    "from langchain_core.documents import Document\n",
    "def relevent_retrieved_docs(list_of_queries: List[str]) -> List[Document]:\n",
    "    retrieved_chunks= []\n",
    "    relevent_documents=dict()\n",
    "    for query in list_of_queries.queries:\n",
    "        relevent_doc = retriever.invoke(query)\n",
    "        retrieved_chunks.extend(relevent_doc)\n",
    "\n",
    "    for chunk in retrieved_chunks:\n",
    "        doc_id = chunk.metadata.get('document_id')\n",
    "        relevent_documents[doc_id] = chunk\n",
    "\n",
    "    return relevent_documents\n",
    "rel_doc = relevent_retrieved_docs(queries)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "45f88f65-6bdf-4240-86fd-eb4ee20ba886",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(rel_doc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "48485971-aa4b-4c97-9057-09bfdea92278",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n",
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n",
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    }
   ],
   "source": [
    "def combined_summary(user_query: str, list_of_relevent_docs: dict):\n",
    "    \n",
    "    map_chain = map_prompt | map_model | normalizer | map_parser\n",
    "    combined_summaries = []\n",
    "    for doc in list_of_relevent_docs:\n",
    "        doc_obj = list_of_relevent_docs[doc]\n",
    "        response = map_chain.invoke({\n",
    "            \"user_query\": user_query, \n",
    "            \"text_chunk\": doc_obj\n",
    "        })\n",
    "        \n",
    "        # Accessing the content safely (handling different model output types)\n",
    "        summary_text = response.summary if hasattr(response, 'content') else str(response.summary)\n",
    "        \n",
    "        # Simple filter to prevent hallucination carry-over\n",
    "        if response.relevant:\n",
    "            combined_summaries.append(summary_text)\n",
    "\n",
    "    # Join with clear delimiters so the final reduce step knows where chunks end\n",
    "    return \"\\n\\n\".join(combined_summaries)\n",
    "\n",
    "summary = combined_summary(user_query=query, list_of_relevent_docs=rel_doc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "341cbf38-c3b4-459a-9dac-e3acb0513237",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open(\"pkl_files/summary.pkl\",'wb') as f:\n",
    "    pickle.dump(summary, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "061efa2d-b539-4f00-9e4c-59df9315f11e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"To calculate the electric field strength between the electrodes, simulations were performed assuming infinitely wide capacitor plates, which yielded a field strength of 90.91 V/m under a voltage difference of 1 V and a spacing of 1.1 cm. This analytical estimation matched closely with the numerically modeled field strength of around 90 V/m along the centerline between the electrodes. Validation of the numerical model involved comparing the simulated electric field with the current flowing through the electrodes, which equaled the conductivity times the normal component of the electric field integrated over the electrode surface, giving 1.75 A/m for the 2D geometry. Experimental measurements confirmed that the electric field decreased from 91.7 V/m at the electrodes to 90.2 V/m at the midpoint between them. Higher voltages led to stronger electric fields, while increasing frequencies caused larger currents. Additionally, the analysis of acellular medium characteristics revealed significant changes in temperature for certain setups, particularly those involving 5 V and 20 Hz or 7.9 Hz, whereas pH and oxygen content remained unaffected. Moreover, hydrogen peroxide concentrations were significantly increased in some cases, especially following 10 minutes of AC stimulation with 5 V and 7.9 Hz. Lastly, long-term cell adhesion studies demonstrated that AC-stimulated liquids had little impact on cell behavior, with minimal variations detected across various conditions.\\n\\nThe electric field strength between the electrodes was calculated and validated in this study by choosing appropriate ES parameters from existing literature, specifically focusing on intensity, frequency, and duration of electric field. An optimal voltage window for maximizing osteoblast density was identified as 1 V to 15 V and 20 Hz, and electric field amplitudes ranging from 10–1000 V/m were deemed sufficient to achieve an effect without causing cell damage. Cell adhesion was significantly increased by 10-minute ES, regardless of AC-stimulation parameters, and enhanced cell attachment to surfaces due to ES has previously been observed in several studies. Additionally, ES influenced the arrangement and density of adhesion receptors-integrins-and focal adhesion-like vinculin. Increased basal Ca2+-activity was detected for settling cells under ES, although its effect on Ca2+-mobilization depended on frequencies and voltages. Intracellular ROS levels rose initially following ES, but returned to baseline values after four hours, indicating potential signaling pathways activated by ES. Overall, these findings suggest that ES may play a significant role in modulating cell behavior, particularly in terms of adhesion, redox status, and Ca2+-signaling, warranting further investigation into optimizing ES parameters for therapeutic applications.\\n\\nThe electric field strength between the electrodes was calculated using numerical simulations based on the electrostatic Laplace equation. Assuming non-frequency dependent dielectric parameters of the cell culture medium, the finite element method was employed to compute the electric field at a voltage difference of 1 V. The results showed that the electric field strength varies across the well, with higher strengths near the edges and lower strengths towards the center. Additionally, the calculation revealed that the electric field distribution remains homogeneous throughout the experiment, indicating minimal influence of the electrode's shape and size on the overall field. Furthermore, the authors utilized a 2D model representing a slice through the geometry at a certain height and the top view, ensuring accurate computation of the electric field within the confines of the well. The calculations also demonstrated that scaling the voltage difference would result in proportional changes to the electric field strength, allowing for easy adaptation to various experimental settings. Overall, these computational models provide valuable insights into the spatial distribution of the electric field generated during the experiments, facilitating informed decision-making regarding future studies and applications involving pulsed electrical stimulation.\""
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pickle\n",
    "with open(\"pkl_files/summary.pkl\",'rb') as f:\n",
    "    summary = pickle.load(f)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "75cf73cd-3a0d-407c-b807-b6b832aa40ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "reduce_system_prompt = \"\"\"\n",
    "<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n",
    "\n",
    "You are a Senior Research Data Auditor. Your task is to synthesize provided summaries into a structured technical narrative while maintaining absolute factual integrity.\n",
    "\n",
    "### NARRATIVE REQUIREMENTS (The \"analysis\" field):\n",
    "1. **Factual Anchor**: Use only the information explicitly stated in the source text. \n",
    "2. **Qualitative vs. Quantitative**: If the source provides specific numbers or units, include and **bold** them. If the source uses descriptive language (e.g., \"enhanced,\" \"decreased,\" \"significant\"), you MUST mirror that language. Never convert qualitative descriptions into numerical estimates.\n",
    "3. **Causal Logic**: Explain the relationship between the inputs/conditions and the resulting outcomes based solely on the provided evidence.\n",
    "4. **Seamless Style**: Write in professional paragraphs without headers. Length should be proportional to the actual data provided.\n",
    "\n",
    "### DATA EXTRACTION RULES (The \"key_metrics\" field):\n",
    "- **STRICT FACTUALITY**: Only extract values that appear verbatim in the source.\n",
    "- **NO INFERENCE**: Do not calculate, estimate, or \"fill in\" missing values (e.g., do not invent a +/- range if none is given).\n",
    "- **Empty State**: If no explicit numerical measurements or specific named parameters are found, return \"key_metrics\": [].\n",
    "\n",
    "### FORMATTING:\n",
    "- Output MUST be a valid JSON object.\n",
    "- The \"analysis\" field must be a single string. Use \\\\n for paragraph breaks.\n",
    "\n",
    "<|eot_id|><|start_header_id|>user<|end_header_id|>\n",
    "Query: {user_query}\n",
    "Source Data:\n",
    "{map_summaries}\n",
    "\n",
    "### RESPONSE (JSON ONLY):\n",
    "<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "9abc4edf-b29d-482c-be63-2335a9a40b1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cde0d352ab3e4348bb91affd67f4a6ce",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Device set to use cuda:0\n"
     ]
    }
   ],
   "source": [
    "from langchain_core.output_parsers import StrOutputParser\n",
    "from langchain_huggingface.llms import HuggingFacePipeline\n",
    "from transformers import AutoTokenizer, pipeline, AutoModelForCausalLM\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "# defining environment variable\n",
    "load_dotenv()\n",
    "\n",
    "model_name = \"meta-llama/Llama-3.1-8B-Instruct\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "#model = AutoModelForCausalLM.from_pretrained(model_name)\n",
    "pipe = pipeline(\"text-generation\",\n",
    "                tokenizer=tokenizer,\n",
    "                model= model_name,\n",
    "                device_map = \"auto\",\n",
    "                dtype = \"auto\",\n",
    "                max_new_tokens=1024,   \n",
    "                do_sample=False,\n",
    "                temperature = 0.5,\n",
    "                repetition_penalty=1.2,\n",
    "                \n",
    "               )\n",
    "\n",
    "model_pipe = HuggingFacePipeline(pipeline=pipe)\n",
    "reduce_model = model_pipe.bind(skip_prompt = True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "5e3e14c3-966f-46ba-87de-389daf451096",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.output_parsers import PydanticOutputParser\n",
    "from pydantic import BaseModel, Field\n",
    "from typing import List, Optional\n",
    "\n",
    "class Metric(BaseModel):\n",
    "    parameter: Optional[str] = Field(\n",
    "        default=None,\n",
    "        description=\"What was measured, if explicitly stated\"\n",
    "    )\n",
    "    value: Optional[str] = Field(\n",
    "        default=None,\n",
    "        description=\"Measured value with unit, if available\"\n",
    "    )\n",
    "\n",
    "class ResearchSynthesisParser(BaseModel):\n",
    "    is_sufficient: Optional[bool] = Field(\n",
    "        default=None,\n",
    "        description=\"True if data was found.\"\n",
    "    )\n",
    "\n",
    "    key_metrics: Optional[List[Metric]] = Field(\n",
    "        default=None,\n",
    "        description=\"Optional list of extracted measurements.\"\n",
    "    )\n",
    "\n",
    "    analysis: str = Field(\n",
    "        description=\"The seamless narrative report (no headers) as multiple paragraphs.\"\n",
    "    )\n",
    "\n",
    "ReduceParser = PydanticOutputParser(pydantic_object = ResearchSynthesisParser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "6b032394-277e-45ef-89f5-8eda9efd498c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from langchain_core.runnables import RunnableLambda\n",
    "\n",
    "def normalize_research_output(raw_output) -> str:\n",
    "    \"\"\"\n",
    "    Normalizes raw LLM outputs for CellSense:\n",
    "    - Removes problematic backslashes (\\)\n",
    "    - Ensures valid JSON\n",
    "    - Enforces schema: analysis (str), key_metrics (list[dict])\n",
    "    \"\"\"\n",
    "\n",
    "    # 1️⃣ If already a dict, just use it\n",
    "    if isinstance(raw_output, dict):\n",
    "        data = raw_output\n",
    "\n",
    "    # 2️⃣ If it's a string, clean it\n",
    "    elif isinstance(raw_output, str):\n",
    "        raw_output = raw_output.strip()\n",
    "\n",
    "        # Remove leading/trailing quotes\n",
    "        if (raw_output.startswith(\"'\") and raw_output.endswith(\"'\")) or \\\n",
    "           (raw_output.startswith('\"') and raw_output.endswith('\"')):\n",
    "            raw_output = raw_output[1:-1]\n",
    "\n",
    "        #  Remove all backslashes (this fixes \\_ and any others)\n",
    "        raw_output = raw_output.replace(\"\\\\\", \"\")\n",
    "\n",
    "        try:\n",
    "            data = json.loads(raw_output)\n",
    "        except json.JSONDecodeError:\n",
    "            return json.dumps({\n",
    "                \"is_sufficient\": False,\n",
    "                \"key_metrics\": [],\n",
    "                \"analysis\": \"JSON parsing failed\"\n",
    "            })\n",
    "\n",
    "    else:\n",
    "        return json.dumps({\n",
    "            \"is_sufficient\": False,\n",
    "            \"key_metrics\": [],\n",
    "            \"analysis\": \"Unsupported output type\"\n",
    "        })\n",
    "\n",
    "    # 3️⃣ Schema hardening\n",
    "    if not isinstance(data.get(\"analysis\"), str):\n",
    "        data[\"analysis\"] = str(data.get(\"analysis\", \"\"))\n",
    "\n",
    "    if not isinstance(data.get(\"key_metrics\"), list):\n",
    "        data[\"key_metrics\"] = []\n",
    "\n",
    "    # Ensure each metric is a dict\n",
    "    data[\"key_metrics\"] = [m for m in data[\"key_metrics\"] if isinstance(m, dict)]\n",
    "\n",
    "    return json.dumps(data, ensure_ascii=False)\n",
    "\n",
    "\n",
    "research_normalizer = RunnableLambda(normalize_research_output)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "38d462fa-7642-4495-a54d-37260d3a45ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "\n",
    "reduce_template = ChatPromptTemplate(\n",
    "    [\n",
    "        (\"system\",  reduce_system_prompt)\n",
    "    ]\n",
    ").partial(format_instructions=ReduceParser.get_format_instructions())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "26435099-33c1-4028-9036-37776351aad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "reduce_chain = reduce_template | reduce_model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "9765bb2d-8eb8-43bf-a8e7-5a5ed037c49f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    }
   ],
   "source": [
    "#query = \"How did pulsed electrical stimulation affect the initial adhesion of suspended osteoblasts?\"\n",
    "query = \"How was the electric field strength between the electrodes calculated and validated in this study?\"\n",
    "Final_response = reduce_chain.invoke({\n",
    "    \"user_query\": query,\n",
    "    \"map_summaries\": summary\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "5ff65ae4-fa49-4ea4-b877-6aedbad7ad24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'{\\n    \"analysis\": \"\\\\nThe electric field strength between the electrodes was first estimated analytically, yielding a value of **90.91 V/m**, considering infinite plate capacitors under a **1 V** voltage difference and a distance of **1.1 cm**.\\\\nThis analytical approximation aligned closely with the computed electric field strength via numerical modeling, approximately **90 V/m**, measured along the central axis connecting the electrodes.\\\\nValidation of the simulation entailed matching it against the current passing through the electrodes, amounting to **1.75 A/m** for a two-dimensional setup.\\\\nExperimental observations indicated that the electric field diminished gradually from **91.7 V/m** at the electrodes down to **90.2 V/m** midway between them.\\\\nHigher applied voltages resulted in more robust electric fields, while elevated frequencies produced greater currents.\\\\nMoreover, detailed examination of the properties of the test environment disclosed notable alterations in temperature under particular configurations, such as when applying **5 V** and either **20 Hz** or **7.9 Hz**; however, both pH and oxygen concentration remained unchanged.\\\\nAdditionally, there was a marked increase in hydrogen peroxide levels in select instances, notably following ten minutes of alternating current stimulation set at **5 V** and **7.9 Hz**.\\\\nLong-term investigations concerning cellular adherence suggested that exposure to AC-stimulated substances did not have a substantial impact on cell behavior, displaying minor fluctuations across different scenarios.\",\\n    \\n    \"key_metrics\": {\\n        \"voltage_difference\": [\"1 V\"],\\n        \"spacing\": [\"1.1 cm\", \"**1.1 cm**\"],\\n        \"field_strength_analytical\": [**90.91 V/m**],\\n        \"field_strength_numerical\": [\"around **90 V/m**\"],\\n        \"current_through_electrodes\": [**1.75 A/m**]\\n    }\\n}'"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Final_response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "01445dcb-9dbe-42b1-a06b-63912a3b6a41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to initialize NVML: Unknown Error\n"
     ]
    }
   ],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c3c5483-ed8f-4111-bba2-284f7170fd94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48da24c2-5ec6-4849-bca1-03650344bb32",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49a34f04-7c71-4b54-bb2b-e727e437eb1e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "105e83ca-ff52-4400-b0cb-32bb3c5394a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aac8451-2e94-411b-9b72-74d4ea021f44",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22735842-f38e-430f-b7a4-0edda1375db9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f91170c-02e9-4ed9-b1e3-fd5ff7d55376",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9ab20e8f",
   "metadata": {},
   "source": [
    "### Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "a09ae3f4-f612-4bb1-a0f8-2dcf346a1f86",
   "metadata": {},
   "outputs": [],
   "source": [
    "system_msg = \"\"\"\n",
    "<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n",
    "\n",
    "You are an Expert Scientific Literature Analyst. Your goal is to provide concise, factual, evidence-based answers.\n",
    "\n",
    "--- INSTRUCTIONS ---\n",
    "1. STRICT: Answer the user's question ONLY using the text in the CONTEXT section below.\n",
    "2. EVIDENCE: If the answer is found, include the source document by using its 'source' metadata (e.g., [Source: document_name.pdf, page 5]).\n",
    "3. UNCERTAINTY: If the answer cannot be found in CONTEXT, respond: 'I apologize, the necessary information could not be found in the available document sections.'\n",
    "4. FORMAT: Use clear, structured output (bullet points, numbered lists, or short paragraphs).\n",
    "\n",
    "--- CONTEXT ---\n",
    "{context}\n",
    "\n",
    "<|eot_id|><|start_header_id|>user<|end_header_id|>\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "4958fb68",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "from langchain_core.output_parsers import StrOutputParser\n",
    "\n",
    "\n",
    "parser = StrOutputParser()\n",
    "template =  ChatPromptTemplate.from_messages(\n",
    "    [\n",
    "        (\n",
    "            \"system\", system_msg\n",
    "        ),\n",
    "        (\"user\", \"{question}\"),\n",
    "    ]\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e153d6d8-de80-4c5e-99bc-071b15db67f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading checkpoint shards: 100%|██████████| 4/4 [00:01<00:00,  2.54it/s]\n",
      "Loading checkpoint shards: 100%|██████████| 4/4 [00:03<00:00,  1.22it/s]\n",
      "Device set to use cuda:0\n"
     ]
    }
   ],
   "source": [
    "from langchain_huggingface.llms import HuggingFacePipeline\n",
    "from transformers import AutoTokenizer, pipeline, AutoModelForCausalLM\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "# defining environment variable\n",
    "load_dotenv()\n",
    "\n",
    "model_name = \"meta-llama/Llama-3.1-8B-Instruct\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "model = AutoModelForCausalLM.from_pretrained(model_name)\n",
    "pipe = pipeline(\"text-generation\",\n",
    "                tokenizer=tokenizer,\n",
    "               model= model_name,\n",
    "                device_map = \"auto\",\n",
    "                dtype = \"auto\",\n",
    "                max_new_tokens=700,   \n",
    "                do_sample=False,\n",
    "                temperature = 0.2,\n",
    "                repetition_penalty=1.5\n",
    "               )\n",
    "\n",
    "model_pipe = HuggingFacePipeline(pipeline=pipe)\n",
    "model = model_pipe.bind(skip_prompt = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "1d9c158b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.runnables import RunnableParallel, RunnablePassthrough\n",
    "\n",
    "parallel_chain = (\n",
    "    {\n",
    "    \"context\": retriever,\n",
    "    \"question\": RunnablePassthrough()\n",
    "    }   \n",
    ")\n",
    "\n",
    "chain = parallel_chain | template | model | parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8850bf3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "\n",
      "Answer:\n",
      "According to Section \"Characteristics\", subsection “Measurement” isn’t present however looking into another relevant context - Subection ‘‘Electric field’’ measurements weren't mentioned explicitly related to how they measure electricity flow. But rather what’s interesting lies elsewhere :  \n",
      "\n",
      "\n",
      "\n",
      "Section ''Results'' > sub-section ``Accllar Media Charactristics — Ac Stimulatoin ''. Here you find out details regarding measuring eletrical Current.\n",
      "\n",
      "\n",
      "\n",
      "Here is your detailed explanation:\n",
      "\n",
      "To determine the validity of simulations used throughout research paper,\n",
      "they needed accurate measures of Electric Field Strength\n",
      "and compare them against those predicted theoretically.\n",
      "\n",
      "\n",
      "They utilized `Tekronik DTS210` oscilloscope device specifically designed for high-speed digital storage scopes capable capturing fast transient signals accurately recording waveforms quickly making precise timing easy-to-read displays allowing users access real-world insights instantly saving valuable lab space since every detail counts especially complex phenomena occurring rapidly before disappearing forever leaving behind nothing tangible except perhaps faint memories fading away slowly until completely erased eventually becoming mere whispers echoes haunting past events now lost beyond recall... \n",
      "\n",
      "\n",
      "However upon closer inspection reveals actual methods employed aren\\'t quite straightforward:\n",
      "\n",
      "\n",
      "Current wasn&#x27;t simply captured passively instead actively induced artificially created controlled environments mimicking natural scenarios closely replicating realistic situations encountered daily life outside laboratory walls yet still maintaining utmost precision required scientific inquiry ensuring reliability accuracy dependability trustworthiness henceforth referred hereinbeforeafterwards always keeping things simple uncomplicated transparent honest reliable trustworthy dependable consistent coherent logical sound sensible reasonable fair justifiable understandable comprehensible accessible approachable friendly welcoming open-minded receptive responsive empathetic considerate benevolent kind generous charitable magnanimous altruistic selfless noble virtuous excellent outstanding remarkable extraordinary exceptional superlativesuperior quality top-notch premium grade first-rate supreme unrivaled unparalleled unbeatable incomparable matchlesssupreme highest caliber finest most superior better-than-best ultimate greatest paramount transcendent sublime magnificent glorious resplendant radiant dazzling brilliant luminariesunparalleled excellence paragon exemplar paradigm epitome embodiment quintessencemost perfect ideal flawless exemplary consummate peerlessness perfection pinnacle apex zenith summit crowning glory apotheosis culmination climax apogee height grandeur majesty splendor brilliance radiance efflorescence exuberant luxuriant lush verdurous opulent sumptious rich profuse abundant copius plentiful lavish munificent liberal bountiful affluent prosperous thriving flourishing blooming burgeoning blossoming sprouting germinating budding growing developing maturing ripening harvesting gathering collecting accumulating aggregating consolidating augmenting increasing expanding enlarging amplifying intensify strengthening fortification reinforcement bolster support augmentation supplementation enrichment embellishment ornament decoration beautifullustrations illustrations pictures photographs graphics visualizations diagrams schematics blueprints plans layouts designs drafts sketches renderings models prototypes mockups simulacra replicas facsimiles copies duplicates analogues surrogates substitutes alternatives variants modifications adaptations permutations combinations fusions hybrids derivatives spin-offs sequels continuums extensions expansions elaborations developments improvements refinements upgrades enhancements revisions updates patches fixes corrections amendments alterations adjustments fine-tuning tweaking polishing honing refining smoothing streamlining optimizing calibrating synchronising harmonisation balancing tuning adjusting regulating controlling governing directing overseeing supervisory monitoring inspecting examining scrutinizing reviewing auditing checking verifying validating confirming certificated authenticating endorsing approving sanctioning ratifiying legalised legitimatisng legitimized authoritatively sanctioned officially recognised acknowledged accepted ratified confirmed verified authenticated endorsed approved certified legalized legitimate authorized permitted allowed permissible permissble tolerible acceptable satisfactory pleasing agreeable likley preferable desirable welcome pleasant enjoyable delightful gratifying satisfying\n"
     ]
    }
   ],
   "source": [
    "response = chain.invoke(\"How was the electrical current measured during stimulation, and why was this measurement important?\")\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9039fe83-3828-4b31-8d8e-066817ba1a21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "response = chain.invoke(\"Why did the authors specifically select the frequencies 7.9 Hz and 20 Hz, and how do these frequencies relate to physiological electrical signals in bone?\")\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cd7a0f1e-a5c2-4d61-88ec-f983f8463002",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " --- \n",
      "\n",
      "---\n",
      "\n",
      "## Step 1\n",
      "To determine how the authors differentiate between immediate impacts caused straight away onto biological entities versus secondary influences brought forth indirectly upon those organisms.\n",
      " \n",
      "First off let’s look if there exists anywhere throughout given context where researchers discuss what exactly constitutes \"direct\" vs “indirect” influence.\n",
      "\n",
      "\n",
      "Upon reviewing said passage one finds out - ‘Direct’ refers here specifically towards alterations happening right immediately without need intermediate substances whereas anything else would fall underneath category labeled simply put – Indirection.\n",
      "\n",
      "\n",
      "\n",
      "Now since now know definitions lets proceed analyzing actual experiment setup details mentioned later down lines describing certain procedures carried-out involving application external stimuli followed closely afterwards measuring resultant outcomes produced thus enabling us infer differences existing possibly arising either path taken\n",
      "\n",
      "\n",
      "\n",
      "Given description clearly outlines steps involved applying varying levels intensity across mediums prior observing responses exhibited subsequent stages thereby allowing identification discerning nature cause action occurring\n",
      "\n",
      "\n",
      "Lastly looking closer specifics regarding measurements conducted reveal attempts quantifying factors influencing outcome variables recorded providing valuable insight distinguishing types interactions taking places course experimentation\n",
      "\n",
      "Therefore combining knowledge gained previously along side findings presented final conclusion drawn shall address original query posed initially asking method employed separating distinct categories actions exerted impacting subjects matter being researched herein.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Based purely textual clues extracted aforementioned passages key takeaway point stands apparent indicating presence systematic attempt made identifying & isolating purest form interaction stimulus-response relationship devoid contamination unwanted confounding elements potentially introduced changing composition solution tested ultimately facilitating accurate assessment true efficacy intervention proposed scenario examined case-by-case basis ensuring reliable conclusions reached end-of-day evaluation performed thoroughness manner undertaken entire exercise scientific inquiry pursued rigorous standards maintained utmost importance maintaining integrity validity obtained insights garnered efforts expended conducting comprehensive examination subject-matter-at-hand thoroughly explored multiple facets phenomenon interest understanding mechanisms operation principles governing complex interplay forces acting play pivotal roles determining ultimate fate success endeavors initiated pursuit advancement field expertise shared collective wisdom accumulated years dedicated service contributions numerous individuals working together collaborative spirit fostering progress advancing frontiers humanity benefiting society large scale impactful ways envisioned future generations inherit legacy built today tomorrow yesterday ongoing journey discovery exploration continuous learning improvement striving perfection relentless drive excellence unwavering commitment betterment lives people everywhere touched work accomplished countless others building bridges connecting communities worldwide creating brighter futures hope inspiration guidance support encouragement empowerment education enlightenment illumination shining lights darkness dispelling shadows casting warmth love kindness compassion empathy generosity selflessness altruism unity harmony cooperation mutual respect trust faith honesty transparency accountability responsibility stewardship sustainability environmental conservation social justice equality fairness freedom opportunity access inclusivity diversity equity accessibility affordability availability reliability consistency dependability predictability stability security safety peace prosperity happiness fulfillment joy satisfaction gratitude appreciation wonder awe curiosity creativity imagination innovation entrepreneurship leadership visionaries pioneers trailblazers game-changers disruptors transformers catalysts agents-change makers shapers movers influencers trendsetters setters pace-setters benchmark achievers record-breakers history-makers legendary icons heroes champions winners losers looser givers takers sharer receivers absorbers reflector reflextion refection reflection reflexions reflections reflective thinking critical thoughtfulness mindfulness awareness consciousness perception observation notice noticing noticed noting noted notes note-taking recording documenting archiving preserving storing safeguarding protecting conserving cherishing honoring valuing respecting appreciating admiring celebrating praising lauding extolling glorification worship reverence devotion dedication loyalty allegiance fidelity fealty pledge promise vow oath covenant agreement contract pact treaty alliance partnership collaboration synergy symbiosis fusion integration assimilation absorption incorporation amalgamation consolidation merger union federation association affiliation membership belonging inclusion participation engagement involvement activism advocacy voice speaking standing-up raising-awareness educating enlightening informing instructing guiding mentoring coaching advising counseling consulting assisting helping serving supporting empowering liberating freeing releasing unleashing untapping unlocking harnessing leveraging maximizing\n"
     ]
    }
   ],
   "source": [
    "response = chain.invoke(\"How do the authors distinguish between direct electrical effects on cells and indirect effects mediated through changes in the culture medium? \")\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e3418d-f210-4176-a1d3-b5fe41ff2cfd",
   "metadata": {},
   "source": [
    "* The model behaving expectedly but the output generation shows too much latency. Also, the answers needed to be complete an non-repetative.\n",
    "* I obserevded that, if the generated answer is shorter than max_tokens then it repeats the answer and if the answer is longer than max_token it generates incomplete answer.\n",
    "* This observation shows we need to fine-tune the model to behave and reponse in specific way."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57c8d1ac-d0e0-459f-b82a-1052b9c7d7b4",
   "metadata": {},
   "source": [
    "* No ReRanking is necessary as Here is only one document and has only few sections. If later adding more documents then Reranking might be necessary to choose from the extracted text."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bb77ad4-7211-48a5-984d-d32619d80226",
   "metadata": {},
   "source": [
    "### Finetunning -LoRA (GPT-OSS-20B)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "71ff5f27-55fe-4288-b75b-a4785c55fb7f",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "pip install \"trl>=0.20.0\" \"peft>=0.17.0\" \"transformers>=4.55.0\" trackio"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0f9ec02-b999-46e2-b553-78736ae73e2d",
   "metadata": {},
   "source": [
    "#### Dataset (PubMedQA + BioInstruct)\n",
    "\n",
    "* To generate answer in gpt-oss format I am finetunig the model using LoRA on PubMedQA dataset\n",
    "* The dataset need to converted in the form of [openAI harmony](https://cookbook.openai.com/articles/openai-harmony#concepts), because openai oss model has been trained to generate response in that prompt format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d1abaafb-edc2-48a6-a3e5-95a53ce88bdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datasets import load_dataset\n",
    "\n",
    "\n",
    "bioins_ds = load_dataset(\"bio-nlp-umass/bioinstruct\")\n",
    "pubmed_ds = load_dataset(\"qiaojin/PubMedQA\", \"pqa_artificial\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "685f629c-8efe-4d38-b64d-415617f18ad6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pubid', 'question', 'context', 'long_answer', 'final_decision']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed_dataset = pubmed_ds['train'].shuffle(seed=42).select(range(2000))\n",
    "pubmed_dataset.column_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "162cf86b-e95b-406f-88be-3420d4f48369",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['instruction', 'input', 'output']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioins_dataset = bioins_ds['train'].shuffle(seed=42).select(range(3000))\n",
    "bioins_dataset.column_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27f8c99f-834d-4fb9-bde8-abbec42e1ddf",
   "metadata": {},
   "source": [
    "#### Creating a function that reformates dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0830437e-8c20-42ff-a17d-4f0f69ef40fe",
   "metadata": {},
   "source": [
    "* The given PubMedQA dataset has question context, and answer but it doesn't have instruction that can be used in analysis channel of harmony template.\n",
    "* I am creating random analysis channel messages for model to use so, that it does not have identical thinking for each query.\n",
    "* FORMAT:```\n",
    "<|channel|>analysis<|message|>User asks: \"What is 2 + 2?\" Simple arithmetic. Provide answer.<|end|>\n",
    "<|start|>assistant<|channel|>final<|message|>2 + 2 = 4.<|return|>```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "265ded28-ff69-4b28-a940-36073e48b492",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "def reformat_to_harmony(dataset):\n",
    "\n",
    "    # randomly selecting the reasoning phrases to prevent overfitting\n",
    "    starts = [\"Assessing the impact of {topic} on patient outcomes.\",\n",
    "              \"Investigating mechanistic insights related to {topic}.\",\n",
    "              \"Reviewing recent findings regarding {topic} interventions.\",\n",
    "              \"Evaluating the reported correlations between biomarkers and {topic}.\",\n",
    "              \"Summarizing evidence from the study concerning {topic}.\",\n",
    "              \"Identifying methodological approaches used to study {topic}.\",\n",
    "              \"Interpreting clinical significance of results related to {topic}.\",\n",
    "             \"Analyzing clinical query regarding {topic}.\",\n",
    "            \"Reviewing provided medical abstract for evidence on {topic}.\",\n",
    "            \"Identifying key variables in the study of {topic}.\",\n",
    "            \"Examining the relationship between the study population and {topic}.\"\n",
    "    ]\n",
    "\n",
    "    second_step = [\n",
    "        \"Cross-referencing provided context for statistical significance and p-values.\",\n",
    "        \"Evaluating the experimental methodology and findings in the abstract.\",\n",
    "        \"Synthesizing the results presented in the provided medical text.\",\n",
    "        \"Checking for potential bias or study limitations in the context.\",\n",
    "        \"Comparing reported findings with established literature on {topic}.\",\n",
    "        \"Analyzing dosage, timing, or exposure variables in relation to {topic}.\",\n",
    "        \"Verifying statistical robustness and confidence intervals in the reported data.\",\n",
    "        \"Assessing whether conclusions are supported by the experimental evidence.\",\n",
    "        \"Highlighting key trends or patterns observed in the study results.\",\n",
    "        \"Checking for consistency between figures, tables, and text in the context.\",\n",
    "        \"Identifying potential confounding factors affecting the interpretation of {topic}.\"\n",
    "    ]\n",
    "\n",
    "    \n",
    "    # creating a topic from the question\n",
    "    question_starters = [\"Is\", \"Can\", \"Does\", \"Do\", \"Are\"]\n",
    "    topic = \" \".join((dataset['question'].split()[:]))\n",
    "    for st in question_starters:\n",
    "        if topic.startswith(st):\n",
    "            topic = topic.replace(st,\"\")\n",
    "            topic = topic.replace(\"?\",\"\")\n",
    "\n",
    "    # making analysis block message\n",
    "    reasoning = (f\"{random.choice(starts).format(topic=topic)} \"\n",
    "        f\"{random.choice(second_step)} \"\n",
    "        f\"Conclusion aligns with: {dataset['final_decision']}.\")\n",
    "\n",
    "    # Building the harmony format\n",
    "    system_msg = (\n",
    "        \"<|start|>system<|message|>\\n\"\n",
    "        \"reasoning_effort: high\\n\"\n",
    "        \"# Valid channels: analysis, final.\\n\"\n",
    "        \"You are a Biomedical Research Assistant. Your role is to help researchers with necessary details asked in question.\"\n",
    "        \"by accurately synthesizing literature and data. Use 'analysis' to evaluate \"\n",
    "        \"the evidence and 'final' for your structured response.\\n\"\n",
    "        \"<|end|>\\n\"\n",
    "    )\n",
    "\n",
    "    user_msg = f\"<|start|>user<|message|>\\nQUESTION: {dataset['question']}\\nCONTEXT: {dataset['context']}\\n<|end|>\\n\"\n",
    "\n",
    "    analysis_block = f\"<|start|>assistant<|channel|>analysis<|message|>\\n{reasoning}\\n<|end|>\\n\"\n",
    "\n",
    "    final_block = f\"<|start|>assistant<|channel|>final<|message|>\\n{dataset['long_answer']}\\n<|end|>\"\n",
    "\n",
    "    return {\"text\": system_msg + user_msg + analysis_block + final_block}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "894cb8ca-e28b-428d-8325-f4aeb356dfb2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "formated_pubmed_ds = pubmed_dataset.map(reformat_to_harmony)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d83025ce-6805-4da5-8e48-dd7acd9f0543",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<|start|>system<|message|>\\nreasoning_effort: high\\n# Valid channels: analysis, final.\\nYou are a Biomedical Research Assistant. Your role is to help researchers with necessary details asked in question.by accurately synthesizing literature and data. Use \\'analysis\\' to evaluate the evidence and \\'final\\' for your structured response.\\n<|end|>\\n<|start|>user<|message|>\\nQUESTION: Are keratin 8 Y54H and G62C mutations associated with inflammatory bowel disease?\\nCONTEXT: {\\'contexts\\': [\\'Keratin 8 is a major component of intermediate filaments in single-layered epithelia of the gastrointestinal tract. Keratin 8 deficient mice display signs of colitis and diarrhoea characteristic for inflammatory bowel disease. Very recently, two keratin 8 mutations, Y54H and G62C, were identified.\\', \\'We investigated if these keratin 8 missense mutations were associated with inflammatory bowel disease.\\', \"In total, 217 German patients with Crohn\\' s disease, 131 German patients with ulcerative colitis, and 560 German control subjects were enrolled in this study.\", \\'Samples were analysed by PCR amplification and subsequent melting curve analysis using fluorescence resonance energy transfer probes.\\', \"The G62C mutation was detected in five (2.3%) patients presenting with Crohn\\'s disease and in three (2.3%) with ulcerative colitis. In comparison, 9 (1.6%) out of 560 controls were heterozygous for this mutation. No patient or control was homozygous for this mutation. Patients carrying one mutant allele did not show any noticeable characteristics in their corresponding phenotype. In contrast, the Y54H mutation was observed in neither any of the 348 patients with inflammatory bowel disease nor in any control subject.\"], \\'labels\\': [\\'BACKGROUND\\', \\'OBJECTIVE\\', \\'METHODS\\', \\'METHODS\\', \\'RESULTS\\'], \\'meshes\\': [\\'Adolescent\\', \\'Adult\\', \\'Aged\\', \\'Aged, 80 and over\\', \\'Case-Control Studies\\', \\'DNA Mutational Analysis\\', \\'Female\\', \\'Heterozygote\\', \\'Humans\\', \\'Inflammatory Bowel Diseases\\', \\'Keratin-8\\', \\'Keratins\\', \\'Male\\', \\'Middle Aged\\', \\'Mutation, Missense\\']}\\n<|end|>\\n<|start|>assistant<|channel|>analysis<|message|>\\nReviewing provided medical abstract for evidence on  keratin 8 Y54H and G62C mutations associated with inflammatory bowel disease. Verifying statistical robustness and confidence intervals in the reported data. Conclusion aligns with: no.\\n<|end|>\\n<|start|>assistant<|channel|>final<|message|>\\nOur data indicate that both keratin 8 mutations, G62C and Y54H, do not play a relevant pathogenic role in inflammatory bowel disease.\\n<|end|>'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "formated_pubmed_ds[0]['text']"
   ]
  },
  {
   "cell_type": "raw",
   "id": "38a01172-cee3-44ce-bdd4-e48d7fdeda24",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "formated_pubmed_ds.push_to_hub(repo_id=\"Vrjb/BioRAG\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65fc8393-c267-482a-b84a-b41e9d58e5c2",
   "metadata": {},
   "source": [
    "#### Now with BioInstruct dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7a599180-a302-412e-a41b-baf425c22699",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'4.',\n",
       " 'Advise',\n",
       " 'Analyze',\n",
       " 'Answer',\n",
       " 'Ask',\n",
       " 'Assess',\n",
       " 'Assist',\n",
       " 'Based',\n",
       " 'Break',\n",
       " 'Briefly',\n",
       " 'Calculate',\n",
       " 'Can',\n",
       " 'Categorize',\n",
       " 'Choose',\n",
       " 'Classify',\n",
       " 'Compare',\n",
       " 'Complete',\n",
       " 'Considering',\n",
       " 'Contrast',\n",
       " 'Convert',\n",
       " 'Correct',\n",
       " 'Create',\n",
       " 'Critically',\n",
       " 'Decide',\n",
       " 'Define',\n",
       " 'Dentists',\n",
       " 'Describe',\n",
       " 'Design',\n",
       " 'Detect',\n",
       " 'Determine',\n",
       " 'Develop',\n",
       " 'Differentiate',\n",
       " 'Discuss',\n",
       " 'Draft',\n",
       " 'Elaborate',\n",
       " 'Estimate',\n",
       " 'Evaluate',\n",
       " 'Explain',\n",
       " 'Explain,',\n",
       " 'Express',\n",
       " 'Extract',\n",
       " 'Find',\n",
       " 'For',\n",
       " 'Formulate',\n",
       " 'From',\n",
       " 'Generate',\n",
       " 'Give',\n",
       " 'Given',\n",
       " 'Help',\n",
       " 'Highlight',\n",
       " 'How',\n",
       " 'Identify',\n",
       " 'In',\n",
       " 'Indicate',\n",
       " 'Interpret',\n",
       " 'Is',\n",
       " 'List',\n",
       " 'Looking',\n",
       " 'Make',\n",
       " 'Mention',\n",
       " 'Name',\n",
       " 'Offer',\n",
       " 'Organize',\n",
       " 'Outline',\n",
       " 'Paraphrase',\n",
       " 'Parse',\n",
       " 'Perform',\n",
       " 'Please',\n",
       " 'Predict',\n",
       " 'Prepare',\n",
       " 'Present',\n",
       " 'Propose',\n",
       " 'Provide',\n",
       " 'Provided',\n",
       " 'Providing',\n",
       " 'Rank',\n",
       " 'Re-phrase',\n",
       " 'Re-write',\n",
       " 'Read',\n",
       " 'Rearrange',\n",
       " 'Recognize',\n",
       " 'Recommend',\n",
       " 'Request',\n",
       " 'Research',\n",
       " 'Restate',\n",
       " 'Review',\n",
       " 'Reword',\n",
       " 'Rewrite',\n",
       " 'Simplify',\n",
       " 'State',\n",
       " 'Suggest',\n",
       " 'Sum',\n",
       " 'Summarize',\n",
       " 'Transform',\n",
       " 'Translate',\n",
       " 'Using',\n",
       " 'Verify',\n",
       " 'Weigh',\n",
       " 'What',\n",
       " 'Which',\n",
       " 'With',\n",
       " 'Write'}"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s= set()\n",
    "for i in bioins_dataset['instruction']:\n",
    "    s.add(i.split()[0])\n",
    "\n",
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bab3f6fd-569c-4fab-aa0a-ce73255472cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reformating_bioins(dataset):\n",
    "    instruction = dataset['instruction']\n",
    "    first_word = instruction.split()[0].rstrip(',.:')\n",
    "\n",
    "    ## In this dataset i dont have question to format the dataset in harmony, so using instructions \n",
    "    ## and taking it's intent like generate, describe and convert it into summarizing, generating.\n",
    "    \n",
    "    # defining intent group\n",
    "    intent_map = {\n",
    "        # Commands & Actions\n",
    "        \"Summarize\": \"Summarizing\", \"Analyze\": \"Analyzing\", \"Explain\": \"Explaining\",\n",
    "        \"Generate\": \"Generating\", \"Extract\": \"Extracting\", \"Compare\": \"Comparing\",\n",
    "        \"Describe\": \"Describing\", \"Evaluate\": \"Evaluating\", \"Classify\": \"Classifying\",\n",
    "        \"Calculate\": \"Calculating\", \"Predict\": \"Predicting\", \"Interpret\": \"Interpreting\",\n",
    "        \n",
    "        # Requests & Polite starts\n",
    "        \"Please\": \"Processing request\", \"Can\": \"Assessing capability\", \n",
    "        \"Ask\": \"Responding\", \"Suggest\": \"Formulating suggestion\",\n",
    "        \n",
    "        # Contextual & Conditional starts\n",
    "        \"Given\": \"Analyzing provided data\", \"Considering\": \"Weighing factors\", \n",
    "        \"Based\": \"Evaluating based on evidence\", \"With\": \"Reviewing context\",\n",
    "        \n",
    "        # Question-based starts\n",
    "        \"What\": \"Identifying\", \"How\": \"Detailing process\", \"Which\": \"Selecting\",\n",
    "        \"Is\": \"Verifying\"\n",
    "    }\n",
    "\n",
    "    ## mapping, and if the first word is not present in map we replace it with following\n",
    "    action = intent_map.get(first_word,\"Addressing biomedical task\")\n",
    "\n",
    "    # building analysis message block\n",
    "    reasoning = (\n",
    "        f\"{action} for the research query. \"\n",
    "        \"Step 1: Identifying key medical entities and constraints. \"\n",
    "        \"Step 2: Synthesizing response based on biomedical standards.\"\n",
    "    )\n",
    "    \n",
    "    # Building the harmony format\n",
    "    system_msg = (\n",
    "        \"<|start|>system<|message|>\\n\"\n",
    "        \"reasoning_effort: high\\n\"\n",
    "        \"# Valid channels: analysis, final.\\n\"\n",
    "        \"You are a Biomedical Research Assistant. Your role is to help researchers with necessary details asked in question.\"\n",
    "        \"by accurately synthesizing literature and data. Use 'analysis' to evaluate \"\n",
    "        \"the evidence and 'final' for your structured response.\\n\"\n",
    "        \"<|end|>\\n\"\n",
    "    )\n",
    "\n",
    "    user_msg = f\"<|start|>user<|message|>\\nQUESTION: {dataset['instruction']}\\nCONTEXT: {dataset['input']}\\n<|end|>\\n\"\n",
    "    analysis_block = f\"<|start|>assistant<|channel|>analysis<|message|>\\n{reasoning}\\n<|end|>\\n\"\n",
    "    final_block = f\"<|start|>assistant<|channel|>final<|message|>\\n{dataset['output']}\\n<|end|>\"\n",
    "    \n",
    "    return {\"text\": system_msg + user_msg + analysis_block + final_block}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f2f9810c-f16a-4237-8228-151b49f1d598",
   "metadata": {},
   "outputs": [],
   "source": [
    "formated_bioins_ds=bioins_dataset.map(reformating_bioins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1ebfb767-00ee-46f7-90c9-413a6ae07f2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(formated_pubmed_ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f8dd3756-266a-4810-a076-680db53871ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5000"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merging dataset\n",
    "\n",
    "def merge(pub, bio):\n",
    "    lis = {pub[i]['text'] for i in range(len(pub))}\n",
    "    for j in range(len(bio)):\n",
    "        lis.add(bio[j]['text'])\n",
    "    return lis\n",
    "\n",
    "full_dataset = merge(pub=formated_pubmed_ds, bio=formated_bioins_ds)\n",
    "len(full_dataset)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "25fbc149-c2e3-4bb0-badd-0655e0f62ea5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from datasets import Dataset\n",
    "\n",
    "ds = Dataset.from_dict({\"text\":full_dataset})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "fe330d51-8614-429e-a28e-e7fc5ec80484",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Uploading the dataset shards:   0%|          | 0/1 [00:00<?, ? shards/s]\n",
      "Creating parquet from Arrow format: 100%|██████████| 1/1 [00:00<00:00, 33.42ba/s]\n",
      "Processing Files (0 / 0): |          |  0.00B /  0.00B            \n",
      "Processing Files (0 / 1):  15%|█▍        |  528kB / 3.55MB,  880kB/s  \n",
      "Processing Files (0 / 1):  45%|████▍     | 1.58MB / 3.55MB, 1.59MB/s  \n",
      "Processing Files (0 / 1):  89%|████████▉ | 3.17MB / 3.55MB, 2.64MB/s  \n",
      "Processing Files (1 / 1): 100%|██████████| 3.55MB / 3.55MB, 2.22MB/s  \n",
      "Processing Files (1 / 1): 100%|██████████| 3.55MB / 3.55MB, 1.97MB/s  \n",
      "New Data Upload: 100%|██████████| 3.55MB / 3.55MB, 1.97MB/s  \n",
      "Uploading the dataset shards: 100%|██████████| 1/1 [00:02<00:00,  2.62s/ shards]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CommitInfo(commit_url='https://huggingface.co/datasets/Vrjb/BioRAG/commit/172c6f3237a51d4c83ca514f3c65d6e2bfe27ad5', commit_message='Upload dataset', commit_description='', oid='172c6f3237a51d4c83ca514f3c65d6e2bfe27ad5', pr_url=None, repo_url=RepoUrl('https://huggingface.co/datasets/Vrjb/BioRAG', endpoint='https://huggingface.co', repo_type='dataset', repo_id='Vrjb/BioRAG'), pr_revision=None, pr_num=None)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds.push_to_hub(\"vrjb/BioRAG\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3b7918c-ca89-45a9-b39e-ce0587ee8be2",
   "metadata": {},
   "source": [
    "### I have created and merged two dataset to fintune the GPT-OSS model in to required prompt template"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "641af63f-d704-4c94-8c3d-ffc89160881d",
   "metadata": {},
   "source": [
    "# Fine-Tunning Using LoRA\n",
    "* [finetune with gpt-oss and HF](https://cookbook.openai.com/articles/gpt-oss/fine-tune-transfomers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "dca13c65-96f4-4028-9248-347391a68628",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Generating train split: 100%|██████████| 5000/5000 [00:00<00:00, 95600.16 examples/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'text': \"<|start|>system<|message|>\\nreasoning_effort: high\\n# Valid channels: analysis, final.\\nYou are a Biomedical Research Assistant. Your role is to help researchers with necessary details asked in question.by accurately synthesizing literature and data. Use 'analysis' to evaluate the evidence and 'final' for your structured response.\\n<|end|>\\n<|start|>user<|message|>\\nQUESTION: Explain how an MRI machine works in simple terms.\\nCONTEXT: MRI (Magnetic Resonance Imaging) is an imaging technique that uses magnetic fields and radio waves to create detailed images of the inside of the body. It works by aligning hydrogen atoms in the body with a strong magnetic field and then applying radiofrequency pulses. The atoms emit signals, which are detected by a scanner and processed to create an image.\\n<|end|>\\n<|start|>assistant<|channel|>analysis<|message|>\\nExplaining for the research query. Step 1: Identifying key medical entities and constraints. Step 2: Synthesizing response based on biomedical standards.\\n<|end|>\\n<|start|>assistant<|channel|>final<|message|>\\nAn MRI machine creates images of the body's insides by using strong magnets and radio waves. It lines up hydrogen atoms in the body, applies radio pulses, and then listens for the signals the atoms give off. These signals are used to create detailed pictures of the inside of your body.\\n<|end|>\"}"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Loading the dataset\n",
    "\n",
    "from datasets import load_dataset\n",
    "\n",
    "dataset = load_dataset(\"vrjb/BioRAG\", split='train')\n",
    "dataset[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f92c2e75-d584-4100-9d3c-37e59ec3eec7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.\n",
      "Knowledge cutoff: 2024-06\n",
      "Current date: 2026-01-07\n",
      "\n",
      "Reasoning: medium\n",
      "\n",
      "# Valid channels: analysis, commentary, final. Channel must be included for every message.<|end|>\n"
     ]
    }
   ],
   "source": [
    "# Loading Tokenizer\n",
    "\n",
    "from transformers import AutoTokenizer\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"openai/gpt-oss-20b\")\n",
    "\n",
    "# Applying chat template methhod to format the messages\n",
    "\n",
    "messages = dataset[0][\"text\"]\n",
    "conversation = tokenizer.apply_chat_template(messages, tokenize=False)\n",
    "print(conversation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "079d9030-3ed5-4aee-8da2-bb4bf0e56c3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to initialize NVML: Unknown Error\n"
     ]
    }
   ],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b2ec339f-20d1-4021-b44d-0ba4a76c91ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading checkpoint shards:  67%|██████▋   | 2/3 [00:08<00:04,  4.06s/it]\n"
     ]
    },
    {
     "ename": "OutOfMemoryError",
     "evalue": "CUDA out of memory. Tried to allocate 1.98 GiB. GPU 0 has a total capacity of 79.25 GiB of which 1.52 GiB is free. Including non-PyTorch memory, this process has 63.43 GiB memory in use. Of the allocated memory 60.30 GiB is allocated by PyTorch, and 2.63 GiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mOutOfMemoryError\u001b[39m                          Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[51]\u001b[39m\u001b[32m, line 15\u001b[39m\n\u001b[32m      6\u001b[39m quantization_config = Mxfp4Config(dequantize=\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[32m      7\u001b[39m model_kwargs = \u001b[38;5;28mdict\u001b[39m(\n\u001b[32m      8\u001b[39m     attn_implementation=\u001b[33m\"\u001b[39m\u001b[33meager\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m      9\u001b[39m     dtype=torch.bfloat16,\n\u001b[32m   (...)\u001b[39m\u001b[32m     12\u001b[39m     device_map=\u001b[33m\"\u001b[39m\u001b[33mauto\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     13\u001b[39m )\n\u001b[32m---> \u001b[39m\u001b[32m15\u001b[39m model = \u001b[43mAutoModelForCausalLM\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfrom_pretrained\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mopenai/gpt-oss-20b\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mmodel_kwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/models/auto/auto_factory.py:604\u001b[39m, in \u001b[36m_BaseAutoModelClass.from_pretrained\u001b[39m\u001b[34m(cls, pretrained_model_name_or_path, *model_args, **kwargs)\u001b[39m\n\u001b[32m    602\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m model_class.config_class == config.sub_configs.get(\u001b[33m\"\u001b[39m\u001b[33mtext_config\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[32m    603\u001b[39m         config = config.get_text_config()\n\u001b[32m--> \u001b[39m\u001b[32m604\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmodel_class\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfrom_pretrained\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    605\u001b[39m \u001b[43m        \u001b[49m\u001b[43mpretrained_model_name_or_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43mmodel_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mhub_kwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\n\u001b[32m    606\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    607\u001b[39m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m    608\u001b[39m     \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mUnrecognized configuration class \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mconfig.\u001b[34m__class__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m for this kind of AutoModel: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mcls\u001b[39m.\u001b[34m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m    609\u001b[39m     \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mModel type should be one of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m, \u001b[39m\u001b[33m'\u001b[39m.join(c.\u001b[34m__name__\u001b[39m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mfor\u001b[39;00m\u001b[38;5;250m \u001b[39mc\u001b[38;5;250m \u001b[39m\u001b[38;5;129;01min\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28mcls\u001b[39m._model_mapping)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    610\u001b[39m )\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/modeling_utils.py:277\u001b[39m, in \u001b[36mrestore_default_dtype.<locals>._wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    275\u001b[39m old_dtype = torch.get_default_dtype()\n\u001b[32m    276\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m277\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    278\u001b[39m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[32m    279\u001b[39m     torch.set_default_dtype(old_dtype)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/modeling_utils.py:5048\u001b[39m, in \u001b[36mPreTrainedModel.from_pretrained\u001b[39m\u001b[34m(cls, pretrained_model_name_or_path, config, cache_dir, ignore_mismatched_sizes, force_download, local_files_only, token, revision, use_safetensors, weights_only, *model_args, **kwargs)\u001b[39m\n\u001b[32m   5038\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m dtype_orig \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m   5039\u001b[39m         torch.set_default_dtype(dtype_orig)\n\u001b[32m   5041\u001b[39m     (\n\u001b[32m   5042\u001b[39m         model,\n\u001b[32m   5043\u001b[39m         missing_keys,\n\u001b[32m   5044\u001b[39m         unexpected_keys,\n\u001b[32m   5045\u001b[39m         mismatched_keys,\n\u001b[32m   5046\u001b[39m         offload_index,\n\u001b[32m   5047\u001b[39m         error_msgs,\n\u001b[32m-> \u001b[39m\u001b[32m5048\u001b[39m     ) = \u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_load_pretrained_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   5049\u001b[39m \u001b[43m        \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5050\u001b[39m \u001b[43m        \u001b[49m\u001b[43mstate_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5051\u001b[39m \u001b[43m        \u001b[49m\u001b[43mcheckpoint_files\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5052\u001b[39m \u001b[43m        \u001b[49m\u001b[43mpretrained_model_name_or_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5053\u001b[39m \u001b[43m        \u001b[49m\u001b[43mignore_mismatched_sizes\u001b[49m\u001b[43m=\u001b[49m\u001b[43mignore_mismatched_sizes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5054\u001b[39m \u001b[43m        \u001b[49m\u001b[43msharded_metadata\u001b[49m\u001b[43m=\u001b[49m\u001b[43msharded_metadata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5055\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdevice_map\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdevice_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5056\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdisk_offload_folder\u001b[49m\u001b[43m=\u001b[49m\u001b[43moffload_folder\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5057\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5058\u001b[39m \u001b[43m        \u001b[49m\u001b[43mhf_quantizer\u001b[49m\u001b[43m=\u001b[49m\u001b[43mhf_quantizer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5059\u001b[39m \u001b[43m        \u001b[49m\u001b[43mkeep_in_fp32_regex\u001b[49m\u001b[43m=\u001b[49m\u001b[43mkeep_in_fp32_regex\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5060\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdevice_mesh\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdevice_mesh\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5061\u001b[39m \u001b[43m        \u001b[49m\u001b[43mkey_mapping\u001b[49m\u001b[43m=\u001b[49m\u001b[43mkey_mapping\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5062\u001b[39m \u001b[43m        \u001b[49m\u001b[43mweights_only\u001b[49m\u001b[43m=\u001b[49m\u001b[43mweights_only\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   5063\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   5064\u001b[39m \u001b[38;5;66;03m# make sure token embedding weights are still tied if needed\u001b[39;00m\n\u001b[32m   5065\u001b[39m model.tie_weights()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/modeling_utils.py:5468\u001b[39m, in \u001b[36mPreTrainedModel._load_pretrained_model\u001b[39m\u001b[34m(cls, model, state_dict, checkpoint_files, pretrained_model_name_or_path, ignore_mismatched_sizes, sharded_metadata, device_map, disk_offload_folder, dtype, hf_quantizer, keep_in_fp32_regex, device_mesh, key_mapping, weights_only)\u001b[39m\n\u001b[32m   5465\u001b[39m         args_list = logging.tqdm(args_list, desc=\u001b[33m\"\u001b[39m\u001b[33mLoading checkpoint shards\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m   5467\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m args \u001b[38;5;129;01min\u001b[39;00m args_list:\n\u001b[32m-> \u001b[39m\u001b[32m5468\u001b[39m         _error_msgs, disk_offload_index = \u001b[43mload_shard_file\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   5469\u001b[39m         error_msgs += _error_msgs\n\u001b[32m   5471\u001b[39m \u001b[38;5;66;03m# Save offloaded index if needed\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/modeling_utils.py:843\u001b[39m, in \u001b[36mload_shard_file\u001b[39m\u001b[34m(args)\u001b[39m\n\u001b[32m    841\u001b[39m \u001b[38;5;66;03m# Skip it with fsdp on ranks other than 0\u001b[39;00m\n\u001b[32m    842\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (is_fsdp_enabled() \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_local_dist_rank_0() \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_quantized):\n\u001b[32m--> \u001b[39m\u001b[32m843\u001b[39m     disk_offload_index = \u001b[43m_load_state_dict_into_meta_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    844\u001b[39m \u001b[43m        \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    845\u001b[39m \u001b[43m        \u001b[49m\u001b[43mstate_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    846\u001b[39m \u001b[43m        \u001b[49m\u001b[43mshard_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    847\u001b[39m \u001b[43m        \u001b[49m\u001b[43mreverse_key_renaming_mapping\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    848\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdevice_map\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdevice_map\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    849\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdisk_offload_folder\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdisk_offload_folder\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    850\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdisk_offload_index\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdisk_offload_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    851\u001b[39m \u001b[43m        \u001b[49m\u001b[43mhf_quantizer\u001b[49m\u001b[43m=\u001b[49m\u001b[43mhf_quantizer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    852\u001b[39m \u001b[43m        \u001b[49m\u001b[43mkeep_in_fp32_regex\u001b[49m\u001b[43m=\u001b[49m\u001b[43mkeep_in_fp32_regex\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    853\u001b[39m \u001b[43m        \u001b[49m\u001b[43mdevice_mesh\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdevice_mesh\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    854\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    856\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m error_msgs, disk_offload_index\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/torch/utils/_contextlib.py:120\u001b[39m, in \u001b[36mcontext_decorator.<locals>.decorate_context\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    117\u001b[39m \u001b[38;5;129m@functools\u001b[39m.wraps(func)\n\u001b[32m    118\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mdecorate_context\u001b[39m(*args, **kwargs):\n\u001b[32m    119\u001b[39m     \u001b[38;5;28;01mwith\u001b[39;00m ctx_factory():\n\u001b[32m--> \u001b[39m\u001b[32m120\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/modeling_utils.py:774\u001b[39m, in \u001b[36m_load_state_dict_into_meta_model\u001b[39m\u001b[34m(model, state_dict, shard_file, reverse_renaming_mapping, device_map, disk_offload_folder, disk_offload_index, hf_quantizer, keep_in_fp32_regex, device_mesh)\u001b[39m\n\u001b[32m    770\u001b[39m     _load_parameter_into_model(model, param_name, param.to(param_device))\n\u001b[32m    772\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    773\u001b[39m     \u001b[38;5;66;03m# TODO naming is stupid it loads it as well\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m774\u001b[39m     \u001b[43mhf_quantizer\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcreate_quantized_param\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam_device\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    776\u001b[39m     \u001b[38;5;66;03m# For quantized modules with FSDP/DeepSpeed Stage 3, we need to quantize the parameter on the GPU\u001b[39;00m\n\u001b[32m    777\u001b[39m     \u001b[38;5;66;03m# and then cast it to CPU to avoid excessive memory usage on each GPU\u001b[39;00m\n\u001b[32m    778\u001b[39m     \u001b[38;5;66;03m# in comparison to the sharded model across GPUs.\u001b[39;00m\n\u001b[32m    779\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m is_fsdp_enabled() \u001b[38;5;129;01mor\u001b[39;00m is_deepspeed_zero3_enabled():\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/quantizers/quantizer_mxfp4.py:246\u001b[39m, in \u001b[36mMxfp4HfQuantizer.create_quantized_param\u001b[39m\u001b[34m(self, model, param_value, param_name, target_device, **kwargs)\u001b[39m\n\u001b[32m    242\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.quantization_config.dequantize:\n\u001b[32m    243\u001b[39m     \u001b[38;5;66;03m# dq_param_name is the name of the parameter without the blocks or scales suffix, it's used in this case since we don't switch linears\u001b[39;00m\n\u001b[32m    244\u001b[39m     \u001b[38;5;66;03m# so we only have the original param name\u001b[39;00m\n\u001b[32m    245\u001b[39m     dq_param_name = param_name[: -\u001b[38;5;28mlen\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33m_blocks\u001b[39m\u001b[33m\"\u001b[39m)]\n\u001b[32m--> \u001b[39m\u001b[32m246\u001b[39m     \u001b[43mdequantize\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodule\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparam_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_device\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdq_param_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mshard_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    247\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    248\u001b[39m     load_and_swizzle_mxfp4(\n\u001b[32m    249\u001b[39m         module,\n\u001b[32m    250\u001b[39m         param_name,\n\u001b[32m   (...)\u001b[39m\u001b[32m    254\u001b[39m         **shard_kwargs,\n\u001b[32m    255\u001b[39m     )\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/integrations/mxfp4.py:351\u001b[39m, in \u001b[36mdequantize\u001b[39m\u001b[34m(module, param_name, param_value, target_device, dq_param_name, **kwargs)\u001b[39m\n\u001b[32m    349\u001b[39m \u001b[38;5;28msetattr\u001b[39m(module, param_name.rsplit(\u001b[33m\"\u001b[39m\u001b[33m.\u001b[39m\u001b[33m\"\u001b[39m, \u001b[32m1\u001b[39m)[\u001b[32m1\u001b[39m], param_value)\n\u001b[32m    350\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(module, blocks_attr) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(module, scales_attr):\n\u001b[32m--> \u001b[39m\u001b[32m351\u001b[39m     dequantized = \u001b[43mconvert_moe_packed_tensors\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mmodule\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mblocks_attr\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mmodule\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscales_attr\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    352\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m target_device == \u001b[33m\"\u001b[39m\u001b[33mcpu\u001b[39m\u001b[33m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m torch.cuda.is_available():\n\u001b[32m    353\u001b[39m         torch.cuda.empty_cache()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/work/CellSense/X20251016-CellSense-VB/cell/lib/python3.11/site-packages/transformers/integrations/mxfp4.py:140\u001b[39m, in \u001b[36mconvert_moe_packed_tensors\u001b[39m\u001b[34m(blocks, scales, dtype, rows_per_chunk)\u001b[39m\n\u001b[32m    138\u001b[39m \u001b[38;5;66;03m# nibble indices -> int64\u001b[39;00m\n\u001b[32m    139\u001b[39m idx_lo = (blk & \u001b[32m0x0F\u001b[39m).to(torch.long)\n\u001b[32m--> \u001b[39m\u001b[32m140\u001b[39m idx_hi = \u001b[43m(\u001b[49m\u001b[43mblk\u001b[49m\u001b[43m \u001b[49m\u001b[43m>>\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m4\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtorch\u001b[49m\u001b[43m.\u001b[49m\u001b[43mlong\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    142\u001b[39m sub = out[r0:r1]\n\u001b[32m    143\u001b[39m sub[:, \u001b[32m0\u001b[39m::\u001b[32m2\u001b[39m] = lut[idx_lo]\n",
      "\u001b[31mOutOfMemoryError\u001b[39m: CUDA out of memory. Tried to allocate 1.98 GiB. GPU 0 has a total capacity of 79.25 GiB of which 1.52 GiB is free. Including non-PyTorch memory, this process has 63.43 GiB memory in use. Of the allocated memory 60.30 GiB is allocated by PyTorch, and 2.63 GiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)"
     ]
    }
   ],
   "source": [
    "# Preparing the model\n",
    "\n",
    "import torch\n",
    "from transformers import AutoModelForCausalLM, Mxfp4Config\n",
    "\n",
    "quantization_config = Mxfp4Config(dequantize=True)\n",
    "model_kwargs = dict(\n",
    "    attn_implementation=\"eager\",\n",
    "    dtype=torch.bfloat16,\n",
    "    quantization_config=quantization_config,\n",
    "    use_cache=False,\n",
    "    device_map=\"auto\",\n",
    ")\n",
    "\n",
    "model = AutoModelForCausalLM.from_pretrained(\"openai/gpt-oss-20b\", **model_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2d48376-7022-4a7e-a33b-7409959139fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing the Raw model's configuration\n",
    "messages = [\n",
    "    {\"role\": \"user\", \"content\": \"¿Cuál es el capital de Australia?\"},\n",
    "]\n",
    "\n",
    "input_ids = tokenizer.apply_chat_template(\n",
    "    messages,\n",
    "    add_generation_prompt=True,\n",
    "    return_tensors=\"pt\",\n",
    ").to(model.device)\n",
    "\n",
    "output_ids = model.generate(input_ids, max_new_tokens=512)\n",
    "response = tokenizer.batch_decode(output_ids)[0]\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d35efd7-2090-4f3b-ae57-2c398533677f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from peft import LoraConfig, get_peft_model\n",
    "\n",
    "peft_config = LoraConfig(\n",
    "    r=8,\n",
    "    lora_alpha=16,\n",
    "    target_modules=\"all-linear\",\n",
    "    target_parameters=[\n",
    "        \"7.mlp.experts.gate_up_proj\",\n",
    "        \"7.mlp.experts.down_proj\",\n",
    "        \"15.mlp.experts.gate_up_proj\",\n",
    "        \"15.mlp.experts.down_proj\",\n",
    "        \"23.mlp.experts.gate_up_proj\",\n",
    "        \"23.mlp.experts.down_proj\",\n",
    "    ],\n",
    ")\n",
    "peft_model = get_peft_model(model, peft_config)\n",
    "peft_model.config.use_cache = False  # Silence warnings, cache is incompatible with checkpointing\n",
    "peft_model.print_trainable_parameters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26650dbc-968f-479a-8d69-55cbd62d1d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "from trl import SFTConfig\n",
    "\n",
    "training_args = SFTConfig(\n",
    "    learning_rate=2e-4,\n",
    "    optim=\"adamw_torch\",\n",
    "    gradient_checkpointing=True,\n",
    "    num_train_epochs=1,\n",
    "    logging_steps=1,\n",
    "    per_device_train_batch_size=1,\n",
    "    gradient_accumulation_steps=2,\n",
    "    max_length=1024,\n",
    "    warmup_ratio=0.03,\n",
    "    lr_scheduler_type=\"cosine_with_min_lr\",\n",
    "    lr_scheduler_kwargs={\"min_lr_rate\": 0.1},\n",
    "    output_dir=\"gpt-oss-20b-BioRAG\",\n",
    "    report_to=\"trackio\",\n",
    "    push_to_hub=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "id": "209cb6d7-3650-40c5-9358-d6e6617e26dc",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "from trl import SFTTrainer\n",
    "\n",
    "trainer = SFTTrainer(\n",
    "    model=peft_model,\n",
    "    args=training_args,\n",
    "    train_dataset=dataset,\n",
    "    processing_class=tokenizer,\n",
    ")\n",
    "trainer.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6110bdf-d89c-4a57-a4c0-8b6c91361ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76ba8302-b0c3-416d-b01a-ae1a7a2353c3",
   "metadata": {},
   "source": [
    "#### Key points for the error \n",
    "* The reason of the above error is `quantization_config = Mxfp4Config(dequantize=True)` line of code. It is upcasting meaning it is changing it's precision from MXFP4 to bffloat16.\n",
    "* The step is necessary for stable gradient accumulation and accuracy for better learning according to this [Fine-Tuning gpt-oss for Accuracy and Performance with Quantization Aware Training](https://developer.nvidia.com/blog/fine-tuning-gpt-oss-for-accuracy-and-performance-with-quantization-aware-training/)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (cell)",
   "language": "python",
   "name": "langchain_classic"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
